<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head><meta http-equiv="content-type" content="text/html; charset=UTF-8" />
<title>Normalizer xref</title>
<link type="text/css" rel="stylesheet" href="../../../../stylesheet.css" />
</head>
<body>
<div id="overview"><a href="../../../../../apidocs/org/biopax/paxtools/normalizer/Normalizer.html">View Javadoc</a></div><pre>
<a class="jxr_linenumber" name="L1" href="#L1">1</a>   <strong class="jxr_keyword">package</strong> org.biopax.paxtools.normalizer;
<a class="jxr_linenumber" name="L2" href="#L2">2</a>   
<a class="jxr_linenumber" name="L3" href="#L3">3</a>   <strong class="jxr_keyword">import</strong> org.apache.commons.lang3.StringUtils;
<a class="jxr_linenumber" name="L4" href="#L4">4</a>   <strong class="jxr_keyword">import</strong> org.biopax.paxtools.controller.ModelUtils;
<a class="jxr_linenumber" name="L5" href="#L5">5</a>   <strong class="jxr_keyword">import</strong> org.biopax.paxtools.controller.ShallowCopy;
<a class="jxr_linenumber" name="L6" href="#L6">6</a>   <strong class="jxr_keyword">import</strong> org.biopax.paxtools.converter.LevelUpgrader;
<a class="jxr_linenumber" name="L7" href="#L7">7</a>   <strong class="jxr_keyword">import</strong> org.biopax.paxtools.io.SimpleIOHandler;
<a class="jxr_linenumber" name="L8" href="#L8">8</a>   <strong class="jxr_keyword">import</strong> org.biopax.paxtools.model.BioPAXElement;
<a class="jxr_linenumber" name="L9" href="#L9">9</a>   <strong class="jxr_keyword">import</strong> org.biopax.paxtools.model.BioPAXLevel;
<a class="jxr_linenumber" name="L10" href="#L10">10</a>  <strong class="jxr_keyword">import</strong> org.biopax.paxtools.model.Model;
<a class="jxr_linenumber" name="L11" href="#L11">11</a>  <strong class="jxr_keyword">import</strong> org.biopax.paxtools.model.level3.*;
<a class="jxr_linenumber" name="L12" href="#L12">12</a>  <strong class="jxr_keyword">import</strong> org.biopax.paxtools.util.BPCollections;
<a class="jxr_linenumber" name="L13" href="#L13">13</a>  <strong class="jxr_keyword">import</strong> org.biopax.paxtools.util.ClassFilterSet;
<a class="jxr_linenumber" name="L14" href="#L14">14</a>  <strong class="jxr_keyword">import</strong> org.slf4j.Logger;
<a class="jxr_linenumber" name="L15" href="#L15">15</a>  <strong class="jxr_keyword">import</strong> org.slf4j.LoggerFactory;
<a class="jxr_linenumber" name="L16" href="#L16">16</a>  
<a class="jxr_linenumber" name="L17" href="#L17">17</a>  <strong class="jxr_keyword">import</strong> java.io.ByteArrayInputStream;
<a class="jxr_linenumber" name="L18" href="#L18">18</a>  <strong class="jxr_keyword">import</strong> java.io.ByteArrayOutputStream;
<a class="jxr_linenumber" name="L19" href="#L19">19</a>  <strong class="jxr_keyword">import</strong> java.io.InputStream;
<a class="jxr_linenumber" name="L20" href="#L20">20</a>  <strong class="jxr_keyword">import</strong> java.io.UnsupportedEncodingException;
<a class="jxr_linenumber" name="L21" href="#L21">21</a>  <strong class="jxr_keyword">import</strong> java.util.*;
<a class="jxr_linenumber" name="L22" href="#L22">22</a>  
<a class="jxr_linenumber" name="L23" href="#L23">23</a>  <em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L24" href="#L24">24</a>  <em class="jxr_javadoccomment"> * BioPAX (Level 3) Normalizer, an advanced BioPAX utility </em>
<a class="jxr_linenumber" name="L25" href="#L25">25</a>  <em class="jxr_javadoccomment"> * to help pathway data integrating and linking.</em>
<a class="jxr_linenumber" name="L26" href="#L26">26</a>  <em class="jxr_javadoccomment"> * </em>
<a class="jxr_linenumber" name="L27" href="#L27">27</a>  <em class="jxr_javadoccomment"> * @author rodche</em>
<a class="jxr_linenumber" name="L28" href="#L28">28</a>  <em class="jxr_javadoccomment"> */</em>
<a class="jxr_linenumber" name="L29" href="#L29">29</a>  <strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">final</strong> <strong class="jxr_keyword">class</strong> <a name="Normalizer" href="../../../../org/biopax/paxtools/normalizer/Normalizer.html#Normalizer">Normalizer</a> {
<a class="jxr_linenumber" name="L30" href="#L30">30</a>  	<strong class="jxr_keyword">private</strong> <strong class="jxr_keyword">static</strong> <strong class="jxr_keyword">final</strong> Logger log = LoggerFactory.getLogger(Normalizer.<strong class="jxr_keyword">class</strong>);
<a class="jxr_linenumber" name="L31" href="#L31">31</a>  	
<a class="jxr_linenumber" name="L32" href="#L32">32</a>  	<strong class="jxr_keyword">private</strong> <a name="SimpleIOHandler" href="../../../../org/biopax/paxtools/io/SimpleIOHandler.html#SimpleIOHandler">SimpleIOHandler</a> biopaxReader;
<a class="jxr_linenumber" name="L33" href="#L33">33</a>  	<strong class="jxr_keyword">private</strong> String description = <span class="jxr_string">""</span>;
<a class="jxr_linenumber" name="L34" href="#L34">34</a>  	<strong class="jxr_keyword">private</strong> <strong class="jxr_keyword">boolean</strong> fixDisplayName;
<a class="jxr_linenumber" name="L35" href="#L35">35</a>  	<strong class="jxr_keyword">private</strong> String xmlBase;
<a class="jxr_linenumber" name="L36" href="#L36">36</a>  	
<a class="jxr_linenumber" name="L37" href="#L37">37</a>  	<em class="jxr_comment">// Normalizer will generate URIs using a strategy specified by the system property</em>
<a class="jxr_linenumber" name="L38" href="#L38">38</a>  	<em class="jxr_comment">// (the default is biopax.normalizer.uri.strategy=md5, to generate 32-byte digest hex string for xrefs's uris)</em>
<a class="jxr_linenumber" name="L39" href="#L39">39</a>  	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">static</strong> <strong class="jxr_keyword">final</strong> String PROPERTY_NORMALIZER_URI_STRATEGY = <span class="jxr_string">"biopax.normalizer.uri.strategy"</span>;
<a class="jxr_linenumber" name="L40" href="#L40">40</a>  	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">static</strong> <strong class="jxr_keyword">final</strong> String VALUE_NORMALIZER_URI_STRATEGY_SIMPLE = <span class="jxr_string">"simple"</span>;
<a class="jxr_linenumber" name="L41" href="#L41">41</a>  	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">static</strong> <strong class="jxr_keyword">final</strong> String VALUE_NORMALIZER_URI_STRATEGY_MD5 = <span class="jxr_string">"md5"</span>; <em class="jxr_comment">//default strategy</em>
<a class="jxr_linenumber" name="L42" href="#L42">42</a>  	
<a class="jxr_linenumber" name="L43" href="#L43">43</a>  	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L44" href="#L44">44</a>  <em class="jxr_javadoccomment">	 * Constructor</em>
<a class="jxr_linenumber" name="L45" href="#L45">45</a>  <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L46" href="#L46">46</a>  	<strong class="jxr_keyword">public</strong> <a name="Normalizer" href="../../../../org/biopax/paxtools/normalizer/Normalizer.html#Normalizer">Normalizer</a>() {
<a class="jxr_linenumber" name="L47" href="#L47">47</a>  		biopaxReader = <strong class="jxr_keyword">new</strong> <a name="SimpleIOHandler" href="../../../../org/biopax/paxtools/io/SimpleIOHandler.html#SimpleIOHandler">SimpleIOHandler</a>(BioPAXLevel.L3);
<a class="jxr_linenumber" name="L48" href="#L48">48</a>  		biopaxReader.mergeDuplicates(<strong class="jxr_keyword">true</strong>);
<a class="jxr_linenumber" name="L49" href="#L49">49</a>  		fixDisplayName = <strong class="jxr_keyword">true</strong>;
<a class="jxr_linenumber" name="L50" href="#L50">50</a>  		xmlBase = <span class="jxr_string">""</span>;
<a class="jxr_linenumber" name="L51" href="#L51">51</a>  	}
<a class="jxr_linenumber" name="L52" href="#L52">52</a>  	
<a class="jxr_linenumber" name="L53" href="#L53">53</a>  	
<a class="jxr_linenumber" name="L54" href="#L54">54</a>  	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L55" href="#L55">55</a>  <em class="jxr_javadoccomment">	 * Normalizes BioPAX OWL data and returns</em>
<a class="jxr_linenumber" name="L56" href="#L56">56</a>  <em class="jxr_javadoccomment">	 * the result as BioPAX OWL (string).</em>
<a class="jxr_linenumber" name="L57" href="#L57">57</a>  <em class="jxr_javadoccomment">	 * </em>
<a class="jxr_linenumber" name="L58" href="#L58">58</a>  <em class="jxr_javadoccomment">	 * This public method is actually intended to use </em>
<a class="jxr_linenumber" name="L59" href="#L59">59</a>  <em class="jxr_javadoccomment">	 * outside the BioPAX Validator framework, with not too large models.</em>
<a class="jxr_linenumber" name="L60" href="#L60">60</a>  <em class="jxr_javadoccomment">	 * </em>
<a class="jxr_linenumber" name="L61" href="#L61">61</a>  <em class="jxr_javadoccomment">	 * @param biopaxOwlData RDF/XML BioPAX content string</em>
<a class="jxr_linenumber" name="L62" href="#L62">62</a>  <em class="jxr_javadoccomment">	 * @return normalized BioPAX RDF/XML</em>
<a class="jxr_linenumber" name="L63" href="#L63">63</a>  <em class="jxr_javadoccomment">	 * @deprecated this method will fail if the data exceeds ~1Gb (max UTF8 java String length)</em>
<a class="jxr_linenumber" name="L64" href="#L64">64</a>  <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L65" href="#L65">65</a>  	<strong class="jxr_keyword">public</strong> String normalize(String biopaxOwlData) {
<a class="jxr_linenumber" name="L66" href="#L66">66</a>  		
<a class="jxr_linenumber" name="L67" href="#L67">67</a>  		<strong class="jxr_keyword">if</strong>(biopaxOwlData == <strong class="jxr_keyword">null</strong> || biopaxOwlData.length() == 0) 
<a class="jxr_linenumber" name="L68" href="#L68">68</a>  			<strong class="jxr_keyword">throw</strong> <strong class="jxr_keyword">new</strong> IllegalArgumentException(<span class="jxr_string">"no data. "</span> + description);
<a class="jxr_linenumber" name="L69" href="#L69">69</a>  		
<a class="jxr_linenumber" name="L70" href="#L70">70</a>  		<em class="jxr_comment">// quick-fix for older BioPAX L3 version (v0.9x) property 'taxonXref' (range: BioSource)</em>
<a class="jxr_linenumber" name="L71" href="#L71">71</a>  		biopaxOwlData = biopaxOwlData.replaceAll(<span class="jxr_string">"taxonXref"</span>,<span class="jxr_string">"xref"</span>);
<a class="jxr_linenumber" name="L72" href="#L72">72</a>  		
<a class="jxr_linenumber" name="L73" href="#L73">73</a>  		<em class="jxr_comment">// build the model</em>
<a class="jxr_linenumber" name="L74" href="#L74">74</a>  		Model model = <strong class="jxr_keyword">null</strong>;
<a class="jxr_linenumber" name="L75" href="#L75">75</a>  		<strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L76" href="#L76">76</a>  			model = biopaxReader.convertFromOWL(
<a class="jxr_linenumber" name="L77" href="#L77">77</a>  				<strong class="jxr_keyword">new</strong> ByteArrayInputStream(biopaxOwlData.getBytes(<span class="jxr_string">"UTF-8"</span>)));
<a class="jxr_linenumber" name="L78" href="#L78">78</a>  		} <strong class="jxr_keyword">catch</strong> (UnsupportedEncodingException e) {
<a class="jxr_linenumber" name="L79" href="#L79">79</a>  			<strong class="jxr_keyword">throw</strong> <strong class="jxr_keyword">new</strong> IllegalArgumentException(<span class="jxr_string">"Failed! "</span> + description, e);
<a class="jxr_linenumber" name="L80" href="#L80">80</a>  		}
<a class="jxr_linenumber" name="L81" href="#L81">81</a>  		
<a class="jxr_linenumber" name="L82" href="#L82">82</a>  		<strong class="jxr_keyword">if</strong>(model == <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L83" href="#L83">83</a>  			<strong class="jxr_keyword">throw</strong> <strong class="jxr_keyword">new</strong> IllegalArgumentException(<span class="jxr_string">"Failed to create Model! "</span> 
<a class="jxr_linenumber" name="L84" href="#L84">84</a>  					+ description);
<a class="jxr_linenumber" name="L85" href="#L85">85</a>  		}
<a class="jxr_linenumber" name="L86" href="#L86">86</a>  		
<a class="jxr_linenumber" name="L87" href="#L87">87</a>  		<em class="jxr_comment">// auto-convert to Level3 model</em>
<a class="jxr_linenumber" name="L88" href="#L88">88</a>  		<strong class="jxr_keyword">if</strong> (model.getLevel() != BioPAXLevel.L3) {
<a class="jxr_linenumber" name="L89" href="#L89">89</a>  			log.info(<span class="jxr_string">"Converting model to BioPAX Level3..."</span>);
<a class="jxr_linenumber" name="L90" href="#L90">90</a>  			model = (<strong class="jxr_keyword">new</strong> <a name="LevelUpgrader" href="../../../../org/biopax/paxtools/converter/LevelUpgrader.html#LevelUpgrader">LevelUpgrader</a>()).filter(model);
<a class="jxr_linenumber" name="L91" href="#L91">91</a>  		}
<a class="jxr_linenumber" name="L92" href="#L92">92</a>  		
<a class="jxr_linenumber" name="L93" href="#L93">93</a>  		normalize(model); <em class="jxr_comment">// L3 only!</em>
<a class="jxr_linenumber" name="L94" href="#L94">94</a>  		
<a class="jxr_linenumber" name="L95" href="#L95">95</a>  		<em class="jxr_comment">// return as BioPAX OWL</em>
<a class="jxr_linenumber" name="L96" href="#L96">96</a>  		<strong class="jxr_keyword">return</strong> convertToOWL(model);
<a class="jxr_linenumber" name="L97" href="#L97">97</a>  	}
<a class="jxr_linenumber" name="L98" href="#L98">98</a>  	
<a class="jxr_linenumber" name="L99" href="#L99">99</a>  
<a class="jxr_linenumber" name="L100" href="#L100">100</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L101" href="#L101">101</a> <em class="jxr_javadoccomment">	 * Normalizes all xrefs (not unification xrefs only) to help normalizing/merging other objects, </em>
<a class="jxr_linenumber" name="L102" href="#L102">102</a> <em class="jxr_javadoccomment">	 * and also because some of the original xref URIs ("normalized")</em>
<a class="jxr_linenumber" name="L103" href="#L103">103</a> <em class="jxr_javadoccomment">	 * are in fact to be used for other biopax types (e.g., CV or ProteinReference); therefore</em>
<a class="jxr_linenumber" name="L104" href="#L104">104</a> <em class="jxr_javadoccomment">	 * this method will replace URI also for "bad" xrefs, i.e., those with empty/illegal 'db' or 'id' values.</em>
<a class="jxr_linenumber" name="L105" href="#L105">105</a> <em class="jxr_javadoccomment">	 * </em>
<a class="jxr_linenumber" name="L106" href="#L106">106</a> <em class="jxr_javadoccomment">	 * @param model biopax model to update</em>
<a class="jxr_linenumber" name="L107" href="#L107">107</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L108" href="#L108">108</a> 	<strong class="jxr_keyword">private</strong> <strong class="jxr_keyword">void</strong> normalizeXrefs(<a name="Model" href="../../../../org/biopax/paxtools/model/Model.html#Model">Model</a> model) {
<a class="jxr_linenumber" name="L109" href="#L109">109</a> 		
<a class="jxr_linenumber" name="L110" href="#L110">110</a> 		<strong class="jxr_keyword">final</strong> NormalizerMap map = <strong class="jxr_keyword">new</strong> NormalizerMap(model);
<a class="jxr_linenumber" name="L111" href="#L111">111</a> 		<strong class="jxr_keyword">final</strong> String xmlBase = getXmlBase(model); <em class="jxr_comment">//current base, the default or model's one, if set.</em>
<a class="jxr_linenumber" name="L112" href="#L112">112</a> 		
<a class="jxr_linenumber" name="L113" href="#L113">113</a> 		<em class="jxr_comment">// use a copy of the xrefs set (to avoid concurrent modif. exception)</em>
<a class="jxr_linenumber" name="L114" href="#L114">114</a> 		Set&lt;? <strong class="jxr_keyword">extends</strong> Xref&gt; xrefs = <strong class="jxr_keyword">new</strong> HashSet&lt;Xref&gt;(model.getObjects(Xref.<strong class="jxr_keyword">class</strong>));
<a class="jxr_linenumber" name="L115" href="#L115">115</a> 		<strong class="jxr_keyword">for</strong>(<a name="Xref" href="../../../../org/biopax/paxtools/model/level3/Xref.html#Xref">Xref</a> ref : xrefs)
<a class="jxr_linenumber" name="L116" href="#L116">116</a> 		{
<a class="jxr_linenumber" name="L117" href="#L117">117</a> 			<em class="jxr_comment">//skip not well-defined ones (incl. PublicationXrefs w/o db/id - won't normalize)</em>
<a class="jxr_linenumber" name="L118" href="#L118">118</a> 			<strong class="jxr_keyword">if</strong>(ref.getDb() == <strong class="jxr_keyword">null</strong> || ref.getId() == <strong class="jxr_keyword">null</strong>)
<a class="jxr_linenumber" name="L119" href="#L119">119</a> 				<strong class="jxr_keyword">continue</strong>;
<a class="jxr_linenumber" name="L120" href="#L120">120</a> 
<a class="jxr_linenumber" name="L121" href="#L121">121</a> 			ref.setDb(ref.getDb().toLowerCase()); <em class="jxr_comment">//set lowercase</em>
<a class="jxr_linenumber" name="L122" href="#L122">122</a> 			String idPart = ref.getId();
<a class="jxr_linenumber" name="L123" href="#L123">123</a> 
<a class="jxr_linenumber" name="L124" href="#L124">124</a> 			<strong class="jxr_keyword">if</strong>(ref instanceof <a name="RelationshipXref" href="../../../../org/biopax/paxtools/model/level3/RelationshipXref.html#RelationshipXref">RelationshipXref</a>) {
<a class="jxr_linenumber" name="L125" href="#L125">125</a> 				<em class="jxr_comment">// only normalize (replace URI of) RXs that could potentially clash with other RX, CVs, ERs;</em>
<a class="jxr_linenumber" name="L126" href="#L126">126</a> 				<em class="jxr_comment">// skip, won't bother, for all other RXs...</em>
<a class="jxr_linenumber" name="L127" href="#L127">127</a> 				<strong class="jxr_keyword">if</strong>(!ref.getUri().startsWith(<span class="jxr_string">"http://identifiers.org/"</span>))
<a class="jxr_linenumber" name="L128" href="#L128">128</a> 					<strong class="jxr_keyword">continue</strong>;
<a class="jxr_linenumber" name="L129" href="#L129">129</a> 
<a class="jxr_linenumber" name="L130" href="#L130">130</a> 				<em class="jxr_comment">//RXs might have the same db, id but different rel. type.</em>
<a class="jxr_linenumber" name="L131" href="#L131">131</a> 				RelationshipTypeVocabulary cv = ((<a name="RelationshipXref" href="../../../../org/biopax/paxtools/model/level3/RelationshipXref.html#RelationshipXref">RelationshipXref</a>) ref).getRelationshipType();
<a class="jxr_linenumber" name="L132" href="#L132">132</a> 				<strong class="jxr_keyword">if</strong>(ref.getIdVersion()!=<strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L133" href="#L133">133</a> 					idPart += <span class="jxr_string">"_"</span> + ref.getIdVersion();
<a class="jxr_linenumber" name="L134" href="#L134">134</a> 				}
<a class="jxr_linenumber" name="L135" href="#L135">135</a> 				<strong class="jxr_keyword">if</strong>(cv != <strong class="jxr_keyword">null</strong> &amp;&amp; !cv.getTerm().isEmpty()) {
<a class="jxr_linenumber" name="L136" href="#L136">136</a> 					idPart += <span class="jxr_string">"_"</span> + StringUtils.join(cv.getTerm(), '_').toLowerCase();
<a class="jxr_linenumber" name="L137" href="#L137">137</a> 				}
<a class="jxr_linenumber" name="L138" href="#L138">138</a> 			}
<a class="jxr_linenumber" name="L139" href="#L139">139</a> 			<strong class="jxr_keyword">else</strong> <strong class="jxr_keyword">if</strong>(ref instanceof <a name="UnificationXref" href="../../../../org/biopax/paxtools/model/level3/UnificationXref.html#UnificationXref">UnificationXref</a>) {
<a class="jxr_linenumber" name="L140" href="#L140">140</a> 				<em class="jxr_comment">// first, try to normalize the db name (using MIRIAM EBI registry)</em>
<a class="jxr_linenumber" name="L141" href="#L141">141</a> 				<strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L142" href="#L142">142</a> 					ref.setDb(MiriamLink.getName(ref.getDb()).toLowerCase());
<a class="jxr_linenumber" name="L143" href="#L143">143</a> 				} <strong class="jxr_keyword">catch</strong> (IllegalArgumentException e) {
<a class="jxr_linenumber" name="L144" href="#L144">144</a> 					<em class="jxr_comment">// - unknown/unmatched db name (normalize using defaults)</em>
<a class="jxr_linenumber" name="L145" href="#L145">145</a> 					<strong class="jxr_keyword">if</strong>(ref.getIdVersion()!=<strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L146" href="#L146">146</a> 						idPart += <span class="jxr_string">"_"</span> + ref.getIdVersion();
<a class="jxr_linenumber" name="L147" href="#L147">147</a> 					}
<a class="jxr_linenumber" name="L148" href="#L148">148</a> 					map.put(ref, Normalizer.uri(xmlBase, ref.getDb(), idPart, ref.getModelInterface()));
<a class="jxr_linenumber" name="L149" href="#L149">149</a> 					<strong class="jxr_keyword">continue</strong>; <em class="jxr_comment">//shortcut (non-standard db name)</em>
<a class="jxr_linenumber" name="L150" href="#L150">150</a> 				}
<a class="jxr_linenumber" name="L151" href="#L151">151</a> 			
<a class="jxr_linenumber" name="L152" href="#L152">152</a> 				<em class="jxr_comment">// a hack for uniprot/isoform xrefs</em>
<a class="jxr_linenumber" name="L153" href="#L153">153</a> 				<strong class="jxr_keyword">if</strong> (ref.getDb().startsWith(<span class="jxr_string">"uniprot"</span>)) {
<a class="jxr_linenumber" name="L154" href="#L154">154</a> 					<em class="jxr_comment">//auto-fix (guess) for possibly incorrect db/id (can be 'uniprot isoform' with/no idVersion, etc..)</em>
<a class="jxr_linenumber" name="L155" href="#L155">155</a> 					<strong class="jxr_keyword">if</strong> (isValidDbId(<span class="jxr_string">"uniprot isoform"</span>, ref.getId())
<a class="jxr_linenumber" name="L156" href="#L156">156</a> 							&amp;&amp; ref.getId().contains(<span class="jxr_string">"-"</span>)) <em class="jxr_comment">//the second condition is important</em>
<a class="jxr_linenumber" name="L157" href="#L157">157</a> 					{	<em class="jxr_comment">//then it's certainly an isoform id; so - fix the db name</em>
<a class="jxr_linenumber" name="L158" href="#L158">158</a> 						ref.setDb(<span class="jxr_string">"uniprot isoform"</span>); <em class="jxr_comment">//fix the db</em>
<a class="jxr_linenumber" name="L159" href="#L159">159</a> 					}
<a class="jxr_linenumber" name="L160" href="#L160">160</a> 					<strong class="jxr_keyword">else</strong> {
<a class="jxr_linenumber" name="L161" href="#L161">161</a> 						<em class="jxr_comment">//id does not end with "-&#92;&#92;d+", i.e., not a isoform id</em>
<a class="jxr_linenumber" name="L162" href="#L162">162</a> 						<em class="jxr_comment">//(idVersion is a different thing, but id db was "uniprot isoform" they probably misused idVersion)</em>
<a class="jxr_linenumber" name="L163" href="#L163">163</a> 						<strong class="jxr_keyword">if</strong>(ref.getDb().equals(<span class="jxr_string">"uniprot isoform"</span>))
<a class="jxr_linenumber" name="L164" href="#L164">164</a> 						{
<a class="jxr_linenumber" name="L165" href="#L165">165</a> 							<strong class="jxr_keyword">if</strong>(ref.getIdVersion() != <strong class="jxr_keyword">null</strong> &amp;&amp; ref.getIdVersion().matches(<span class="jxr_string">"^&#92;&#92;d+$"</span>))
<a class="jxr_linenumber" name="L166" href="#L166">166</a> 								idPart = ref.getId()+<span class="jxr_string">"-"</span>+ref.getIdVersion(); <em class="jxr_comment">//guess, by idVersion, they actually meant isoform</em>
<a class="jxr_linenumber" name="L167" href="#L167">167</a> 							<strong class="jxr_keyword">if</strong>(isValidDbId(ref.getDb(), idPart)) {
<a class="jxr_linenumber" name="L168" href="#L168">168</a> 								ref.setId(idPart); <em class="jxr_comment">//moved the isoform # to the ID</em>
<a class="jxr_linenumber" name="L169" href="#L169">169</a> 								ref.setIdVersion(<strong class="jxr_keyword">null</strong>);
<a class="jxr_linenumber" name="L170" href="#L170">170</a> 							}
<a class="jxr_linenumber" name="L171" href="#L171">171</a> 							<strong class="jxr_keyword">else</strong> <strong class="jxr_keyword">if</strong>(!isValidDbId(ref.getDb(), ref.getId())) {
<a class="jxr_linenumber" name="L172" href="#L172">172</a> 								<em class="jxr_comment">//certainly not isoform (might not even uniprot, but try...)</em>
<a class="jxr_linenumber" name="L173" href="#L173">173</a> 								ref.setDb(<span class="jxr_string">"uniprot knowledgebase"</span>); <em class="jxr_comment">//guess, fix</em>
<a class="jxr_linenumber" name="L174" href="#L174">174</a> 							}
<a class="jxr_linenumber" name="L175" href="#L175">175</a> 							idPart = ref.getId();
<a class="jxr_linenumber" name="L176" href="#L176">176</a> 						}
<a class="jxr_linenumber" name="L177" href="#L177">177</a> 					}
<a class="jxr_linenumber" name="L178" href="#L178">178</a> 				}
<a class="jxr_linenumber" name="L179" href="#L179">179</a> 			}
<a class="jxr_linenumber" name="L180" href="#L180">180</a> 
<a class="jxr_linenumber" name="L181" href="#L181">181</a> 			<em class="jxr_comment">// shelve it for URI replace</em>
<a class="jxr_linenumber" name="L182" href="#L182">182</a> 			map.put(ref, Normalizer.uri(xmlBase, ref.getDb(), idPart, ref.getModelInterface()));
<a class="jxr_linenumber" name="L183" href="#L183">183</a> 		}
<a class="jxr_linenumber" name="L184" href="#L184">184</a> 		
<a class="jxr_linenumber" name="L185" href="#L185">185</a> 		<em class="jxr_comment">// execute replace xrefs</em>
<a class="jxr_linenumber" name="L186" href="#L186">186</a> 		map.doSubs();
<a class="jxr_linenumber" name="L187" href="#L187">187</a> 	}
<a class="jxr_linenumber" name="L188" href="#L188">188</a> 
<a class="jxr_linenumber" name="L189" href="#L189">189</a> 	<em class="jxr_comment">/*</em>
<a class="jxr_linenumber" name="L190" href="#L190">190</a> <em class="jxr_comment">	 * @param db must be valid MIRIAM db name or sinonym</em>
<a class="jxr_linenumber" name="L191" href="#L191">191</a> <em class="jxr_comment">	 * @trows IllegalArgumentException when db is unknown name.</em>
<a class="jxr_linenumber" name="L192" href="#L192">192</a> <em class="jxr_comment">	 */</em>
<a class="jxr_linenumber" name="L193" href="#L193">193</a> 	<strong class="jxr_keyword">private</strong> <strong class="jxr_keyword">boolean</strong> isValidDbId(String db, String id) {
<a class="jxr_linenumber" name="L194" href="#L194">194</a> 		<strong class="jxr_keyword">return</strong> MiriamLink.checkRegExp(id, db);
<a class="jxr_linenumber" name="L195" href="#L195">195</a> 	}
<a class="jxr_linenumber" name="L196" href="#L196">196</a> 
<a class="jxr_linenumber" name="L197" href="#L197">197</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L198" href="#L198">198</a> <em class="jxr_javadoccomment">	 * Consistently generates a new BioPAX element URI </em>
<a class="jxr_linenumber" name="L199" href="#L199">199</a> <em class="jxr_javadoccomment">	 * using given URI namespace (xml:base), BioPAX class, </em>
<a class="jxr_linenumber" name="L200" href="#L200">200</a> <em class="jxr_javadoccomment">	 * and two different identifiers (at least one is required).</em>
<a class="jxr_linenumber" name="L201" href="#L201">201</a> <em class="jxr_javadoccomment">	 * Miriam registry is used to get the standard db name and </em>
<a class="jxr_linenumber" name="L202" href="#L202">202</a> <em class="jxr_javadoccomment">	 * identifiers.org URI, if possible, only for relationship type vocabulary, </em>
<a class="jxr_linenumber" name="L203" href="#L203">203</a> <em class="jxr_javadoccomment">	 * publication xref, and entity reference types.</em>
<a class="jxr_linenumber" name="L204" href="#L204">204</a> <em class="jxr_javadoccomment">	 * </em>
<a class="jxr_linenumber" name="L205" href="#L205">205</a> <em class="jxr_javadoccomment">	 * @param xmlBase xml:base (common URI prefix for a BioPAX model), case-sensitive</em>
<a class="jxr_linenumber" name="L206" href="#L206">206</a> <em class="jxr_javadoccomment">	 * @param dbName a bio data collection name or synonym, case-insensitive</em>
<a class="jxr_linenumber" name="L207" href="#L207">207</a> <em class="jxr_javadoccomment">	 * @param idPart optional (can be null), e.g., xref.id, case-sensitive</em>
<a class="jxr_linenumber" name="L208" href="#L208">208</a> <em class="jxr_javadoccomment">	 * @param type BioPAX class</em>
<a class="jxr_linenumber" name="L209" href="#L209">209</a> <em class="jxr_javadoccomment">	 * @return URI</em>
<a class="jxr_linenumber" name="L210" href="#L210">210</a> <em class="jxr_javadoccomment">	 * @throws IllegalArgumentException if either type is null or both 'dbName' and 'idPart' are all nulls.</em>
<a class="jxr_linenumber" name="L211" href="#L211">211</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L212" href="#L212">212</a> 	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">static</strong> String uri(<strong class="jxr_keyword">final</strong> String xmlBase, 
<a class="jxr_linenumber" name="L213" href="#L213">213</a> 		String dbName, String idPart, Class&lt;? <strong class="jxr_keyword">extends</strong> BioPAXElement&gt; type)
<a class="jxr_linenumber" name="L214" href="#L214">214</a> 	{
<a class="jxr_linenumber" name="L215" href="#L215">215</a> 		<strong class="jxr_keyword">if</strong>(type == <strong class="jxr_keyword">null</strong> || (dbName == <strong class="jxr_keyword">null</strong> &amp;&amp; idPart == <strong class="jxr_keyword">null</strong>))
<a class="jxr_linenumber" name="L216" href="#L216">216</a> 			<strong class="jxr_keyword">throw</strong> <strong class="jxr_keyword">new</strong> IllegalArgumentException(<span class="jxr_string">"'Either type' is null, or both dbName and idPart are nulls."</span>);
<a class="jxr_linenumber" name="L217" href="#L217">217</a> 
<a class="jxr_linenumber" name="L218" href="#L218">218</a> 		<strong class="jxr_keyword">if</strong> (idPart != <strong class="jxr_keyword">null</strong>) idPart = idPart.trim();
<a class="jxr_linenumber" name="L219" href="#L219">219</a> 		<strong class="jxr_keyword">if</strong> (dbName != <strong class="jxr_keyword">null</strong>) dbName = dbName.trim();
<a class="jxr_linenumber" name="L220" href="#L220">220</a> 
<a class="jxr_linenumber" name="L221" href="#L221">221</a> 		<em class="jxr_comment">// try to find a standard URI, if exists, for a publication xref, or at least a standard name:</em>
<a class="jxr_linenumber" name="L222" href="#L222">222</a> 		<strong class="jxr_keyword">if</strong> (dbName != <strong class="jxr_keyword">null</strong>)
<a class="jxr_linenumber" name="L223" href="#L223">223</a> 		{
<a class="jxr_linenumber" name="L224" href="#L224">224</a> 			<strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L225" href="#L225">225</a> 				<em class="jxr_comment">// try to get the preferred/standard name</em>
<a class="jxr_linenumber" name="L226" href="#L226">226</a> 				<em class="jxr_comment">// for any type, for consistency</em>
<a class="jxr_linenumber" name="L227" href="#L227">227</a> 				dbName = MiriamLink.getName(dbName);
<a class="jxr_linenumber" name="L228" href="#L228">228</a> 				<em class="jxr_comment">// a shortcut: a standard and resolvable URI exists for some BioPAX types</em>
<a class="jxr_linenumber" name="L229" href="#L229">229</a> 				<strong class="jxr_keyword">if</strong> ((type.equals(PublicationXref.<strong class="jxr_keyword">class</strong>) &amp;&amp; <span class="jxr_string">"pubmed"</span>.equalsIgnoreCase(dbName))
<a class="jxr_linenumber" name="L230" href="#L230">230</a> 					|| type.equals(RelationshipTypeVocabulary.<strong class="jxr_keyword">class</strong>)
<a class="jxr_linenumber" name="L231" href="#L231">231</a> 					|| ProteinReference.<strong class="jxr_keyword">class</strong>.isAssignableFrom(type)
<a class="jxr_linenumber" name="L232" href="#L232">232</a> 					|| SmallMoleculeReference.<strong class="jxr_keyword">class</strong>.isAssignableFrom(type)
<a class="jxr_linenumber" name="L233" href="#L233">233</a> 					|| (type.equals(BioSource.<strong class="jxr_keyword">class</strong>) &amp;&amp; <span class="jxr_string">"taxonomy"</span>.equalsIgnoreCase(dbName)
<a class="jxr_linenumber" name="L234" href="#L234">234</a> 						&amp;&amp; idPart!=<strong class="jxr_keyword">null</strong> &amp;&amp; idPart.matches(<span class="jxr_string">"^&#92;&#92;d+$"</span>))
<a class="jxr_linenumber" name="L235" href="#L235">235</a> 				)
<a class="jxr_linenumber" name="L236" href="#L236">236</a> 				{	<em class="jxr_comment">//get the standard URI and quit (success), or fail and continue making a new URI below...</em>
<a class="jxr_linenumber" name="L237" href="#L237">237</a> 					<strong class="jxr_keyword">return</strong> MiriamLink.getIdentifiersOrgURI(dbName, idPart);
<a class="jxr_linenumber" name="L238" href="#L238">238</a> 				}
<a class="jxr_linenumber" name="L239" href="#L239">239</a> 			} <strong class="jxr_keyword">catch</strong> (IllegalArgumentException e) {
<a class="jxr_linenumber" name="L240" href="#L240">240</a> 				log.info(String.format(<span class="jxr_string">"uri(for a %s): db:%s, id:%s are not standard; %s)"</span>,
<a class="jxr_linenumber" name="L241" href="#L241">241</a> 						type.getSimpleName(), dbName, idPart, e.getMessage()));
<a class="jxr_linenumber" name="L242" href="#L242">242</a> 			}
<a class="jxr_linenumber" name="L243" href="#L243">243</a> 		}
<a class="jxr_linenumber" name="L244" href="#L244">244</a> 
<a class="jxr_linenumber" name="L245" href="#L245">245</a> 		<em class="jxr_comment">// If not returned above this point - no standard URI (Identifiers.org) was found -</em>
<a class="jxr_linenumber" name="L246" href="#L246">246</a> 		<em class="jxr_comment">// then let's consistently build a new URI from args, anyway, the other way around:</em>
<a class="jxr_linenumber" name="L247" href="#L247">247</a> 		
<a class="jxr_linenumber" name="L248" href="#L248">248</a> 		StringBuilder sb = <strong class="jxr_keyword">new</strong> StringBuilder();		
<a class="jxr_linenumber" name="L249" href="#L249">249</a> 		<strong class="jxr_keyword">if</strong> (dbName != <strong class="jxr_keyword">null</strong>) <em class="jxr_comment">//lowercase for consistency</em>
<a class="jxr_linenumber" name="L250" href="#L250">250</a> 			sb.append(dbName.toLowerCase()); 	
<a class="jxr_linenumber" name="L251" href="#L251">251</a> 		
<a class="jxr_linenumber" name="L252" href="#L252">252</a> 		<strong class="jxr_keyword">if</strong> (idPart != <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L253" href="#L253">253</a> 			<strong class="jxr_keyword">if</strong> (dbName != <strong class="jxr_keyword">null</strong>) sb.append(<span class="jxr_string">"_"</span>);
<a class="jxr_linenumber" name="L254" href="#L254">254</a> 			sb.append(idPart);
<a class="jxr_linenumber" name="L255" href="#L255">255</a> 		}
<a class="jxr_linenumber" name="L256" href="#L256">256</a> 		
<a class="jxr_linenumber" name="L257" href="#L257">257</a> 		String localPart = sb.toString();
<a class="jxr_linenumber" name="L258" href="#L258">258</a> 		String strategy = System.getProperty(PROPERTY_NORMALIZER_URI_STRATEGY, VALUE_NORMALIZER_URI_STRATEGY_MD5);
<a class="jxr_linenumber" name="L259" href="#L259">259</a> 		<strong class="jxr_keyword">if</strong>(VALUE_NORMALIZER_URI_STRATEGY_SIMPLE.equals(strategy) || Xref.<strong class="jxr_keyword">class</strong>.isAssignableFrom(type))
<a class="jxr_linenumber" name="L260" href="#L260">260</a> 		<em class="jxr_comment">//i.e., for xrefs, always use the simple URI strategy (makes them human-readable)</em>
<a class="jxr_linenumber" name="L261" href="#L261">261</a> 		{
<a class="jxr_linenumber" name="L262" href="#L262">262</a> 			<em class="jxr_comment">//simply replace "unsafe" symbols with underscore (some uri clashes might be possible but rare...)</em>
<a class="jxr_linenumber" name="L263" href="#L263">263</a> 			localPart = localPart.replaceAll(<span class="jxr_string">"[^-&#92;&#92;w]"</span>, <span class="jxr_string">"_"</span>);
<a class="jxr_linenumber" name="L264" href="#L264">264</a> 		}
<a class="jxr_linenumber" name="L265" href="#L265">265</a> 		<strong class="jxr_keyword">else</strong>
<a class="jxr_linenumber" name="L266" href="#L266">266</a> 		{
<a class="jxr_linenumber" name="L267" href="#L267">267</a> 			<em class="jxr_comment">//replace the local part with its md5 sum string (32-byte)</em>
<a class="jxr_linenumber" name="L268" href="#L268">268</a> 			localPart = ModelUtils.md5hex(localPart);
<a class="jxr_linenumber" name="L269" href="#L269">269</a> 		}
<a class="jxr_linenumber" name="L270" href="#L270">270</a> 		
<a class="jxr_linenumber" name="L271" href="#L271">271</a> 		<em class="jxr_comment">// create URI using the xml:base and digest of other values:</em>
<a class="jxr_linenumber" name="L272" href="#L272">272</a> 		<strong class="jxr_keyword">return</strong> ((xmlBase != <strong class="jxr_keyword">null</strong>) ? xmlBase : <span class="jxr_string">""</span>) + type.getSimpleName() + <span class="jxr_string">"_"</span> + localPart;
<a class="jxr_linenumber" name="L273" href="#L273">273</a> 	}
<a class="jxr_linenumber" name="L274" href="#L274">274</a> 
<a class="jxr_linenumber" name="L275" href="#L275">275</a> 	
<a class="jxr_linenumber" name="L276" href="#L276">276</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L277" href="#L277">277</a> <em class="jxr_javadoccomment">	 * Description of the model to normalize.</em>
<a class="jxr_linenumber" name="L278" href="#L278">278</a> <em class="jxr_javadoccomment">	 * </em>
<a class="jxr_linenumber" name="L279" href="#L279">279</a> <em class="jxr_javadoccomment">	 * @return info about the BioPAX model</em>
<a class="jxr_linenumber" name="L280" href="#L280">280</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L281" href="#L281">281</a> 	<strong class="jxr_keyword">public</strong> String getDescription() {
<a class="jxr_linenumber" name="L282" href="#L282">282</a> 		<strong class="jxr_keyword">return</strong> description;
<a class="jxr_linenumber" name="L283" href="#L283">283</a> 	}
<a class="jxr_linenumber" name="L284" href="#L284">284</a> 	
<a class="jxr_linenumber" name="L285" href="#L285">285</a> 	
<a class="jxr_linenumber" name="L286" href="#L286">286</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L287" href="#L287">287</a> <em class="jxr_javadoccomment">	 * @param description the description to set</em>
<a class="jxr_linenumber" name="L288" href="#L288">288</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L289" href="#L289">289</a> 	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">void</strong> setDescription(String description) {
<a class="jxr_linenumber" name="L290" href="#L290">290</a> 		<strong class="jxr_keyword">this</strong>.description = description;
<a class="jxr_linenumber" name="L291" href="#L291">291</a> 	}
<a class="jxr_linenumber" name="L292" href="#L292">292</a> 
<a class="jxr_linenumber" name="L293" href="#L293">293</a> 	
<a class="jxr_linenumber" name="L294" href="#L294">294</a> 	<strong class="jxr_keyword">private</strong> <strong class="jxr_keyword">void</strong> fixDisplayName(<a name="Model" href="../../../../org/biopax/paxtools/model/Model.html#Model">Model</a> model) {
<a class="jxr_linenumber" name="L295" href="#L295">295</a> 		log.info(<span class="jxr_string">"Trying to auto-fix 'null' displayName..."</span>);
<a class="jxr_linenumber" name="L296" href="#L296">296</a> 		<em class="jxr_comment">// where it's null, set to the shortest name if possible</em>
<a class="jxr_linenumber" name="L297" href="#L297">297</a> 		<strong class="jxr_keyword">for</strong> (<a name="Named" href="../../../../org/biopax/paxtools/model/level3/Named.html#Named">Named</a> e : model.getObjects(Named.<strong class="jxr_keyword">class</strong>)) {
<a class="jxr_linenumber" name="L298" href="#L298">298</a> 			<strong class="jxr_keyword">if</strong> (e.getDisplayName() == <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L299" href="#L299">299</a> 				<strong class="jxr_keyword">if</strong> (e.getStandardName() != <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L300" href="#L300">300</a> 					e.setDisplayName(e.getStandardName());
<a class="jxr_linenumber" name="L301" href="#L301">301</a> 					log.info(e + <span class="jxr_string">" displayName auto-fix: "</span>
<a class="jxr_linenumber" name="L302" href="#L302">302</a> 							+ e.getDisplayName() + <span class="jxr_string">". "</span> + description);
<a class="jxr_linenumber" name="L303" href="#L303">303</a> 				} <strong class="jxr_keyword">else</strong> <strong class="jxr_keyword">if</strong> (!e.getName().isEmpty()) {
<a class="jxr_linenumber" name="L304" href="#L304">304</a> 					String dsp = e.getName().iterator().next();
<a class="jxr_linenumber" name="L305" href="#L305">305</a> 					<strong class="jxr_keyword">for</strong> (String name : e.getName()) {
<a class="jxr_linenumber" name="L306" href="#L306">306</a> 						<strong class="jxr_keyword">if</strong> (name.length() &lt; dsp.length())
<a class="jxr_linenumber" name="L307" href="#L307">307</a> 							dsp = name;
<a class="jxr_linenumber" name="L308" href="#L308">308</a> 					}
<a class="jxr_linenumber" name="L309" href="#L309">309</a> 					e.setDisplayName(dsp);
<a class="jxr_linenumber" name="L310" href="#L310">310</a> 					log.info(e + <span class="jxr_string">" displayName auto-fix: "</span> + dsp
<a class="jxr_linenumber" name="L311" href="#L311">311</a> 						+ <span class="jxr_string">". "</span> + description);
<a class="jxr_linenumber" name="L312" href="#L312">312</a> 				}
<a class="jxr_linenumber" name="L313" href="#L313">313</a> 			}
<a class="jxr_linenumber" name="L314" href="#L314">314</a> 		}
<a class="jxr_linenumber" name="L315" href="#L315">315</a> 		<em class="jxr_comment">// if required, set PE name to (already fixed) ER's name...</em>
<a class="jxr_linenumber" name="L316" href="#L316">316</a> 		<strong class="jxr_keyword">for</strong>(<a name="EntityReference" href="../../../../org/biopax/paxtools/model/level3/EntityReference.html#EntityReference">EntityReference</a> er : model.getObjects(EntityReference.<strong class="jxr_keyword">class</strong>)) {
<a class="jxr_linenumber" name="L317" href="#L317">317</a> 			<strong class="jxr_keyword">for</strong>(<a name="SimplePhysicalEntity" href="../../../../org/biopax/paxtools/model/level3/SimplePhysicalEntity.html#SimplePhysicalEntity">SimplePhysicalEntity</a> spe : er.getEntityReferenceOf()) {
<a class="jxr_linenumber" name="L318" href="#L318">318</a> 				<strong class="jxr_keyword">if</strong>(spe.getDisplayName() == <strong class="jxr_keyword">null</strong> || spe.getDisplayName().trim().length() == 0) {
<a class="jxr_linenumber" name="L319" href="#L319">319</a> 					<strong class="jxr_keyword">if</strong>(er.getDisplayName() != <strong class="jxr_keyword">null</strong> &amp;&amp; er.getDisplayName().trim().length() &gt; 0) {
<a class="jxr_linenumber" name="L320" href="#L320">320</a> 						spe.setDisplayName(er.getDisplayName());
<a class="jxr_linenumber" name="L321" href="#L321">321</a> 					}
<a class="jxr_linenumber" name="L322" href="#L322">322</a> 				}
<a class="jxr_linenumber" name="L323" href="#L323">323</a> 			}
<a class="jxr_linenumber" name="L324" href="#L324">324</a> 		}
<a class="jxr_linenumber" name="L325" href="#L325">325</a> 	}
<a class="jxr_linenumber" name="L326" href="#L326">326</a> 
<a class="jxr_linenumber" name="L327" href="#L327">327</a> 
<a class="jxr_linenumber" name="L328" href="#L328">328</a> 	<strong class="jxr_keyword">private</strong> String convertToOWL(<a name="Model" href="../../../../org/biopax/paxtools/model/Model.html#Model">Model</a> model) {
<a class="jxr_linenumber" name="L329" href="#L329">329</a> 		ByteArrayOutputStream out = <strong class="jxr_keyword">new</strong> ByteArrayOutputStream();
<a class="jxr_linenumber" name="L330" href="#L330">330</a> 		(<strong class="jxr_keyword">new</strong> <a name="SimpleIOHandler" href="../../../../org/biopax/paxtools/io/SimpleIOHandler.html#SimpleIOHandler">SimpleIOHandler</a>(model.getLevel())).convertToOWL(model, out);
<a class="jxr_linenumber" name="L331" href="#L331">331</a> 		<strong class="jxr_keyword">return</strong> out.toString();
<a class="jxr_linenumber" name="L332" href="#L332">332</a> 	}
<a class="jxr_linenumber" name="L333" href="#L333">333</a> 
<a class="jxr_linenumber" name="L334" href="#L334">334</a> 
<a class="jxr_linenumber" name="L335" href="#L335">335</a> 	<strong class="jxr_keyword">private</strong> Collection&lt;UnificationXref&gt; getUnificationXrefsSorted(<a name="XReferrable" href="../../../../org/biopax/paxtools/model/level3/XReferrable.html#XReferrable">XReferrable</a> r) {
<a class="jxr_linenumber" name="L336" href="#L336">336</a> 
<a class="jxr_linenumber" name="L337" href="#L337">337</a> 		List&lt;UnificationXref&gt; urefs = <strong class="jxr_keyword">new</strong> ArrayList&lt;UnificationXref&gt;();
<a class="jxr_linenumber" name="L338" href="#L338">338</a> 		<strong class="jxr_keyword">for</strong>(<a name="UnificationXref" href="../../../../org/biopax/paxtools/model/level3/UnificationXref.html#UnificationXref">UnificationXref</a> ux : <strong class="jxr_keyword">new</strong> ClassFilterSet&lt;Xref,UnificationXref&gt;(r.getXref(), UnificationXref.<strong class="jxr_keyword">class</strong>))
<a class="jxr_linenumber" name="L339" href="#L339">339</a> 		{
<a class="jxr_linenumber" name="L340" href="#L340">340</a> 			<strong class="jxr_keyword">if</strong>(ux.getDb() != <strong class="jxr_keyword">null</strong> &amp;&amp; ux.getId() != <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L341" href="#L341">341</a> 				urefs.add(ux);
<a class="jxr_linenumber" name="L342" href="#L342">342</a> 			} 
<a class="jxr_linenumber" name="L343" href="#L343">343</a> 		}
<a class="jxr_linenumber" name="L344" href="#L344">344</a> 
<a class="jxr_linenumber" name="L345" href="#L345">345</a> 		Collections.sort(urefs, <strong class="jxr_keyword">new</strong> Comparator&lt;UnificationXref&gt;() {
<a class="jxr_linenumber" name="L346" href="#L346">346</a> 			<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">int</strong> compare(<a name="Unification<a name="UnificationXref" href="../../../../org/biopax/paxtools/model/level3/UnificationXref.html#UnificationXref">UnificationXref</a>../../../org/biopax/paxtools/model/level3/UnificationXref.html#UnificationXref">UnificationXref</a> o1, <a name="UnificationXref" href="../../../../org/biopax/paxtools/model/level3/UnificationXref.html#UnificationXref">UnificationXref</a> o2) {
<a class="jxr_linenumber" name="L347" href="#L347">347</a> 				String s1 = o1.getDb() + o1.getId();
<a class="jxr_linenumber" name="L348" href="#L348">348</a> 				String s2 = o2.getDb() + o2.getId();
<a class="jxr_linenumber" name="L349" href="#L349">349</a> 				<strong class="jxr_keyword">return</strong> s1.compareTo(s2);
<a class="jxr_linenumber" name="L350" href="#L350">350</a> 			}
<a class="jxr_linenumber" name="L351" href="#L351">351</a> 		});
<a class="jxr_linenumber" name="L352" href="#L352">352</a> 
<a class="jxr_linenumber" name="L353" href="#L353">353</a> 		<strong class="jxr_keyword">return</strong> urefs;
<a class="jxr_linenumber" name="L354" href="#L354">354</a> 	}
<a class="jxr_linenumber" name="L355" href="#L355">355</a> 
<a class="jxr_linenumber" name="L356" href="#L356">356</a> 	
<a class="jxr_linenumber" name="L357" href="#L357">357</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L358" href="#L358">358</a> <em class="jxr_javadoccomment">	 * Finds one preferred unification xref, if possible.</em>
<a class="jxr_linenumber" name="L359" href="#L359">359</a> <em class="jxr_javadoccomment">	 * Preferred db values are:</em>
<a class="jxr_linenumber" name="L360" href="#L360">360</a> <em class="jxr_javadoccomment">	 * "ncbi gene" - for NucleicAcidReference and the sub-classes;</em>
<a class="jxr_linenumber" name="L361" href="#L361">361</a> <em class="jxr_javadoccomment">	 * "uniprot" or "refseq" for ProteinReference;</em>
<a class="jxr_linenumber" name="L362" href="#L362">362</a> <em class="jxr_javadoccomment">	 * "chebi" - for SmallMoleculeReference;</em>
<a class="jxr_linenumber" name="L363" href="#L363">363</a> <em class="jxr_javadoccomment">	 * </em>
<a class="jxr_linenumber" name="L364" href="#L364">364</a> <em class="jxr_javadoccomment">	 * @param bpe BioPAX object that can have xrefs</em>
<a class="jxr_linenumber" name="L365" href="#L365">365</a> <em class="jxr_javadoccomment">	 * @return the "best" first unification xref </em>
<a class="jxr_linenumber" name="L366" href="#L366">366</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L367" href="#L367">367</a> 	<strong class="jxr_keyword">private</strong> <a name="UnificationXref" href="../../../../org/biopax/paxtools/model/level3/UnificationXref.html#UnificationXref">UnificationXref</a> findPreferredUnificationXref(<a name="XReferrable" href="../../../../org/biopax/paxtools/model/level3/XReferrable.html#XReferrable">XReferrable</a> bpe)
<a class="jxr_linenumber" name="L368" href="#L368">368</a> 	{
<a class="jxr_linenumber" name="L369" href="#L369">369</a> 		UnificationXref toReturn = <strong class="jxr_keyword">null</strong>;
<a class="jxr_linenumber" name="L370" href="#L370">370</a> 
<a class="jxr_linenumber" name="L371" href="#L371">371</a> 		Collection&lt;UnificationXref&gt; orderedUrefs = getUnificationXrefsSorted(bpe);
<a class="jxr_linenumber" name="L372" href="#L372">372</a> 
<a class="jxr_linenumber" name="L373" href="#L373">373</a> 		<em class="jxr_comment">//use preferred db prefix for different type of ER</em>
<a class="jxr_linenumber" name="L374" href="#L374">374</a> 		<strong class="jxr_keyword">if</strong>(bpe instanceof <a name="ProteinReference" href="../../../../org/biopax/paxtools/model/level3/ProteinReference.html#ProteinReference">ProteinReference</a>) {
<a class="jxr_linenumber" name="L375" href="#L375">375</a> 			<em class="jxr_comment">// prefer xref having db starts with "uniprot" (preferred) or "refseq";</em>
<a class="jxr_linenumber" name="L376" href="#L376">376</a> 			toReturn = findSingleUnificationXref(orderedUrefs, <span class="jxr_string">"uniprot"</span>);
<a class="jxr_linenumber" name="L377" href="#L377">377</a> 			<em class="jxr_comment">// if null, next try refseq.</em>
<a class="jxr_linenumber" name="L378" href="#L378">378</a> 			<strong class="jxr_keyword">if</strong>(toReturn==<strong class="jxr_keyword">null</strong>)
<a class="jxr_linenumber" name="L379" href="#L379">379</a> 				toReturn = findSingleUnificationXref(orderedUrefs, <span class="jxr_string">"refseq"</span>);
<a class="jxr_linenumber" name="L380" href="#L380">380</a> 		} <strong class="jxr_keyword">else</strong> <strong class="jxr_keyword">if</strong>(bpe instanceof <a name="SmallMoleculeReference" href="../../../../org/biopax/paxtools/model/level3/SmallMoleculeReference.html#SmallMoleculeReference">SmallMoleculeReference</a>) {
<a class="jxr_linenumber" name="L381" href="#L381">381</a> 			<em class="jxr_comment">// - "chebi", then  "pubchem";</em>
<a class="jxr_linenumber" name="L382" href="#L382">382</a> 			toReturn = findSingleUnificationXref(orderedUrefs, <span class="jxr_string">"chebi"</span>);
<a class="jxr_linenumber" name="L383" href="#L383">383</a> 			<strong class="jxr_keyword">if</strong>(toReturn==<strong class="jxr_keyword">null</strong>)
<a class="jxr_linenumber" name="L384" href="#L384">384</a> 				toReturn = findSingleUnificationXref(orderedUrefs, <span class="jxr_string">"pubchem"</span>);
<a class="jxr_linenumber" name="L385" href="#L385">385</a> 		} <strong class="jxr_keyword">else</strong> <strong class="jxr_keyword">if</strong>(bpe instanceof <a name="NucleicAcidReference" href="../../../../org/biopax/paxtools/model/level3/NucleicAcidReference.html#NucleicAcidReference">NucleicAcidReference</a>) {
<a class="jxr_linenumber" name="L386" href="#L386">386</a> 			<em class="jxr_comment">//that includes NucleicAcidRegionReference, etc. sub-classes;</em>
<a class="jxr_linenumber" name="L387" href="#L387">387</a> 			toReturn = findSingleUnificationXref(orderedUrefs, <span class="jxr_string">"ncbi gene"</span>);
<a class="jxr_linenumber" name="L388" href="#L388">388</a> 			<strong class="jxr_keyword">if</strong>(toReturn==<strong class="jxr_keyword">null</strong>)
<a class="jxr_linenumber" name="L389" href="#L389">389</a> 				toReturn = findSingleUnificationXref(orderedUrefs, <span class="jxr_string">"entrez"</span>);
<a class="jxr_linenumber" name="L390" href="#L390">390</a> 		} <strong class="jxr_keyword">else</strong> {
<a class="jxr_linenumber" name="L391" href="#L391">391</a> 			<em class="jxr_comment">//for other XReferrable types (BioSource or ControlledVocabulary)</em>
<a class="jxr_linenumber" name="L392" href="#L392">392</a> 			<em class="jxr_comment">//use if there's only one xref (return null if many)</em>
<a class="jxr_linenumber" name="L393" href="#L393">393</a> 			<strong class="jxr_keyword">if</strong>(orderedUrefs.size()==1)
<a class="jxr_linenumber" name="L394" href="#L394">394</a> 				toReturn = orderedUrefs.iterator().next();
<a class="jxr_linenumber" name="L395" href="#L395">395</a> 		}
<a class="jxr_linenumber" name="L396" href="#L396">396</a> 		
<a class="jxr_linenumber" name="L397" href="#L397">397</a> 		<strong class="jxr_keyword">return</strong> toReturn;
<a class="jxr_linenumber" name="L398" href="#L398">398</a> 	}
<a class="jxr_linenumber" name="L399" href="#L399">399</a> 
<a class="jxr_linenumber" name="L400" href="#L400">400</a> 	<strong class="jxr_keyword">private</strong> <a name="UnificationXref" href="../../../../org/biopax/paxtools/model/level3/UnificationXref.html#UnificationXref">UnificationXref</a> findSingleUnificationXref(Collection&lt;UnificationXref&gt; uXrefs, String dbStartsWith) {
<a class="jxr_linenumber" name="L401" href="#L401">401</a> 		UnificationXref ret = <strong class="jxr_keyword">null</strong>;
<a class="jxr_linenumber" name="L402" href="#L402">402</a> 
<a class="jxr_linenumber" name="L403" href="#L403">403</a> 		<strong class="jxr_keyword">for</strong>(<a name="Xref" href="../../../../org/biopax/paxtools/model/level3/Xref.html#Xref">Xref</a> x : uXrefs) {
<a class="jxr_linenumber" name="L404" href="#L404">404</a> 			<strong class="jxr_keyword">if</strong>(x instanceof <a name="UnificationXref" href="../../../../org/biopax/paxtools/model/level3/UnificationXref.html#UnificationXref">UnificationXref</a> &amp;&amp; x.getId() != <strong class="jxr_keyword">null</strong> &amp;&amp; x.getDb() != <strong class="jxr_keyword">null</strong>
<a class="jxr_linenumber" name="L405" href="#L405">405</a> 					&amp;&amp; x.getDb().toLowerCase().startsWith(dbStartsWith))
<a class="jxr_linenumber" name="L406" href="#L406">406</a> 			{
<a class="jxr_linenumber" name="L407" href="#L407">407</a> 				<strong class="jxr_keyword">if</strong>(ret == <strong class="jxr_keyword">null</strong>)
<a class="jxr_linenumber" name="L408" href="#L408">408</a> 				{
<a class="jxr_linenumber" name="L409" href="#L409">409</a> 					ret = (<a name="UnificationXref" href="../../../../org/biopax/paxtools/model/level3/UnificationXref.html#UnificationXref">UnificationXref</a>)x;
<a class="jxr_linenumber" name="L410" href="#L410">410</a> 				}
<a class="jxr_linenumber" name="L411" href="#L411">411</a> 				<strong class="jxr_keyword">else</strong> <strong class="jxr_keyword">if</strong> (ret.getDb().equalsIgnoreCase(x.getDb()) &amp;&amp; !ret.getId().equals(x.getId())) {
<a class="jxr_linenumber" name="L412" href="#L412">412</a> 					<em class="jxr_comment">//exactly the same kind (same 'db') xref is found;</em>
<a class="jxr_linenumber" name="L413" href="#L413">413</a> 					<em class="jxr_comment">//we can't test here if several IDs map to the same thing;</em>
<a class="jxr_linenumber" name="L414" href="#L414">414</a> 					<em class="jxr_comment">//so, give up - for safety - and return null.</em>
<a class="jxr_linenumber" name="L415" href="#L415">415</a> 					ret = <strong class="jxr_keyword">null</strong>;
<a class="jxr_linenumber" name="L416" href="#L416">416</a> 					<strong class="jxr_keyword">break</strong>;
<a class="jxr_linenumber" name="L417" href="#L417">417</a> 				}
<a class="jxr_linenumber" name="L418" href="#L418">418</a> 			}
<a class="jxr_linenumber" name="L419" href="#L419">419</a> 		}
<a class="jxr_linenumber" name="L420" href="#L420">420</a> 
<a class="jxr_linenumber" name="L421" href="#L421">421</a> 		<strong class="jxr_keyword">return</strong> ret;
<a class="jxr_linenumber" name="L422" href="#L422">422</a> 	}
<a class="jxr_linenumber" name="L423" href="#L423">423</a> 
<a class="jxr_linenumber" name="L424" href="#L424">424</a> 	
<a class="jxr_linenumber" name="L425" href="#L425">425</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L426" href="#L426">426</a> <em class="jxr_javadoccomment">	 * BioPAX normalization </em>
<a class="jxr_linenumber" name="L427" href="#L427">427</a> <em class="jxr_javadoccomment">	 * (modifies the original Model)</em>
<a class="jxr_linenumber" name="L428" href="#L428">428</a> <em class="jxr_javadoccomment">	 * </em>
<a class="jxr_linenumber" name="L429" href="#L429">429</a> <em class="jxr_javadoccomment">	 * @param model BioPAX model to normalize</em>
<a class="jxr_linenumber" name="L430" href="#L430">430</a> <em class="jxr_javadoccomment">	 * @throws NullPointerException if model is null</em>
<a class="jxr_linenumber" name="L431" href="#L431">431</a> <em class="jxr_javadoccomment">	 * @throws IllegalArgumentException if model is not Level3 BioPAX</em>
<a class="jxr_linenumber" name="L432" href="#L432">432</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L433" href="#L433">433</a> 	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">void</strong> normalize(<a name="Model" href="../../../../org/biopax/paxtools/model/Model.html#Model">Model</a> model) {
<a class="jxr_linenumber" name="L434" href="#L434">434</a> 		
<a class="jxr_linenumber" name="L435" href="#L435">435</a> 		<strong class="jxr_keyword">if</strong>(model.getLevel() != BioPAXLevel.L3)
<a class="jxr_linenumber" name="L436" href="#L436">436</a> 			<strong class="jxr_keyword">throw</strong> <strong class="jxr_keyword">new</strong> IllegalArgumentException(<span class="jxr_string">"Not Level3 model. "</span> +
<a class="jxr_linenumber" name="L437" href="#L437">437</a> 				<span class="jxr_string">"Consider converting it first (e.g., with the PaxTools)."</span>);
<a class="jxr_linenumber" name="L438" href="#L438">438</a> 		
<a class="jxr_linenumber" name="L439" href="#L439">439</a> 		<em class="jxr_comment">//if set, update the xml:base</em>
<a class="jxr_linenumber" name="L440" href="#L440">440</a> 		<strong class="jxr_keyword">if</strong>(xmlBase != <strong class="jxr_keyword">null</strong> &amp;&amp; !xmlBase.isEmpty())
<a class="jxr_linenumber" name="L441" href="#L441">441</a> 			model.setXmlBase(xmlBase);
<a class="jxr_linenumber" name="L442" href="#L442">442</a> 
<a class="jxr_linenumber" name="L443" href="#L443">443</a> 		<em class="jxr_comment">// Normalize/merge xrefs, first, and then CVs</em>
<a class="jxr_linenumber" name="L444" href="#L444">444</a> 		<em class="jxr_comment">// (also because some of original xrefs might have "normalized" URIs </em>
<a class="jxr_linenumber" name="L445" href="#L445">445</a> 		<em class="jxr_comment">// that, in fact, must be used for other biopax types, such as CV or ProteinReference)</em>
<a class="jxr_linenumber" name="L446" href="#L446">446</a> 		log.info(<span class="jxr_string">"Normalizing xrefs..."</span> + description);
<a class="jxr_linenumber" name="L447" href="#L447">447</a> 		normalizeXrefs(model);
<a class="jxr_linenumber" name="L448" href="#L448">448</a> 		
<a class="jxr_linenumber" name="L449" href="#L449">449</a> 		<em class="jxr_comment">// fix displayName where possible</em>
<a class="jxr_linenumber" name="L450" href="#L450">450</a> 		<strong class="jxr_keyword">if</strong>(fixDisplayName) {
<a class="jxr_linenumber" name="L451" href="#L451">451</a> 			log.info(<span class="jxr_string">"Normalizing display names..."</span> + description);
<a class="jxr_linenumber" name="L452" href="#L452">452</a> 			fixDisplayName(model);
<a class="jxr_linenumber" name="L453" href="#L453">453</a> 		}
<a class="jxr_linenumber" name="L454" href="#L454">454</a> 			
<a class="jxr_linenumber" name="L455" href="#L455">455</a> 		log.info(<span class="jxr_string">"Normalizing CVs..."</span> + description);
<a class="jxr_linenumber" name="L456" href="#L456">456</a> 		normalizeCVs(model);
<a class="jxr_linenumber" name="L457" href="#L457">457</a> 		
<a class="jxr_linenumber" name="L458" href="#L458">458</a> 		<em class="jxr_comment">//normalize BioSource objects (better, as it is here, go after Xrefs and CVs)</em>
<a class="jxr_linenumber" name="L459" href="#L459">459</a> 		log.info(<span class="jxr_string">"Normalizing organisms..."</span> + description);
<a class="jxr_linenumber" name="L460" href="#L460">460</a> 		normalizeBioSources(model);
<a class="jxr_linenumber" name="L461" href="#L461">461</a> 
<a class="jxr_linenumber" name="L462" href="#L462">462</a> 		<em class="jxr_comment">// auto-generate missing entity references:</em>
<a class="jxr_linenumber" name="L463" href="#L463">463</a> 		<strong class="jxr_keyword">for</strong>(<a name="SimplePhysicalEntity" href="../../../../org/biopax/paxtools/model/level3/SimplePhysicalEntity.html#SimplePhysicalEntity">SimplePhysicalEntity</a> spe : <strong class="jxr_keyword">new</strong> HashSet&lt;SimplePhysicalEntity&gt;(model.getObjects(SimplePhysicalEntity.<strong class="jxr_keyword">class</strong>))) {
<a class="jxr_linenumber" name="L464" href="#L464">464</a> 			<em class="jxr_comment">//it skips if spe has entityReference or memberPE already</em>
<a class="jxr_linenumber" name="L465" href="#L465">465</a> 			ModelUtils.addMissingEntityReference(model, spe);
<a class="jxr_linenumber" name="L466" href="#L466">466</a> 		}
<a class="jxr_linenumber" name="L467" href="#L467">467</a> 
<a class="jxr_linenumber" name="L468" href="#L468">468</a> 		log.info(<span class="jxr_string">"Normalizing entity references..."</span> + description);
<a class="jxr_linenumber" name="L469" href="#L469">469</a> 		normalizeERs(model);
<a class="jxr_linenumber" name="L470" href="#L470">470</a> 		
<a class="jxr_linenumber" name="L471" href="#L471">471</a> 		<em class="jxr_comment">// find/add lost (in replace) children</em>
<a class="jxr_linenumber" name="L472" href="#L472">472</a> 		log.info(<span class="jxr_string">"Repairing..."</span> + description);
<a class="jxr_linenumber" name="L473" href="#L473">473</a> 		model.repair(); <em class="jxr_comment">// it does not remove dangling utility class objects (can be done separately, later, if needed)</em>
<a class="jxr_linenumber" name="L474" href="#L474">474</a> 		
<a class="jxr_linenumber" name="L475" href="#L475">475</a> 		log.info(<span class="jxr_string">"Optional tasks (reasoning)..."</span> + description);
<a class="jxr_linenumber" name="L476" href="#L476">476</a> 	}
<a class="jxr_linenumber" name="L477" href="#L477">477</a> 
<a class="jxr_linenumber" name="L478" href="#L478">478</a> 	
<a class="jxr_linenumber" name="L479" href="#L479">479</a> 	<strong class="jxr_keyword">private</strong> <strong class="jxr_keyword">void</strong> normalizeCVs(<a name="Model" href="../../../../org/biopax/paxtools/model/Model.html#Model">Model</a> model) {
<a class="jxr_linenumber" name="L480" href="#L480">480</a> 		
<a class="jxr_linenumber" name="L481" href="#L481">481</a> 		NormalizerMap map = <strong class="jxr_keyword">new</strong> NormalizerMap(model);
<a class="jxr_linenumber" name="L482" href="#L482">482</a> 		
<a class="jxr_linenumber" name="L483" href="#L483">483</a> 		<em class="jxr_comment">// process ControlledVocabulary objects (all sub-classes)</em>
<a class="jxr_linenumber" name="L484" href="#L484">484</a> 		<strong class="jxr_keyword">for</strong>(<a name="ControlledVocabulary" href="../../../../org/biopax/paxtools/model/level3/ControlledVocabulary.html#ControlledVocabulary">ControlledVocabulary</a> cv : model.getObjects(ControlledVocabulary.<strong class="jxr_keyword">class</strong>))
<a class="jxr_linenumber" name="L485" href="#L485">485</a> 		{
<a class="jxr_linenumber" name="L486" href="#L486">486</a> 			<em class="jxr_comment">//it does not check/fix the CV terms though (but a validation rule can do if run before the normalizer)...</em>
<a class="jxr_linenumber" name="L487" href="#L487">487</a> 			UnificationXref uref = findPreferredUnificationXref(cv); <em class="jxr_comment">//usually, there's only one such xref</em>
<a class="jxr_linenumber" name="L488" href="#L488">488</a> 			<strong class="jxr_keyword">if</strong> (uref != <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L489" href="#L489">489</a> 				<em class="jxr_comment">// so let's generate a consistent URI</em>
<a class="jxr_linenumber" name="L490" href="#L490">490</a> 				map.put(cv, uri(xmlBase, uref.getDb(), uref.getId(), cv.getModelInterface()));
<a class="jxr_linenumber" name="L491" href="#L491">491</a> 			} <strong class="jxr_keyword">else</strong> <strong class="jxr_keyword">if</strong>(!cv.getTerm().isEmpty()) {
<a class="jxr_linenumber" name="L492" href="#L492">492</a> 				map.put(cv, uri(xmlBase, <strong class="jxr_keyword">null</strong>, cv.getTerm().iterator().next(), cv.getModelInterface()));
<a class="jxr_linenumber" name="L493" href="#L493">493</a> 			} <strong class="jxr_keyword">else</strong> log.info(<span class="jxr_string">"Cannot normalize "</span> + cv.getModelInterface().getSimpleName() 
<a class="jxr_linenumber" name="L494" href="#L494">494</a> 				+ <span class="jxr_string">" : no unification xrefs nor terms found in "</span> + cv.getUri()
<a class="jxr_linenumber" name="L495" href="#L495">495</a> 				+ <span class="jxr_string">". "</span> + description);
<a class="jxr_linenumber" name="L496" href="#L496">496</a> 		} 
<a class="jxr_linenumber" name="L497" href="#L497">497</a> 		
<a class="jxr_linenumber" name="L498" href="#L498">498</a> 		<em class="jxr_comment">// replace/update elements in the model</em>
<a class="jxr_linenumber" name="L499" href="#L499">499</a> 		map.doSubs();
<a class="jxr_linenumber" name="L500" href="#L500">500</a> 	}
<a class="jxr_linenumber" name="L501" href="#L501">501</a> 	
<a class="jxr_linenumber" name="L502" href="#L502">502</a> 	
<a class="jxr_linenumber" name="L503" href="#L503">503</a> 	<strong class="jxr_keyword">private</strong> <strong class="jxr_keyword">void</strong> normalizeBioSources(<a name="Model" href="../../../../org/biopax/paxtools/model/Model.html#Model">Model</a> model) {
<a class="jxr_linenumber" name="L504" href="#L504">504</a> 		
<a class="jxr_linenumber" name="L505" href="#L505">505</a> 		NormalizerMap map = <strong class="jxr_keyword">new</strong> NormalizerMap(model);
<a class="jxr_linenumber" name="L506" href="#L506">506</a> 		
<a class="jxr_linenumber" name="L507" href="#L507">507</a> 		<strong class="jxr_keyword">for</strong>(<a name="BioSource" href="../../../../org/biopax/paxtools/model/level3/BioSource.html#BioSource">BioSource</a> bs : model.getObjects(BioSource.<strong class="jxr_keyword">class</strong>))
<a class="jxr_linenumber" name="L508" href="#L508">508</a> 		{
<a class="jxr_linenumber" name="L509" href="#L509">509</a> 			UnificationXref uref = findPreferredUnificationXref(bs);
<a class="jxr_linenumber" name="L510" href="#L510">510</a> 			<em class="jxr_comment">//normally, the xref db is 'Taxonomy' (or a valid synonym)</em>
<a class="jxr_linenumber" name="L511" href="#L511">511</a> 			<strong class="jxr_keyword">if</strong> (uref != <strong class="jxr_keyword">null</strong>
<a class="jxr_linenumber" name="L512" href="#L512">512</a> 				&amp;&amp; (uref.getDb().toLowerCase().contains(<span class="jxr_string">"taxonomy"</span>) || uref.getDb().equalsIgnoreCase(<span class="jxr_string">"newt"</span>)))
<a class="jxr_linenumber" name="L513" href="#L513">513</a> 			{	
<a class="jxr_linenumber" name="L514" href="#L514">514</a> 				String 	idPart = uref.getId();
<a class="jxr_linenumber" name="L515" href="#L515">515</a> 
<a class="jxr_linenumber" name="L516" href="#L516">516</a> 				<em class="jxr_comment">//tissue/cellType terms can be added below:</em>
<a class="jxr_linenumber" name="L517" href="#L517">517</a> 				<strong class="jxr_keyword">if</strong>(bs.getTissue()!=<strong class="jxr_keyword">null</strong> &amp;&amp; !bs.getTissue().getTerm().isEmpty()) 
<a class="jxr_linenumber" name="L518" href="#L518">518</a> 					idPart += <span class="jxr_string">"_"</span> + bs.getTissue().getTerm().iterator().next();
<a class="jxr_linenumber" name="L519" href="#L519">519</a> 				<strong class="jxr_keyword">if</strong>(bs.getCellType()!=<strong class="jxr_keyword">null</strong> &amp;&amp; !bs.getCellType().getTerm().isEmpty()) 
<a class="jxr_linenumber" name="L520" href="#L520">520</a> 					idPart += <span class="jxr_string">"_"</span> + bs.getCellType().getTerm().iterator().next();
<a class="jxr_linenumber" name="L521" href="#L521">521</a> 				
<a class="jxr_linenumber" name="L522" href="#L522">522</a> 				String uri = (idPart.equals(uref.getId()) <em class="jxr_comment">//- no tissue or celltype were attached</em>
<a class="jxr_linenumber" name="L523" href="#L523">523</a> 						&amp;&amp; idPart.matches(<span class="jxr_string">"^&#92;&#92;d+$"</span>)) <em class="jxr_comment">//- is positive integer id</em>
<a class="jxr_linenumber" name="L524" href="#L524">524</a> 					? uri(xmlBase, uref.getDb(), idPart, BioSource.<strong class="jxr_keyword">class</strong>)
<a class="jxr_linenumber" name="L525" href="#L525">525</a> 						: <span class="jxr_string">"http://identifiers.org/taxonomy/"</span> + idPart;
<a class="jxr_linenumber" name="L526" href="#L526">526</a> 					<em class="jxr_comment">// the latter is intentionally invalid identifiers.org/taxonomy URI - good (and important) for merging</em>
<a class="jxr_linenumber" name="L527" href="#L527">527</a> 				map.put(bs, uri);
<a class="jxr_linenumber" name="L528" href="#L528">528</a> 
<a class="jxr_linenumber" name="L529" href="#L529">529</a> 			} <strong class="jxr_keyword">else</strong> 
<a class="jxr_linenumber" name="L530" href="#L530">530</a> 				log.debug(<span class="jxr_string">"Won't normalize BioSource"</span> 
<a class="jxr_linenumber" name="L531" href="#L531">531</a> 					+ <span class="jxr_string">" : no taxonomy unification xref found in "</span> + bs.getUri()
<a class="jxr_linenumber" name="L532" href="#L532">532</a> 					+ <span class="jxr_string">". "</span> + description);
<a class="jxr_linenumber" name="L533" href="#L533">533</a> 		} 
<a class="jxr_linenumber" name="L534" href="#L534">534</a> 		
<a class="jxr_linenumber" name="L535" href="#L535">535</a> 		map.doSubs();
<a class="jxr_linenumber" name="L536" href="#L536">536</a> 	}
<a class="jxr_linenumber" name="L537" href="#L537">537</a> 
<a class="jxr_linenumber" name="L538" href="#L538">538</a> 	<strong class="jxr_keyword">private</strong> <strong class="jxr_keyword">void</strong> normalizeERs(<a name="Model" href="../../../../org/biopax/paxtools/model/Model.html#Model">Model</a> model) {
<a class="jxr_linenumber" name="L539" href="#L539">539</a> 		
<a class="jxr_linenumber" name="L540" href="#L540">540</a> 		NormalizerMap map = <strong class="jxr_keyword">new</strong> NormalizerMap(model);
<a class="jxr_linenumber" name="L541" href="#L541">541</a> 		
<a class="jxr_linenumber" name="L542" href="#L542">542</a> 		<em class="jxr_comment">// process the rest of utility classes (selectively though)</em>
<a class="jxr_linenumber" name="L543" href="#L543">543</a> 		<strong class="jxr_keyword">for</strong> (<a name="EntityReference" href="../../../../org/biopax/paxtools/model/level3/EntityReference.html#EntityReference">EntityReference</a> bpe : model.getObjects(EntityReference.<strong class="jxr_keyword">class</strong>)) {
<a class="jxr_linenumber" name="L544" href="#L544">544</a> 			
<a class="jxr_linenumber" name="L545" href="#L545">545</a> 			<em class="jxr_comment">//skip those with already normalized URIs</em>
<a class="jxr_linenumber" name="L546" href="#L546">546</a> 			<strong class="jxr_keyword">if</strong>(bpe.getUri().startsWith(<span class="jxr_string">"http://identifiers.org/"</span>)) {
<a class="jxr_linenumber" name="L547" href="#L547">547</a> 				log.info(<span class="jxr_string">"Skip already normalized: "</span> + bpe.getUri());
<a class="jxr_linenumber" name="L548" href="#L548">548</a> 				<strong class="jxr_keyword">continue</strong>;
<a class="jxr_linenumber" name="L549" href="#L549">549</a> 			}			
<a class="jxr_linenumber" name="L550" href="#L550">550</a> 			
<a class="jxr_linenumber" name="L551" href="#L551">551</a> 			UnificationXref uref = findPreferredUnificationXref(bpe);
<a class="jxr_linenumber" name="L552" href="#L552">552</a> 			<strong class="jxr_keyword">if</strong> (uref != <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L553" href="#L553">553</a> 				<em class="jxr_comment">// Create (with a new URI made from a unif. xref) </em>
<a class="jxr_linenumber" name="L554" href="#L554">554</a> 				<em class="jxr_comment">// and save the replacement object, if possible, </em>
<a class="jxr_linenumber" name="L555" href="#L555">555</a> 				<em class="jxr_comment">// but do not replace yet (will call doSubs later, for all).</em>
<a class="jxr_linenumber" name="L556" href="#L556">556</a> 				<strong class="jxr_keyword">final</strong> String db = uref.getDb();
<a class="jxr_linenumber" name="L557" href="#L557">557</a> 				<strong class="jxr_keyword">final</strong> String id = uref.getId();
<a class="jxr_linenumber" name="L558" href="#L558">558</a> 				<em class="jxr_comment">// get the standard ID</em>
<a class="jxr_linenumber" name="L559" href="#L559">559</a> 				String uri = <strong class="jxr_keyword">null</strong>;
<a class="jxr_linenumber" name="L560" href="#L560">560</a> 				<strong class="jxr_keyword">try</strong> { <em class="jxr_comment">// make a new ID for the element</em>
<a class="jxr_linenumber" name="L561" href="#L561">561</a> 					uri = MiriamLink.getIdentifiersOrgURI(db, id);
<a class="jxr_linenumber" name="L562" href="#L562">562</a> 				} <strong class="jxr_keyword">catch</strong> (Exception e) {
<a class="jxr_linenumber" name="L563" href="#L563">563</a> 					log.error(<span class="jxr_string">"Cannot get a Miriam standard ID for "</span> + bpe 
<a class="jxr_linenumber" name="L564" href="#L564">564</a> 							+ <span class="jxr_string">" ("</span> + bpe.getModelInterface().getSimpleName()
<a class="jxr_linenumber" name="L565" href="#L565">565</a> 							+ <span class="jxr_string">") "</span> + <span class="jxr_string">", using "</span> + db + <span class="jxr_string">":"</span> + id 
<a class="jxr_linenumber" name="L566" href="#L566">566</a> 							+ <span class="jxr_string">". "</span> + e.getMessage());
<a class="jxr_linenumber" name="L567" href="#L567">567</a> 					<strong class="jxr_keyword">return</strong>;
<a class="jxr_linenumber" name="L568" href="#L568">568</a> 				}
<a class="jxr_linenumber" name="L569" href="#L569">569</a> 
<a class="jxr_linenumber" name="L570" href="#L570">570</a> 				<strong class="jxr_keyword">if</strong>(uri != <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L571" href="#L571">571</a> 					map.put(bpe, uri);
<a class="jxr_linenumber" name="L572" href="#L572">572</a> 				}	
<a class="jxr_linenumber" name="L573" href="#L573">573</a> 			} <strong class="jxr_keyword">else</strong>
<a class="jxr_linenumber" name="L574" href="#L574">574</a> 				log.info(<span class="jxr_string">"Cannot normalize EntityReference: "</span>
<a class="jxr_linenumber" name="L575" href="#L575">575</a> 					+ <span class="jxr_string">"no unification xrefs found in "</span> + bpe.getUri()
<a class="jxr_linenumber" name="L576" href="#L576">576</a> 					+ <span class="jxr_string">". "</span> + description);
<a class="jxr_linenumber" name="L577" href="#L577">577</a> 		}
<a class="jxr_linenumber" name="L578" href="#L578">578</a> 		
<a class="jxr_linenumber" name="L579" href="#L579">579</a> 		<em class="jxr_comment">// replace/update elements in the model</em>
<a class="jxr_linenumber" name="L580" href="#L580">580</a> 		map.doSubs();
<a class="jxr_linenumber" name="L581" href="#L581">581</a> 	}
<a class="jxr_linenumber" name="L582" href="#L582">582</a> 	
<a class="jxr_linenumber" name="L583" href="#L583">583</a> 	
<a class="jxr_linenumber" name="L584" href="#L584">584</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L585" href="#L585">585</a> <em class="jxr_javadoccomment">	 * Auto-generates standard and other names for the datasource</em>
<a class="jxr_linenumber" name="L586" href="#L586">586</a> <em class="jxr_javadoccomment">	 * from either its ID (if URN) or one of its existing names (preferably - standard name)</em>
<a class="jxr_linenumber" name="L587" href="#L587">587</a> <em class="jxr_javadoccomment">	 * </em>
<a class="jxr_linenumber" name="L588" href="#L588">588</a> <em class="jxr_javadoccomment">	 * @param pro data source (BioPAX Provenance)</em>
<a class="jxr_linenumber" name="L589" href="#L589">589</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L590" href="#L590">590</a> 	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">static</strong> <strong class="jxr_keyword">void</strong> autoName(<a name="Provenance" href="../../../../org/biopax/paxtools/model/level3/Provenance.html#Provenance">Provenance</a> pro) {
<a class="jxr_linenumber" name="L591" href="#L591">591</a> 		<strong class="jxr_keyword">if</strong>(!(pro.getUri().startsWith(<span class="jxr_string">"urn:miriam:"</span>) || pro.getUri().startsWith(<span class="jxr_string">"http://identifiers.org/"</span>))
<a class="jxr_linenumber" name="L592" href="#L592">592</a> 				&amp;&amp; pro.getName().isEmpty()) {
<a class="jxr_linenumber" name="L593" href="#L593">593</a> 			log.info(<span class="jxr_string">"Skipping: cannot normalize Provenance: "</span> + pro.getUri());
<a class="jxr_linenumber" name="L594" href="#L594">594</a> 		}
<a class="jxr_linenumber" name="L595" href="#L595">595</a> 		<strong class="jxr_keyword">else</strong> { <em class="jxr_comment">// i.e., 'name' is not empty or ID is the URN</em>
<a class="jxr_linenumber" name="L596" href="#L596">596</a> 			<strong class="jxr_keyword">final</strong> SortedSet&lt;String&gt; names = <strong class="jxr_keyword">new</strong> TreeSet&lt;String&gt;();
<a class="jxr_linenumber" name="L597" href="#L597">597</a> 			
<a class="jxr_linenumber" name="L598" href="#L598">598</a> 			String key = <strong class="jxr_keyword">null</strong>;
<a class="jxr_linenumber" name="L599" href="#L599">599</a> 			<strong class="jxr_keyword">if</strong>(pro.getUri().startsWith(<span class="jxr_string">"urn:miriam:"</span>) || pro.getUri().startsWith(<span class="jxr_string">"http://identifiers.org/"</span>)) {
<a class="jxr_linenumber" name="L600" href="#L600">600</a> 				key = pro.getUri();
<a class="jxr_linenumber" name="L601" href="#L601">601</a> 			} <strong class="jxr_keyword">else</strong> <strong class="jxr_keyword">if</strong> (pro.getStandardName() != <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L602" href="#L602">602</a> 				key = pro.getStandardName();
<a class="jxr_linenumber" name="L603" href="#L603">603</a> 			} <strong class="jxr_keyword">else</strong> {
<a class="jxr_linenumber" name="L604" href="#L604">604</a> 				key = pro.getDisplayName(); <em class="jxr_comment">// can be null</em>
<a class="jxr_linenumber" name="L605" href="#L605">605</a> 			}
<a class="jxr_linenumber" name="L606" href="#L606">606</a> 			
<a class="jxr_linenumber" name="L607" href="#L607">607</a> 			<strong class="jxr_keyword">if</strong> (key != <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L608" href="#L608">608</a> 				<strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L609" href="#L609">609</a> 					names.addAll(Arrays.asList(MiriamLink.getNames(key)));
<a class="jxr_linenumber" name="L610" href="#L610">610</a> 					pro.setStandardName(MiriamLink.getName(key));
<a class="jxr_linenumber" name="L611" href="#L611">611</a> 					<em class="jxr_comment">// get the datasource description</em>
<a class="jxr_linenumber" name="L612" href="#L612">612</a> 					String description = MiriamLink.getDataTypeDef(pro.getStandardName());
<a class="jxr_linenumber" name="L613" href="#L613">613</a> 					pro.addComment(description);
<a class="jxr_linenumber" name="L614" href="#L614">614</a> 				} <strong class="jxr_keyword">catch</strong> (IllegalArgumentException e) {
<a class="jxr_linenumber" name="L615" href="#L615">615</a> 					<em class="jxr_comment">// ignore (then, names is still empty...)</em>
<a class="jxr_linenumber" name="L616" href="#L616">616</a> 				}
<a class="jxr_linenumber" name="L617" href="#L617">617</a> 			} 
<a class="jxr_linenumber" name="L618" href="#L618">618</a> 			
<a class="jxr_linenumber" name="L619" href="#L619">619</a> 			<em class="jxr_comment">// when the above failed (no match in Miriam), or key was null -</em>
<a class="jxr_linenumber" name="L620" href="#L620">620</a> 			<strong class="jxr_keyword">if</strong>(names.isEmpty()) {
<a class="jxr_linenumber" name="L621" href="#L621">621</a> 				<em class="jxr_comment">// finally, trying to find all valid names for each existing one</em>
<a class="jxr_linenumber" name="L622" href="#L622">622</a> 					<strong class="jxr_keyword">for</strong> (String name : pro.getName()) {
<a class="jxr_linenumber" name="L623" href="#L623">623</a> 						<strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L624" href="#L624">624</a> 							names.addAll(Arrays.asList(MiriamLink.getNames(name)));
<a class="jxr_linenumber" name="L625" href="#L625">625</a> 						} <strong class="jxr_keyword">catch</strong> (IllegalArgumentException e) {
<a class="jxr_linenumber" name="L626" href="#L626">626</a> 							<em class="jxr_comment">// ignore</em>
<a class="jxr_linenumber" name="L627" href="#L627">627</a> 						}
<a class="jxr_linenumber" name="L628" href="#L628">628</a> 					}
<a class="jxr_linenumber" name="L629" href="#L629">629</a> 					<em class="jxr_comment">// pick up the first name, get the standard name</em>
<a class="jxr_linenumber" name="L630" href="#L630">630</a> 					<strong class="jxr_keyword">if</strong>(!names.isEmpty())
<a class="jxr_linenumber" name="L631" href="#L631">631</a> 						pro.setStandardName(<a name="MiriamLink" href="../../../../org/biopax/paxtools/normalizer/MiriamLink.html#MiriamLink">MiriamLink</a>
<a class="jxr_linenumber" name="L632" href="#L632">632</a> 							.getName(names.iterator().next()));
<a class="jxr_linenumber" name="L633" href="#L633">633</a> 			}
<a class="jxr_linenumber" name="L634" href="#L634">634</a> 			
<a class="jxr_linenumber" name="L635" href="#L635">635</a> 			<em class="jxr_comment">// and add all the synonyms if any</em>
<a class="jxr_linenumber" name="L636" href="#L636">636</a> 			<strong class="jxr_keyword">for</strong>(String name : names)
<a class="jxr_linenumber" name="L637" href="#L637">637</a> 				pro.addName(name);
<a class="jxr_linenumber" name="L638" href="#L638">638</a> 			
<a class="jxr_linenumber" name="L639" href="#L639">639</a> 			<em class="jxr_comment">//set display name if not set (standard name is set already)</em>
<a class="jxr_linenumber" name="L640" href="#L640">640</a> 			<strong class="jxr_keyword">if</strong>(pro.getDisplayName() == <strong class="jxr_keyword">null</strong>)
<a class="jxr_linenumber" name="L641" href="#L641">641</a> 				pro.setDisplayName(pro.getStandardName());			
<a class="jxr_linenumber" name="L642" href="#L642">642</a> 		}
<a class="jxr_linenumber" name="L643" href="#L643">643</a> 	}
<a class="jxr_linenumber" name="L644" href="#L644">644</a> 	
<a class="jxr_linenumber" name="L645" href="#L645">645</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L646" href="#L646">646</a> <em class="jxr_javadoccomment">	 * Converts BioPAX L1 or L2 RDF/XML string data to BioPAX L3 string.</em>
<a class="jxr_linenumber" name="L647" href="#L647">647</a> <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L648" href="#L648">648</a> <em class="jxr_javadoccomment">	 * WARN: this is not for huge (larger than 1GB) BioPAX RDF/XML data</em>
<a class="jxr_linenumber" name="L649" href="#L649">649</a> <em class="jxr_javadoccomment">	 * due to use of (UTF-8) String and Byte Array internally.</em>
<a class="jxr_linenumber" name="L650" href="#L650">650</a> <em class="jxr_javadoccomment">	 * This can be and is used by online web apps, such as the</em>
<a class="jxr_linenumber" name="L651" href="#L651">651</a> <em class="jxr_javadoccomment">	 * BioPAX Validator.</em>
<a class="jxr_linenumber" name="L652" href="#L652">652</a> <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L653" href="#L653">653</a> <em class="jxr_javadoccomment">	 * @param biopaxData String</em>
<a class="jxr_linenumber" name="L654" href="#L654">654</a> <em class="jxr_javadoccomment">	 * @return BioPAX Level3 RDF/XML string</em>
<a class="jxr_linenumber" name="L655" href="#L655">655</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L656" href="#L656">656</a> 	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">static</strong> String convertToLevel3(<strong class="jxr_keyword">final</strong> String biopaxData) {
<a class="jxr_linenumber" name="L657" href="#L657">657</a> 		String toReturn = <span class="jxr_string">""</span>;
<a class="jxr_linenumber" name="L658" href="#L658">658</a> 		
<a class="jxr_linenumber" name="L659" href="#L659">659</a> 		<strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L660" href="#L660">660</a> 			ByteArrayOutputStream os = <strong class="jxr_keyword">new</strong> ByteArrayOutputStream();
<a class="jxr_linenumber" name="L661" href="#L661">661</a> 			InputStream is = <strong class="jxr_keyword">new</strong> ByteArrayInputStream(biopaxData.getBytes());
<a class="jxr_linenumber" name="L662" href="#L662">662</a> 			SimpleIOHandler io = <strong class="jxr_keyword">new</strong> <a name="SimpleIOHandler" href="../../../../org/biopax/paxtools/io/SimpleIOHandler.html#SimpleIOHandler">SimpleIOHandler</a>();
<a class="jxr_linenumber" name="L663" href="#L663">663</a> 			io.mergeDuplicates(<strong class="jxr_keyword">true</strong>);
<a class="jxr_linenumber" name="L664" href="#L664">664</a> 			Model model = io.convertFromOWL(is);
<a class="jxr_linenumber" name="L665" href="#L665">665</a> 			<strong class="jxr_keyword">if</strong> (model.getLevel() != BioPAXLevel.L3) {
<a class="jxr_linenumber" name="L666" href="#L666">666</a> 				log.info(<span class="jxr_string">"Converting to BioPAX Level3... "</span> + model.getXmlBase());
<a class="jxr_linenumber" name="L667" href="#L667">667</a> 				model = (<strong class="jxr_keyword">new</strong> <a name="LevelUpgrader" href="../../../../org/biopax/paxtools/converter/LevelUpgrader.html#LevelUpgrader">LevelUpgrader</a>()).filter(model);
<a class="jxr_linenumber" name="L668" href="#L668">668</a> 				<strong class="jxr_keyword">if</strong> (model != <strong class="jxr_keyword">null</strong>) {
<a class="jxr_linenumber" name="L669" href="#L669">669</a> 					io.setFactory(model.getLevel().getDefaultFactory());
<a class="jxr_linenumber" name="L670" href="#L670">670</a> 					io.convertToOWL(model, os);
<a class="jxr_linenumber" name="L671" href="#L671">671</a> 					toReturn = os.toString();
<a class="jxr_linenumber" name="L672" href="#L672">672</a> 				}
<a class="jxr_linenumber" name="L673" href="#L673">673</a> 			} <strong class="jxr_keyword">else</strong> {
<a class="jxr_linenumber" name="L674" href="#L674">674</a> 				toReturn = biopaxData;
<a class="jxr_linenumber" name="L675" href="#L675">675</a> 			}
<a class="jxr_linenumber" name="L676" href="#L676">676</a> 		} <strong class="jxr_keyword">catch</strong>(Exception e) {
<a class="jxr_linenumber" name="L677" href="#L677">677</a> 			<strong class="jxr_keyword">throw</strong> <strong class="jxr_keyword">new</strong> RuntimeException(
<a class="jxr_linenumber" name="L678" href="#L678">678</a> 				<span class="jxr_string">"Cannot convert to BioPAX Level3"</span>, e);
<a class="jxr_linenumber" name="L679" href="#L679">679</a> 		}
<a class="jxr_linenumber" name="L680" href="#L680">680</a> 
<a class="jxr_linenumber" name="L681" href="#L681">681</a> 		<strong class="jxr_keyword">return</strong> toReturn;
<a class="jxr_linenumber" name="L682" href="#L682">682</a> 	}
<a class="jxr_linenumber" name="L683" href="#L683">683</a> 	
<a class="jxr_linenumber" name="L684" href="#L684">684</a> 	
<a class="jxr_linenumber" name="L685" href="#L685">685</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L686" href="#L686">686</a> <em class="jxr_javadoccomment">	 * Gets the xml:base to use with newly created BioPAX elements.</em>
<a class="jxr_linenumber" name="L687" href="#L687">687</a> <em class="jxr_javadoccomment">	 * </em>
<a class="jxr_linenumber" name="L688" href="#L688">688</a> <em class="jxr_javadoccomment">	 * @param modelToNormalize</em>
<a class="jxr_linenumber" name="L689" href="#L689">689</a> <em class="jxr_javadoccomment">	 * @return BioPAX model's xml:base or empty string if it's null</em>
<a class="jxr_linenumber" name="L690" href="#L690">690</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L691" href="#L691">691</a> 	<strong class="jxr_keyword">private</strong> String getXmlBase(<a name="Model" href="../../../../org/biopax/paxtools/model/Model.html#Model">Model</a> modelToNormalize) {
<a class="jxr_linenumber" name="L692" href="#L692">692</a> 		<strong class="jxr_keyword">if</strong>(xmlBase != <strong class="jxr_keyword">null</strong>)
<a class="jxr_linenumber" name="L693" href="#L693">693</a> 			<strong class="jxr_keyword">return</strong> xmlBase;
<a class="jxr_linenumber" name="L694" href="#L694">694</a> 		<strong class="jxr_keyword">else</strong>
<a class="jxr_linenumber" name="L695" href="#L695">695</a> 			<strong class="jxr_keyword">return</strong> <span class="jxr_string">""</span>;
<a class="jxr_linenumber" name="L696" href="#L696">696</a> 	}
<a class="jxr_linenumber" name="L697" href="#L697">697</a> 	
<a class="jxr_linenumber" name="L698" href="#L698">698</a> 	
<a class="jxr_linenumber" name="L699" href="#L699">699</a> 	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">boolean</strong> isFixDisplayName() {
<a class="jxr_linenumber" name="L700" href="#L700">700</a> 		<strong class="jxr_keyword">return</strong> fixDisplayName;
<a class="jxr_linenumber" name="L701" href="#L701">701</a> 	}
<a class="jxr_linenumber" name="L702" href="#L702">702</a> 	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">void</strong> setFixDisplayName(<strong class="jxr_keyword">boolean</strong> fixDisplayName) {
<a class="jxr_linenumber" name="L703" href="#L703">703</a> 		<strong class="jxr_keyword">this</strong>.fixDisplayName = fixDisplayName;
<a class="jxr_linenumber" name="L704" href="#L704">704</a> 	}
<a class="jxr_linenumber" name="L705" href="#L705">705</a> 
<a class="jxr_linenumber" name="L706" href="#L706">706</a> 	<strong class="jxr_keyword">public</strong> String getXmlBase() {
<a class="jxr_linenumber" name="L707" href="#L707">707</a> 		<strong class="jxr_keyword">return</strong> xmlBase;
<a class="jxr_linenumber" name="L708" href="#L708">708</a> 	}
<a class="jxr_linenumber" name="L709" href="#L709">709</a> 	<strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">void</strong> setXmlBase(String xmlBase) {
<a class="jxr_linenumber" name="L710" href="#L710">710</a> 		<strong class="jxr_keyword">this</strong>.xmlBase = xmlBase;
<a class="jxr_linenumber" name="L711" href="#L711">711</a> 	}
<a class="jxr_linenumber" name="L712" href="#L712">712</a> 
<a class="jxr_linenumber" name="L713" href="#L713">713</a> 	
<a class="jxr_linenumber" name="L714" href="#L714">714</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L715" href="#L715">715</a> <em class="jxr_javadoccomment">	 * Helper class, to associate original </em>
<a class="jxr_linenumber" name="L716" href="#L716">716</a> <em class="jxr_javadoccomment">	 * and new (either copy or existing) objects </em>
<a class="jxr_linenumber" name="L717" href="#L717">717</a> <em class="jxr_javadoccomment">	 * within some biopax model and then execute </em>
<a class="jxr_linenumber" name="L718" href="#L718">718</a> <em class="jxr_javadoccomment">	 * batch replace.</em>
<a class="jxr_linenumber" name="L719" href="#L719">719</a> <em class="jxr_javadoccomment">	 * </em>
<a class="jxr_linenumber" name="L720" href="#L720">720</a> <em class="jxr_javadoccomment">	 * @author rodche</em>
<a class="jxr_linenumber" name="L721" href="#L721">721</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L722" href="#L722">722</a> 	<strong class="jxr_keyword">private</strong> <strong class="jxr_keyword">static</strong> <strong class="jxr_keyword">class</strong> NormalizerMap {
<a class="jxr_linenumber" name="L723" href="#L723">723</a> 
<a class="jxr_linenumber" name="L724" href="#L724">724</a> 		<em class="jxr_comment">//a model to modify by replacing biopax objects</em>
<a class="jxr_linenumber" name="L725" href="#L725">725</a> 		<strong class="jxr_keyword">final</strong> <a name="Model" href="../../../../org/biopax/paxtools/model/Model.html#Model">Model</a> model;
<a class="jxr_linenumber" name="L726" href="#L726">726</a> 		
<a class="jxr_linenumber" name="L727" href="#L727">727</a> 		<em class="jxr_comment">// this is the biopax elements substitution map (old-&gt;new)</em>
<a class="jxr_linenumber" name="L728" href="#L728">728</a> 		<strong class="jxr_keyword">final</strong> Map&lt;BioPAXElement,BioPAXElement&gt; subs;
<a class="jxr_linenumber" name="L729" href="#L729">729</a> 		
<a class="jxr_linenumber" name="L730" href="#L730">730</a> 		<em class="jxr_comment">//the next map is to make sure the subs.values()  all have different URIs</em>
<a class="jxr_linenumber" name="L731" href="#L731">731</a> 		<strong class="jxr_keyword">final</strong> Map&lt;String,BioPAXElement&gt; uriToSub;
<a class="jxr_linenumber" name="L732" href="#L732">732</a> 		
<a class="jxr_linenumber" name="L733" href="#L733">733</a> 		<strong class="jxr_keyword">final</strong> <a name="ShallowCopy" href="../../../../org/biopax/paxtools/controller/ShallowCopy.html#ShallowCopy">ShallowCopy</a> copier;
<a class="jxr_linenumber" name="L734" href="#L734">734</a> 				
<a class="jxr_linenumber" name="L735" href="#L735">735</a> 		NormalizerMap(<a name="Model" href="../../../../org/biopax/paxtools/model/Model.html#Model">Model</a> model) {
<a class="jxr_linenumber" name="L736" href="#L736">736</a> 			subs = BPCollections.I.createMap();
<a class="jxr_linenumber" name="L737" href="#L737">737</a> 			uriToSub = BPCollections.I.createMap();
<a class="jxr_linenumber" name="L738" href="#L738">738</a> 			<strong class="jxr_keyword">this</strong>.model = model;
<a class="jxr_linenumber" name="L739" href="#L739">739</a> 			copier = <strong class="jxr_keyword">new</strong> <a name="ShallowCopy" href="../../../../org/biopax/paxtools/controller/ShallowCopy.html#ShallowCopy">ShallowCopy</a>();
<a class="jxr_linenumber" name="L740" href="#L740">740</a> 		}
<a class="jxr_linenumber" name="L741" href="#L741">741</a> 
<a class="jxr_linenumber" name="L742" href="#L742">742</a> 
<a class="jxr_linenumber" name="L743" href="#L743">743</a> 		<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L744" href="#L744">744</a> <em class="jxr_javadoccomment">		 * Creates (by URI) and saves the replacement object,</em>
<a class="jxr_linenumber" name="L745" href="#L745">745</a> <em class="jxr_javadoccomment">		 * but does not replace yet (call doSubs to replace).</em>
<a class="jxr_linenumber" name="L746" href="#L746">746</a> <em class="jxr_javadoccomment">		 * </em>
<a class="jxr_linenumber" name="L747" href="#L747">747</a> <em class="jxr_javadoccomment">		 * @param bpe</em>
<a class="jxr_linenumber" name="L748" href="#L748">748</a> <em class="jxr_javadoccomment">		 * @param newUri</em>
<a class="jxr_linenumber" name="L749" href="#L749">749</a> <em class="jxr_javadoccomment">		 */</em>
<a class="jxr_linenumber" name="L750" href="#L750">750</a> 		<strong class="jxr_keyword">void</strong> put(<a name="BioPAXElement" href="../../../../org/biopax/paxtools/model/BioPAXElement.html#BioPAXElement">BioPAXElement</a> bpe, String newUri)
<a class="jxr_linenumber" name="L751" href="#L751">751</a> 		{
<a class="jxr_linenumber" name="L752" href="#L752">752</a> 			<strong class="jxr_keyword">if</strong>(model.containsID(newUri)) {
<a class="jxr_linenumber" name="L753" href="#L753">753</a> 				<em class="jxr_comment">// will use existing original (model) object that has the new Uri</em>
<a class="jxr_linenumber" name="L754" href="#L754">754</a> 				map(bpe, model.getByID(newUri));
<a class="jxr_linenumber" name="L755" href="#L755">755</a> 			} <strong class="jxr_keyword">else</strong> <strong class="jxr_keyword">if</strong>(uriToSub.containsKey(newUri)) {
<a class="jxr_linenumber" name="L756" href="#L756">756</a> 				<em class="jxr_comment">// re-use the new object that's already added to replace another original</em>
<a class="jxr_linenumber" name="L757" href="#L757">757</a> 				map(bpe, uriToSub.get(newUri));
<a class="jxr_linenumber" name="L758" href="#L758">758</a> 			} <strong class="jxr_keyword">else</strong> {
<a class="jxr_linenumber" name="L759" href="#L759">759</a> 				<em class="jxr_comment">// will use the object's shallow copy that gets new Uri</em>
<a class="jxr_linenumber" name="L760" href="#L760">760</a> 				BioPAXElement copy = copier.copy(bpe, newUri);
<a class="jxr_linenumber" name="L761" href="#L761">761</a> 				map(bpe, copy);
<a class="jxr_linenumber" name="L762" href="#L762">762</a> 			}
<a class="jxr_linenumber" name="L763" href="#L763">763</a> 		}		
<a class="jxr_linenumber" name="L764" href="#L764">764</a> 		
<a class="jxr_linenumber" name="L765" href="#L765">765</a> 		<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L766" href="#L766">766</a> <em class="jxr_javadoccomment">		 * Executes the batch replace - migrating  </em>
<a class="jxr_linenumber" name="L767" href="#L767">767</a> <em class="jxr_javadoccomment">		 * to the normalized equivalent objects.</em>
<a class="jxr_linenumber" name="L768" href="#L768">768</a> <em class="jxr_javadoccomment">		 */</em>
<a class="jxr_linenumber" name="L769" href="#L769">769</a> 		<strong class="jxr_keyword">void</strong> doSubs() {
<a class="jxr_linenumber" name="L770" href="#L770">770</a> 			<strong class="jxr_keyword">for</strong>(<a name="BioPAXElement" href="../../../../org/biopax/paxtools/model/BioPAXElement.html#BioPAXElement">BioPAXElement</a> e : subs.keySet()) {
<a class="jxr_linenumber" name="L771" href="#L771">771</a> 				model.remove(e);
<a class="jxr_linenumber" name="L772" href="#L772">772</a> 			}
<a class="jxr_linenumber" name="L773" href="#L773">773</a> 			
<a class="jxr_linenumber" name="L774" href="#L774">774</a> 			<strong class="jxr_keyword">try</strong> {
<a class="jxr_linenumber" name="L775" href="#L775">775</a> 				ModelUtils.replace(model, subs);
<a class="jxr_linenumber" name="L776" href="#L776">776</a> 			} <strong class="jxr_keyword">catch</strong> (Exception e) {
<a class="jxr_linenumber" name="L777" href="#L777">777</a> 				log.error(<span class="jxr_string">"Failed to replace BioPAX elements."</span>, e);
<a class="jxr_linenumber" name="L778" href="#L778">778</a> 				<strong class="jxr_keyword">return</strong>;
<a class="jxr_linenumber" name="L779" href="#L779">779</a> 			}
<a class="jxr_linenumber" name="L780" href="#L780">780</a> 			
<a class="jxr_linenumber" name="L781" href="#L781">781</a> 			<strong class="jxr_keyword">for</strong>(<a name="BioPAXElement" href="../../../../org/biopax/paxtools/model/BioPAXElement.html#BioPAXElement">BioPAXElement</a> e : subs.values()) {
<a class="jxr_linenumber" name="L782" href="#L782">782</a> 				<strong class="jxr_keyword">if</strong>(!model.contains(e))
<a class="jxr_linenumber" name="L783" href="#L783">783</a> 					model.add(e);
<a class="jxr_linenumber" name="L784" href="#L784">784</a> 			}
<a class="jxr_linenumber" name="L785" href="#L785">785</a> 		
<a class="jxr_linenumber" name="L786" href="#L786">786</a> 			<strong class="jxr_keyword">for</strong>(<a name="BioPAXElement" href="../../../../org/biopax/paxtools/model/BioPAXElement.html#BioPAXElement">BioPAXElement</a> e : model.getObjects()) {
<a class="jxr_linenumber" name="L787" href="#L787">787</a> 				ModelUtils.fixDanglingInverseProperties(e, model);
<a class="jxr_linenumber" name="L788" href="#L788">788</a> 			}
<a class="jxr_linenumber" name="L789" href="#L789">789</a> 		}
<a class="jxr_linenumber" name="L790" href="#L790">790</a> 
<a class="jxr_linenumber" name="L791" href="#L791">791</a> 		<strong class="jxr_keyword">private</strong> <strong class="jxr_keyword">void</strong> map(<a name="BioPAXElem<a name="BioPAXElement" href="../../../../org/biopax/paxtools/model/BioPAXElement.html#BioPAXElement">BioPAXElement</a>/../../../org/biopax/paxtools/model/BioPAXElement.html#BioPAXElement">BioPAXElement</a> bpe, <a name="BioPAXElement" href="../../../../org/biopax/paxtools/model/BioPAXElement.html#BioPAXElement">BioPAXElement</a> newBpe) {
<a class="jxr_linenumber" name="L792" href="#L792">792</a> 			subs.put(bpe, newBpe);
<a class="jxr_linenumber" name="L793" href="#L793">793</a> 			uriToSub.put(newBpe.getUri(), newBpe);
<a class="jxr_linenumber" name="L794" href="#L794">794</a> 		}
<a class="jxr_linenumber" name="L795" href="#L795">795</a> 	}
<a class="jxr_linenumber" name="L796" href="#L796">796</a> 	
<a class="jxr_linenumber" name="L797" href="#L797">797</a> }
</pre>
<hr/>
<div id="footer">Copyright &#169; 2023 <a href="http://www.biopax.org">BioPAX</a>. All rights reserved.</div>
</body>
</html>
