<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<html lang="en">
<head>
<title>Source code</title>
<link rel="stylesheet" type="text/css" href="../../../../../stylesheet.css" title="Style">
</head>
<body>
<div class="sourceContainer">
<pre><span class="sourceLineNo">001</span>package org.biopax.paxtools.normalizer;<a name="line.1"></a>
<span class="sourceLineNo">002</span><a name="line.2"></a>
<span class="sourceLineNo">003</span>import org.apache.commons.lang.StringUtils;<a name="line.3"></a>
<span class="sourceLineNo">004</span>import org.biopax.paxtools.controller.ModelUtils;<a name="line.4"></a>
<span class="sourceLineNo">005</span>import org.biopax.paxtools.controller.ShallowCopy;<a name="line.5"></a>
<span class="sourceLineNo">006</span>import org.biopax.paxtools.converter.LevelUpgrader;<a name="line.6"></a>
<span class="sourceLineNo">007</span>import org.biopax.paxtools.io.SimpleIOHandler;<a name="line.7"></a>
<span class="sourceLineNo">008</span>import org.biopax.paxtools.model.BioPAXElement;<a name="line.8"></a>
<span class="sourceLineNo">009</span>import org.biopax.paxtools.model.BioPAXLevel;<a name="line.9"></a>
<span class="sourceLineNo">010</span>import org.biopax.paxtools.model.Model;<a name="line.10"></a>
<span class="sourceLineNo">011</span>import org.biopax.paxtools.model.level3.*;<a name="line.11"></a>
<span class="sourceLineNo">012</span>import org.biopax.paxtools.util.BPCollections;<a name="line.12"></a>
<span class="sourceLineNo">013</span>import org.biopax.paxtools.util.ClassFilterSet;<a name="line.13"></a>
<span class="sourceLineNo">014</span>import org.slf4j.Logger;<a name="line.14"></a>
<span class="sourceLineNo">015</span>import org.slf4j.LoggerFactory;<a name="line.15"></a>
<span class="sourceLineNo">016</span><a name="line.16"></a>
<span class="sourceLineNo">017</span>import java.io.ByteArrayInputStream;<a name="line.17"></a>
<span class="sourceLineNo">018</span>import java.io.ByteArrayOutputStream;<a name="line.18"></a>
<span class="sourceLineNo">019</span>import java.io.InputStream;<a name="line.19"></a>
<span class="sourceLineNo">020</span>import java.io.UnsupportedEncodingException;<a name="line.20"></a>
<span class="sourceLineNo">021</span>import java.util.*;<a name="line.21"></a>
<span class="sourceLineNo">022</span><a name="line.22"></a>
<span class="sourceLineNo">023</span>/**<a name="line.23"></a>
<span class="sourceLineNo">024</span> * BioPAX (Level 3) Normalizer, an advanced BioPAX utility <a name="line.24"></a>
<span class="sourceLineNo">025</span> * to help pathway data integrating and linking.<a name="line.25"></a>
<span class="sourceLineNo">026</span> * <a name="line.26"></a>
<span class="sourceLineNo">027</span> * @author rodche<a name="line.27"></a>
<span class="sourceLineNo">028</span> */<a name="line.28"></a>
<span class="sourceLineNo">029</span>public final class Normalizer {<a name="line.29"></a>
<span class="sourceLineNo">030</span>        private static final Logger log = LoggerFactory.getLogger(Normalizer.class);<a name="line.30"></a>
<span class="sourceLineNo">031</span>        <a name="line.31"></a>
<span class="sourceLineNo">032</span>        private SimpleIOHandler biopaxReader;<a name="line.32"></a>
<span class="sourceLineNo">033</span>        private String description = "";<a name="line.33"></a>
<span class="sourceLineNo">034</span>        private boolean fixDisplayName;<a name="line.34"></a>
<span class="sourceLineNo">035</span>        private String xmlBase;<a name="line.35"></a>
<span class="sourceLineNo">036</span>        <a name="line.36"></a>
<span class="sourceLineNo">037</span>        // Normalizer will generate URIs using a strategy specified by the system property<a name="line.37"></a>
<span class="sourceLineNo">038</span>        // (the default is biopax.normalizer.uri.strategy=md5, to generate 32-byte digest hex string for xrefs's uris)<a name="line.38"></a>
<span class="sourceLineNo">039</span>        public static final String PROPERTY_NORMALIZER_URI_STRATEGY = "biopax.normalizer.uri.strategy";<a name="line.39"></a>
<span class="sourceLineNo">040</span>        public static final String VALUE_NORMALIZER_URI_STRATEGY_SIMPLE = "simple";<a name="line.40"></a>
<span class="sourceLineNo">041</span>        public static final String VALUE_NORMALIZER_URI_STRATEGY_MD5 = "md5"; //default strategy<a name="line.41"></a>
<span class="sourceLineNo">042</span>        <a name="line.42"></a>
<span class="sourceLineNo">043</span>        /**<a name="line.43"></a>
<span class="sourceLineNo">044</span>         * Constructor<a name="line.44"></a>
<span class="sourceLineNo">045</span>         */<a name="line.45"></a>
<span class="sourceLineNo">046</span>        public Normalizer() {<a name="line.46"></a>
<span class="sourceLineNo">047</span>                biopaxReader = new SimpleIOHandler(BioPAXLevel.L3);<a name="line.47"></a>
<span class="sourceLineNo">048</span>                biopaxReader.mergeDuplicates(true);<a name="line.48"></a>
<span class="sourceLineNo">049</span>                fixDisplayName = true;<a name="line.49"></a>
<span class="sourceLineNo">050</span>                xmlBase = "";<a name="line.50"></a>
<span class="sourceLineNo">051</span>        }<a name="line.51"></a>
<span class="sourceLineNo">052</span>        <a name="line.52"></a>
<span class="sourceLineNo">053</span>        <a name="line.53"></a>
<span class="sourceLineNo">054</span>        /**<a name="line.54"></a>
<span class="sourceLineNo">055</span>         * Normalizes BioPAX OWL data and returns<a name="line.55"></a>
<span class="sourceLineNo">056</span>         * the result as BioPAX OWL (string).<a name="line.56"></a>
<span class="sourceLineNo">057</span>         * <a name="line.57"></a>
<span class="sourceLineNo">058</span>         * This public method is actually intended to use <a name="line.58"></a>
<span class="sourceLineNo">059</span>         * outside the BioPAX Validator framework, with not too large models.<a name="line.59"></a>
<span class="sourceLineNo">060</span>         * <a name="line.60"></a>
<span class="sourceLineNo">061</span>         * @param biopaxOwlData RDF/XML BioPAX content string<a name="line.61"></a>
<span class="sourceLineNo">062</span>         * @return normalized BioPAX RDF/XML<a name="line.62"></a>
<span class="sourceLineNo">063</span>         * @deprecated this method will fail if the data exceeds ~1Gb (max UTF8 java String length)<a name="line.63"></a>
<span class="sourceLineNo">064</span>         */<a name="line.64"></a>
<span class="sourceLineNo">065</span>        public String normalize(String biopaxOwlData) {<a name="line.65"></a>
<span class="sourceLineNo">066</span>                <a name="line.66"></a>
<span class="sourceLineNo">067</span>                if(biopaxOwlData == null || biopaxOwlData.length() == 0) <a name="line.67"></a>
<span class="sourceLineNo">068</span>                        throw new IllegalArgumentException("no data. " + description);<a name="line.68"></a>
<span class="sourceLineNo">069</span>                <a name="line.69"></a>
<span class="sourceLineNo">070</span>                // quick-fix for older BioPAX L3 version (v0.9x) property 'taxonXref' (range: BioSource)<a name="line.70"></a>
<span class="sourceLineNo">071</span>                biopaxOwlData = biopaxOwlData.replaceAll("taxonXref","xref");<a name="line.71"></a>
<span class="sourceLineNo">072</span>                <a name="line.72"></a>
<span class="sourceLineNo">073</span>                // build the model<a name="line.73"></a>
<span class="sourceLineNo">074</span>                Model model = null;<a name="line.74"></a>
<span class="sourceLineNo">075</span>                try {<a name="line.75"></a>
<span class="sourceLineNo">076</span>                        model = biopaxReader.convertFromOWL(<a name="line.76"></a>
<span class="sourceLineNo">077</span>                                new ByteArrayInputStream(biopaxOwlData.getBytes("UTF-8")));<a name="line.77"></a>
<span class="sourceLineNo">078</span>                } catch (UnsupportedEncodingException e) {<a name="line.78"></a>
<span class="sourceLineNo">079</span>                        throw new IllegalArgumentException("Failed! " + description, e);<a name="line.79"></a>
<span class="sourceLineNo">080</span>                }<a name="line.80"></a>
<span class="sourceLineNo">081</span>                <a name="line.81"></a>
<span class="sourceLineNo">082</span>                if(model == null) {<a name="line.82"></a>
<span class="sourceLineNo">083</span>                        throw new IllegalArgumentException("Failed to create Model! " <a name="line.83"></a>
<span class="sourceLineNo">084</span>                                        + description);<a name="line.84"></a>
<span class="sourceLineNo">085</span>                }<a name="line.85"></a>
<span class="sourceLineNo">086</span>                <a name="line.86"></a>
<span class="sourceLineNo">087</span>                // auto-convert to Level3 model<a name="line.87"></a>
<span class="sourceLineNo">088</span>                if (model.getLevel() != BioPAXLevel.L3) {<a name="line.88"></a>
<span class="sourceLineNo">089</span>                        log.info("Converting model to BioPAX Level3...");<a name="line.89"></a>
<span class="sourceLineNo">090</span>                        model = (new LevelUpgrader()).filter(model);<a name="line.90"></a>
<span class="sourceLineNo">091</span>                }<a name="line.91"></a>
<span class="sourceLineNo">092</span>                <a name="line.92"></a>
<span class="sourceLineNo">093</span>                normalize(model); // L3 only!<a name="line.93"></a>
<span class="sourceLineNo">094</span>                <a name="line.94"></a>
<span class="sourceLineNo">095</span>                // return as BioPAX OWL<a name="line.95"></a>
<span class="sourceLineNo">096</span>                return convertToOWL(model);<a name="line.96"></a>
<span class="sourceLineNo">097</span>        }<a name="line.97"></a>
<span class="sourceLineNo">098</span>        <a name="line.98"></a>
<span class="sourceLineNo">099</span><a name="line.99"></a>
<span class="sourceLineNo">100</span>        /**<a name="line.100"></a>
<span class="sourceLineNo">101</span>         * Normalizes all xrefs (not unification xrefs only) to help normalizing/merging other objects, <a name="line.101"></a>
<span class="sourceLineNo">102</span>         * and also because some of the original xref URIs ("normalized")<a name="line.102"></a>
<span class="sourceLineNo">103</span>         * are in fact to be used for other biopax types (e.g., CV or ProteinReference); therefore<a name="line.103"></a>
<span class="sourceLineNo">104</span>         * this method will replace URI also for "bad" xrefs, i.e., those with empty/illegal 'db' or 'id' values.<a name="line.104"></a>
<span class="sourceLineNo">105</span>         * <a name="line.105"></a>
<span class="sourceLineNo">106</span>         * @param model biopax model to update<a name="line.106"></a>
<span class="sourceLineNo">107</span>         */<a name="line.107"></a>
<span class="sourceLineNo">108</span>        private void normalizeXrefs(Model model) {<a name="line.108"></a>
<span class="sourceLineNo">109</span>                <a name="line.109"></a>
<span class="sourceLineNo">110</span>                final NormalizerMap map = new NormalizerMap(model);<a name="line.110"></a>
<span class="sourceLineNo">111</span>                final String xmlBase = getXmlBase(model); //current base, the default or model's one, if set.<a name="line.111"></a>
<span class="sourceLineNo">112</span>                <a name="line.112"></a>
<span class="sourceLineNo">113</span>                // use a copy of the xrefs set (to avoid concurrent modif. exception)<a name="line.113"></a>
<span class="sourceLineNo">114</span>                Set&lt;? extends Xref&gt; xrefs = new HashSet&lt;Xref&gt;(model.getObjects(Xref.class));<a name="line.114"></a>
<span class="sourceLineNo">115</span>                for(Xref ref : xrefs)<a name="line.115"></a>
<span class="sourceLineNo">116</span>                {<a name="line.116"></a>
<span class="sourceLineNo">117</span>                        //skip not well-defined ones (incl. PublicationXrefs w/o db/id - won't normalize)<a name="line.117"></a>
<span class="sourceLineNo">118</span>                        if(ref.getDb() == null || ref.getId() == null)<a name="line.118"></a>
<span class="sourceLineNo">119</span>                                continue;<a name="line.119"></a>
<span class="sourceLineNo">120</span><a name="line.120"></a>
<span class="sourceLineNo">121</span>                        ref.setDb(ref.getDb().toLowerCase()); //set lowercase<a name="line.121"></a>
<span class="sourceLineNo">122</span>                        String idPart = ref.getId();<a name="line.122"></a>
<span class="sourceLineNo">123</span><a name="line.123"></a>
<span class="sourceLineNo">124</span>                        if(ref instanceof RelationshipXref) {<a name="line.124"></a>
<span class="sourceLineNo">125</span>                                // only normalize (replace URI of) RXs that could potentially clash with other RX, CVs, ERs;<a name="line.125"></a>
<span class="sourceLineNo">126</span>                                // skip, won't bother, for all other RXs...<a name="line.126"></a>
<span class="sourceLineNo">127</span>                                if(!ref.getUri().startsWith("http://identifiers.org/"))<a name="line.127"></a>
<span class="sourceLineNo">128</span>                                        continue;<a name="line.128"></a>
<span class="sourceLineNo">129</span><a name="line.129"></a>
<span class="sourceLineNo">130</span>                                //RXs might have the same db, id but different rel. type.<a name="line.130"></a>
<span class="sourceLineNo">131</span>                                RelationshipTypeVocabulary cv = ((RelationshipXref) ref).getRelationshipType();<a name="line.131"></a>
<span class="sourceLineNo">132</span>                                if(ref.getIdVersion()!=null) {<a name="line.132"></a>
<span class="sourceLineNo">133</span>                                        idPart += "_" + ref.getIdVersion();<a name="line.133"></a>
<span class="sourceLineNo">134</span>                                }<a name="line.134"></a>
<span class="sourceLineNo">135</span>                                if(cv != null &amp;&amp; !cv.getTerm().isEmpty()) {<a name="line.135"></a>
<span class="sourceLineNo">136</span>                                        idPart += "_" + StringUtils.join(cv.getTerm(), '_').toLowerCase();<a name="line.136"></a>
<span class="sourceLineNo">137</span>                                }<a name="line.137"></a>
<span class="sourceLineNo">138</span>                        }<a name="line.138"></a>
<span class="sourceLineNo">139</span>                        else if(ref instanceof UnificationXref) {<a name="line.139"></a>
<span class="sourceLineNo">140</span>                                // first, try to normalize the db name (using MIRIAM EBI registry)<a name="line.140"></a>
<span class="sourceLineNo">141</span>                                try {<a name="line.141"></a>
<span class="sourceLineNo">142</span>                                        ref.setDb(MiriamLink.getName(ref.getDb()).toLowerCase());<a name="line.142"></a>
<span class="sourceLineNo">143</span>                                } catch (IllegalArgumentException e) {<a name="line.143"></a>
<span class="sourceLineNo">144</span>                                        // - unknown/unmatched db name (normalize using defaults)<a name="line.144"></a>
<span class="sourceLineNo">145</span>                                        if(ref.getIdVersion()!=null) {<a name="line.145"></a>
<span class="sourceLineNo">146</span>                                                idPart += "_" + ref.getIdVersion();<a name="line.146"></a>
<span class="sourceLineNo">147</span>                                        }<a name="line.147"></a>
<span class="sourceLineNo">148</span>                                        map.put(ref, Normalizer.uri(xmlBase, ref.getDb(), idPart, ref.getModelInterface()));<a name="line.148"></a>
<span class="sourceLineNo">149</span>                                        continue; //shortcut (non-standard db name)<a name="line.149"></a>
<span class="sourceLineNo">150</span>                                }<a name="line.150"></a>
<span class="sourceLineNo">151</span>                        <a name="line.151"></a>
<span class="sourceLineNo">152</span>                                // a hack for uniprot/isoform xrefs<a name="line.152"></a>
<span class="sourceLineNo">153</span>                                if (ref.getDb().startsWith("uniprot")) {<a name="line.153"></a>
<span class="sourceLineNo">154</span>                                        //auto-fix (guess) for possibly incorrect db/id (can be 'uniprot isoform' with/no idVersion, etc..)<a name="line.154"></a>
<span class="sourceLineNo">155</span>                                        if (isValidDbId("uniprot isoform", ref.getId())<a name="line.155"></a>
<span class="sourceLineNo">156</span>                                                        &amp;&amp; ref.getId().contains("-")) //the second condition is important<a name="line.156"></a>
<span class="sourceLineNo">157</span>                                        {       //then it's certainly an isoform id; so - fix the db name<a name="line.157"></a>
<span class="sourceLineNo">158</span>                                                ref.setDb("uniprot isoform"); //fix the db<a name="line.158"></a>
<span class="sourceLineNo">159</span>                                        }<a name="line.159"></a>
<span class="sourceLineNo">160</span>                                        else {<a name="line.160"></a>
<span class="sourceLineNo">161</span>                                                //id does not end with "-\\d+", i.e., not a isoform id<a name="line.161"></a>
<span class="sourceLineNo">162</span>                                                //(idVersion is a different thing, but id db was "uniprot isoform" they probably misused idVersion)<a name="line.162"></a>
<span class="sourceLineNo">163</span>                                                if(ref.getDb().equals("uniprot isoform"))<a name="line.163"></a>
<span class="sourceLineNo">164</span>                                                {<a name="line.164"></a>
<span class="sourceLineNo">165</span>                                                        if(ref.getIdVersion() != null &amp;&amp; ref.getIdVersion().matches("^\\d+$"))<a name="line.165"></a>
<span class="sourceLineNo">166</span>                                                                idPart = ref.getId()+"-"+ref.getIdVersion(); //guess, by idVersion, they actually meant isoform<a name="line.166"></a>
<span class="sourceLineNo">167</span>                                                        if(isValidDbId(ref.getDb(), idPart)) {<a name="line.167"></a>
<span class="sourceLineNo">168</span>                                                                ref.setId(idPart); //moved the isoform # to the ID<a name="line.168"></a>
<span class="sourceLineNo">169</span>                                                                ref.setIdVersion(null);<a name="line.169"></a>
<span class="sourceLineNo">170</span>                                                        }<a name="line.170"></a>
<span class="sourceLineNo">171</span>                                                        else if(!isValidDbId(ref.getDb(), ref.getId())) {<a name="line.171"></a>
<span class="sourceLineNo">172</span>                                                                //certainly not isoform (might not even uniprot, but try...)<a name="line.172"></a>
<span class="sourceLineNo">173</span>                                                                ref.setDb("uniprot knowledgebase"); //guess, fix<a name="line.173"></a>
<span class="sourceLineNo">174</span>                                                        }<a name="line.174"></a>
<span class="sourceLineNo">175</span>                                                        idPart = ref.getId();<a name="line.175"></a>
<span class="sourceLineNo">176</span>                                                }<a name="line.176"></a>
<span class="sourceLineNo">177</span>                                        }<a name="line.177"></a>
<span class="sourceLineNo">178</span>                                }<a name="line.178"></a>
<span class="sourceLineNo">179</span>                        }<a name="line.179"></a>
<span class="sourceLineNo">180</span><a name="line.180"></a>
<span class="sourceLineNo">181</span>                        // shelve it for URI replace<a name="line.181"></a>
<span class="sourceLineNo">182</span>                        map.put(ref, Normalizer.uri(xmlBase, ref.getDb(), idPart, ref.getModelInterface()));<a name="line.182"></a>
<span class="sourceLineNo">183</span>                }<a name="line.183"></a>
<span class="sourceLineNo">184</span>                <a name="line.184"></a>
<span class="sourceLineNo">185</span>                // execute replace xrefs<a name="line.185"></a>
<span class="sourceLineNo">186</span>                map.doSubs();<a name="line.186"></a>
<span class="sourceLineNo">187</span>        }<a name="line.187"></a>
<span class="sourceLineNo">188</span><a name="line.188"></a>
<span class="sourceLineNo">189</span>        /*<a name="line.189"></a>
<span class="sourceLineNo">190</span>         * @param db must be valid MIRIAM db name or sinonym<a name="line.190"></a>
<span class="sourceLineNo">191</span>         * @trows IllegalArgumentException when db is unknown name.<a name="line.191"></a>
<span class="sourceLineNo">192</span>         */<a name="line.192"></a>
<span class="sourceLineNo">193</span>        private boolean isValidDbId(String db, String id) {<a name="line.193"></a>
<span class="sourceLineNo">194</span>                return MiriamLink.checkRegExp(id, db);<a name="line.194"></a>
<span class="sourceLineNo">195</span>        }<a name="line.195"></a>
<span class="sourceLineNo">196</span><a name="line.196"></a>
<span class="sourceLineNo">197</span>        /**<a name="line.197"></a>
<span class="sourceLineNo">198</span>         * Consistently generates a new BioPAX element URI <a name="line.198"></a>
<span class="sourceLineNo">199</span>         * using given URI namespace (xml:base), BioPAX class, <a name="line.199"></a>
<span class="sourceLineNo">200</span>         * and two different identifiers (at least one is required).<a name="line.200"></a>
<span class="sourceLineNo">201</span>         * Miriam registry is used to get the standard db name and <a name="line.201"></a>
<span class="sourceLineNo">202</span>         * identifiers.org URI, if possible, only for relationship type vocabulary, <a name="line.202"></a>
<span class="sourceLineNo">203</span>         * publication xref, and entity reference types.<a name="line.203"></a>
<span class="sourceLineNo">204</span>         * <a name="line.204"></a>
<span class="sourceLineNo">205</span>         * @param xmlBase xml:base (common URI prefix for a BioPAX model), case-sensitive<a name="line.205"></a>
<span class="sourceLineNo">206</span>         * @param dbName a bio data collection name or synonym, case-insensitive<a name="line.206"></a>
<span class="sourceLineNo">207</span>         * @param idPart optional (can be null), e.g., xref.id, case-sensitive<a name="line.207"></a>
<span class="sourceLineNo">208</span>         * @param type BioPAX class<a name="line.208"></a>
<span class="sourceLineNo">209</span>         * @return URI<a name="line.209"></a>
<span class="sourceLineNo">210</span>         * @throws IllegalArgumentException if either type is null or both 'dbName' and 'idPart' are all nulls.<a name="line.210"></a>
<span class="sourceLineNo">211</span>         */<a name="line.211"></a>
<span class="sourceLineNo">212</span>        public static String uri(final String xmlBase, <a name="line.212"></a>
<span class="sourceLineNo">213</span>                String dbName, final String idPart, Class&lt;? extends BioPAXElement&gt; type)<a name="line.213"></a>
<span class="sourceLineNo">214</span>        {<a name="line.214"></a>
<span class="sourceLineNo">215</span>                if(type == null || (dbName == null &amp;&amp; idPart == null))<a name="line.215"></a>
<span class="sourceLineNo">216</span>                        throw new IllegalArgumentException("'Either type' is null, or both dbName and idPart are nulls.");<a name="line.216"></a>
<span class="sourceLineNo">217</span>                        <a name="line.217"></a>
<span class="sourceLineNo">218</span>                // try to find a standard URI, if exists, for a publication xref, <a name="line.218"></a>
<span class="sourceLineNo">219</span>                // or at least a standard name:<a name="line.219"></a>
<span class="sourceLineNo">220</span>                if (dbName != null) {<a name="line.220"></a>
<span class="sourceLineNo">221</span>                        try {<a name="line.221"></a>
<span class="sourceLineNo">222</span>                                // try to get the preferred/standard name<a name="line.222"></a>
<span class="sourceLineNo">223</span>                                // for any type, for consistency<a name="line.223"></a>
<span class="sourceLineNo">224</span>                                dbName = MiriamLink.getName(dbName);<a name="line.224"></a>
<span class="sourceLineNo">225</span>                                // a shortcut: a standard and resolvable URI exists for some BioPAX types<a name="line.225"></a>
<span class="sourceLineNo">226</span>                                if ((type.equals(PublicationXref.class) &amp;&amp; "pubmed".equalsIgnoreCase(dbName))<a name="line.226"></a>
<span class="sourceLineNo">227</span>                                        || type.equals(RelationshipTypeVocabulary.class)<a name="line.227"></a>
<span class="sourceLineNo">228</span>                                        || ProteinReference.class.isAssignableFrom(type)<a name="line.228"></a>
<span class="sourceLineNo">229</span>                                        || SmallMoleculeReference.class.isAssignableFrom(type)<a name="line.229"></a>
<span class="sourceLineNo">230</span>                                        || (type.equals(BioSource.class) &amp;&amp; "taxonomy".equalsIgnoreCase(dbName) &amp;&amp; idPart.matches("^\\d+$"))<a name="line.230"></a>
<span class="sourceLineNo">231</span>                                )<a name="line.231"></a>
<span class="sourceLineNo">232</span>                                {       //get the standard URI and quit (success), or fail and continue making a new URI below...<a name="line.232"></a>
<span class="sourceLineNo">233</span>                                        return MiriamLink.getIdentifiersOrgURI(dbName, idPart);<a name="line.233"></a>
<span class="sourceLineNo">234</span>                                }<a name="line.234"></a>
<span class="sourceLineNo">235</span>                        } catch (IllegalArgumentException e) {<a name="line.235"></a>
<span class="sourceLineNo">236</span>                                log.info(String.format("uri(for a %s): db:%s, id:%s are not standard; %s)",<a name="line.236"></a>
<span class="sourceLineNo">237</span>                                                type.getSimpleName(), dbName, idPart, e.getMessage()));<a name="line.237"></a>
<span class="sourceLineNo">238</span>                        }<a name="line.238"></a>
<span class="sourceLineNo">239</span>                }<a name="line.239"></a>
<span class="sourceLineNo">240</span><a name="line.240"></a>
<span class="sourceLineNo">241</span>                // If not returned above this point - no standard URI (Identifiers.org) was found -<a name="line.241"></a>
<span class="sourceLineNo">242</span>                // then let's consistently build a new URI from args, anyway, the other way around:<a name="line.242"></a>
<span class="sourceLineNo">243</span>                <a name="line.243"></a>
<span class="sourceLineNo">244</span>                StringBuilder sb = new StringBuilder();         <a name="line.244"></a>
<span class="sourceLineNo">245</span>                if (dbName != null) //lowercase for consistency<a name="line.245"></a>
<span class="sourceLineNo">246</span>                        sb.append(dbName.toLowerCase());        <a name="line.246"></a>
<span class="sourceLineNo">247</span>                <a name="line.247"></a>
<span class="sourceLineNo">248</span>                if (idPart != null) {<a name="line.248"></a>
<span class="sourceLineNo">249</span>                        if (dbName != null) sb.append("_");<a name="line.249"></a>
<span class="sourceLineNo">250</span>                        sb.append(idPart);<a name="line.250"></a>
<span class="sourceLineNo">251</span>                }<a name="line.251"></a>
<span class="sourceLineNo">252</span>                <a name="line.252"></a>
<span class="sourceLineNo">253</span>                String localPart = sb.toString();<a name="line.253"></a>
<span class="sourceLineNo">254</span>                String strategy = System.getProperty(PROPERTY_NORMALIZER_URI_STRATEGY, VALUE_NORMALIZER_URI_STRATEGY_MD5);<a name="line.254"></a>
<span class="sourceLineNo">255</span>                if(VALUE_NORMALIZER_URI_STRATEGY_SIMPLE.equals(strategy) || Xref.class.isAssignableFrom(type))<a name="line.255"></a>
<span class="sourceLineNo">256</span>                //i.e., for xrefs, always use the simple URI strategy (makes them human-readable)<a name="line.256"></a>
<span class="sourceLineNo">257</span>                {<a name="line.257"></a>
<span class="sourceLineNo">258</span>                        //simply replace "unsafe" symbols with underscore (some uri clashes might be possible but rare...)<a name="line.258"></a>
<span class="sourceLineNo">259</span>                        localPart = localPart.replaceAll("[^-\\w]", "_");<a name="line.259"></a>
<span class="sourceLineNo">260</span>                }<a name="line.260"></a>
<span class="sourceLineNo">261</span>                else<a name="line.261"></a>
<span class="sourceLineNo">262</span>                {<a name="line.262"></a>
<span class="sourceLineNo">263</span>                        //replace the local part with its md5 sum string (32-byte)<a name="line.263"></a>
<span class="sourceLineNo">264</span>                        localPart = ModelUtils.md5hex(localPart);<a name="line.264"></a>
<span class="sourceLineNo">265</span>                }<a name="line.265"></a>
<span class="sourceLineNo">266</span>                <a name="line.266"></a>
<span class="sourceLineNo">267</span>                // create URI using the xml:base and digest of other values:<a name="line.267"></a>
<span class="sourceLineNo">268</span>                return ((xmlBase != null) ? xmlBase : "") + type.getSimpleName() + "_" + localPart;<a name="line.268"></a>
<span class="sourceLineNo">269</span>        }<a name="line.269"></a>
<span class="sourceLineNo">270</span><a name="line.270"></a>
<span class="sourceLineNo">271</span>        <a name="line.271"></a>
<span class="sourceLineNo">272</span>        /**<a name="line.272"></a>
<span class="sourceLineNo">273</span>         * Description of the model to normalize.<a name="line.273"></a>
<span class="sourceLineNo">274</span>         * <a name="line.274"></a>
<span class="sourceLineNo">275</span>         * @return info about the BioPAX model<a name="line.275"></a>
<span class="sourceLineNo">276</span>         */<a name="line.276"></a>
<span class="sourceLineNo">277</span>        public String getDescription() {<a name="line.277"></a>
<span class="sourceLineNo">278</span>                return description;<a name="line.278"></a>
<span class="sourceLineNo">279</span>        }<a name="line.279"></a>
<span class="sourceLineNo">280</span>        <a name="line.280"></a>
<span class="sourceLineNo">281</span>        <a name="line.281"></a>
<span class="sourceLineNo">282</span>        /**<a name="line.282"></a>
<span class="sourceLineNo">283</span>         * @param description the description to set<a name="line.283"></a>
<span class="sourceLineNo">284</span>         */<a name="line.284"></a>
<span class="sourceLineNo">285</span>        public void setDescription(String description) {<a name="line.285"></a>
<span class="sourceLineNo">286</span>                this.description = description;<a name="line.286"></a>
<span class="sourceLineNo">287</span>        }<a name="line.287"></a>
<span class="sourceLineNo">288</span><a name="line.288"></a>
<span class="sourceLineNo">289</span>        <a name="line.289"></a>
<span class="sourceLineNo">290</span>        private void fixDisplayName(Model model) {<a name="line.290"></a>
<span class="sourceLineNo">291</span>                log.info("Trying to auto-fix 'null' displayName...");<a name="line.291"></a>
<span class="sourceLineNo">292</span>                // where it's null, set to the shortest name if possible<a name="line.292"></a>
<span class="sourceLineNo">293</span>                for (Named e : model.getObjects(Named.class)) {<a name="line.293"></a>
<span class="sourceLineNo">294</span>                        if (e.getDisplayName() == null) {<a name="line.294"></a>
<span class="sourceLineNo">295</span>                                if (e.getStandardName() != null) {<a name="line.295"></a>
<span class="sourceLineNo">296</span>                                        e.setDisplayName(e.getStandardName());<a name="line.296"></a>
<span class="sourceLineNo">297</span>                                        log.info(e + " displayName auto-fix: "<a name="line.297"></a>
<span class="sourceLineNo">298</span>                                                        + e.getDisplayName() + ". " + description);<a name="line.298"></a>
<span class="sourceLineNo">299</span>                                } else if (!e.getName().isEmpty()) {<a name="line.299"></a>
<span class="sourceLineNo">300</span>                                        String dsp = e.getName().iterator().next();<a name="line.300"></a>
<span class="sourceLineNo">301</span>                                        for (String name : e.getName()) {<a name="line.301"></a>
<span class="sourceLineNo">302</span>                                                if (name.length() &lt; dsp.length())<a name="line.302"></a>
<span class="sourceLineNo">303</span>                                                        dsp = name;<a name="line.303"></a>
<span class="sourceLineNo">304</span>                                        }<a name="line.304"></a>
<span class="sourceLineNo">305</span>                                        e.setDisplayName(dsp);<a name="line.305"></a>
<span class="sourceLineNo">306</span>                                        log.info(e + " displayName auto-fix: " + dsp<a name="line.306"></a>
<span class="sourceLineNo">307</span>                                                + ". " + description);<a name="line.307"></a>
<span class="sourceLineNo">308</span>                                }<a name="line.308"></a>
<span class="sourceLineNo">309</span>                        }<a name="line.309"></a>
<span class="sourceLineNo">310</span>                }<a name="line.310"></a>
<span class="sourceLineNo">311</span>                // if required, set PE name to (already fixed) ER's name...<a name="line.311"></a>
<span class="sourceLineNo">312</span>                for(EntityReference er : model.getObjects(EntityReference.class)) {<a name="line.312"></a>
<span class="sourceLineNo">313</span>                        for(SimplePhysicalEntity spe : er.getEntityReferenceOf()) {<a name="line.313"></a>
<span class="sourceLineNo">314</span>                                if(spe.getDisplayName() == null || spe.getDisplayName().trim().length() == 0) {<a name="line.314"></a>
<span class="sourceLineNo">315</span>                                        if(er.getDisplayName() != null &amp;&amp; er.getDisplayName().trim().length() &gt; 0) {<a name="line.315"></a>
<span class="sourceLineNo">316</span>                                                spe.setDisplayName(er.getDisplayName());<a name="line.316"></a>
<span class="sourceLineNo">317</span>                                        }<a name="line.317"></a>
<span class="sourceLineNo">318</span>                                }<a name="line.318"></a>
<span class="sourceLineNo">319</span>                        }<a name="line.319"></a>
<span class="sourceLineNo">320</span>                }<a name="line.320"></a>
<span class="sourceLineNo">321</span>        }<a name="line.321"></a>
<span class="sourceLineNo">322</span><a name="line.322"></a>
<span class="sourceLineNo">323</span><a name="line.323"></a>
<span class="sourceLineNo">324</span>        private String convertToOWL(Model model) {<a name="line.324"></a>
<span class="sourceLineNo">325</span>                ByteArrayOutputStream out = new ByteArrayOutputStream();<a name="line.325"></a>
<span class="sourceLineNo">326</span>                (new SimpleIOHandler(model.getLevel())).convertToOWL(model, out);<a name="line.326"></a>
<span class="sourceLineNo">327</span>                return out.toString();<a name="line.327"></a>
<span class="sourceLineNo">328</span>        }<a name="line.328"></a>
<span class="sourceLineNo">329</span><a name="line.329"></a>
<span class="sourceLineNo">330</span><a name="line.330"></a>
<span class="sourceLineNo">331</span>        private Collection&lt;UnificationXref&gt; getUnificationXrefsSorted(XReferrable r) {<a name="line.331"></a>
<span class="sourceLineNo">332</span><a name="line.332"></a>
<span class="sourceLineNo">333</span>                Collection&lt;UnificationXref&gt; urefs = new TreeSet&lt;UnificationXref&gt;(<a name="line.333"></a>
<span class="sourceLineNo">334</span>                                new Comparator&lt;UnificationXref&gt;() {<a name="line.334"></a>
<span class="sourceLineNo">335</span>                                        public int compare(UnificationXref o1, UnificationXref o2) {<a name="line.335"></a>
<span class="sourceLineNo">336</span>                                                String s1 = o1.getDb() + o1.getId();<a name="line.336"></a>
<span class="sourceLineNo">337</span>                                                String s2 = o2.getDb() + o2.getId();<a name="line.337"></a>
<span class="sourceLineNo">338</span>                                                return s1.compareTo(s2);<a name="line.338"></a>
<span class="sourceLineNo">339</span>                                        }<a name="line.339"></a>
<span class="sourceLineNo">340</span>                                }<a name="line.340"></a>
<span class="sourceLineNo">341</span>                );<a name="line.341"></a>
<span class="sourceLineNo">342</span><a name="line.342"></a>
<span class="sourceLineNo">343</span>                for(UnificationXref ux : new ClassFilterSet&lt;Xref,UnificationXref&gt;(r.getXref(), UnificationXref.class))<a name="line.343"></a>
<span class="sourceLineNo">344</span>                {<a name="line.344"></a>
<span class="sourceLineNo">345</span>                        if(ux.getDb() != null &amp;&amp; ux.getId() != null) {<a name="line.345"></a>
<span class="sourceLineNo">346</span>                                urefs.add(ux);<a name="line.346"></a>
<span class="sourceLineNo">347</span>                        } <a name="line.347"></a>
<span class="sourceLineNo">348</span>                }<a name="line.348"></a>
<span class="sourceLineNo">349</span>                <a name="line.349"></a>
<span class="sourceLineNo">350</span>                return urefs;<a name="line.350"></a>
<span class="sourceLineNo">351</span>        }<a name="line.351"></a>
<span class="sourceLineNo">352</span><a name="line.352"></a>
<span class="sourceLineNo">353</span>        <a name="line.353"></a>
<span class="sourceLineNo">354</span>        /**<a name="line.354"></a>
<span class="sourceLineNo">355</span>         * Finds one preferred unification xref, if possible.<a name="line.355"></a>
<span class="sourceLineNo">356</span>         * Preferred db values are:<a name="line.356"></a>
<span class="sourceLineNo">357</span>         * "ncbi gene" - for NucleicAcidReference and the sub-classes;<a name="line.357"></a>
<span class="sourceLineNo">358</span>         * "uniprot" or "refseq" for ProteinReference;<a name="line.358"></a>
<span class="sourceLineNo">359</span>         * "chebi" - for SmallMoleculeReference;<a name="line.359"></a>
<span class="sourceLineNo">360</span>         * <a name="line.360"></a>
<span class="sourceLineNo">361</span>         * @param bpe BioPAX object that can have xrefs<a name="line.361"></a>
<span class="sourceLineNo">362</span>         * @return the "best" first unification xref <a name="line.362"></a>
<span class="sourceLineNo">363</span>         */<a name="line.363"></a>
<span class="sourceLineNo">364</span>        private UnificationXref findPreferredUnificationXref(XReferrable bpe)<a name="line.364"></a>
<span class="sourceLineNo">365</span>        {<a name="line.365"></a>
<span class="sourceLineNo">366</span>                UnificationXref toReturn = null;<a name="line.366"></a>
<span class="sourceLineNo">367</span><a name="line.367"></a>
<span class="sourceLineNo">368</span>                Collection&lt;UnificationXref&gt; orderedUrefs = getUnificationXrefsSorted(bpe);<a name="line.368"></a>
<span class="sourceLineNo">369</span><a name="line.369"></a>
<span class="sourceLineNo">370</span>                //use preferred db prefix for different type of ER<a name="line.370"></a>
<span class="sourceLineNo">371</span>                if(bpe instanceof ProteinReference) {<a name="line.371"></a>
<span class="sourceLineNo">372</span>                        // prefer xref having db starts with "uniprot" (preferred) or "refseq";<a name="line.372"></a>
<span class="sourceLineNo">373</span>                        toReturn = findSingleUnificationXref(orderedUrefs, "uniprot");<a name="line.373"></a>
<span class="sourceLineNo">374</span>                        // if null, next try refseq.<a name="line.374"></a>
<span class="sourceLineNo">375</span>                        if(toReturn==null)<a name="line.375"></a>
<span class="sourceLineNo">376</span>                                toReturn = findSingleUnificationXref(orderedUrefs, "refseq");<a name="line.376"></a>
<span class="sourceLineNo">377</span>                } else if(bpe instanceof SmallMoleculeReference) {<a name="line.377"></a>
<span class="sourceLineNo">378</span>                        // - "chebi", then  "pubchem";<a name="line.378"></a>
<span class="sourceLineNo">379</span>                        toReturn = findSingleUnificationXref(orderedUrefs, "chebi");<a name="line.379"></a>
<span class="sourceLineNo">380</span>                        if(toReturn==null)<a name="line.380"></a>
<span class="sourceLineNo">381</span>                                toReturn = findSingleUnificationXref(orderedUrefs, "pubchem");<a name="line.381"></a>
<span class="sourceLineNo">382</span>                } else if(bpe instanceof NucleicAcidReference) {<a name="line.382"></a>
<span class="sourceLineNo">383</span>                        //that includes NucleicAcidRegionReference, etc. sub-classes;<a name="line.383"></a>
<span class="sourceLineNo">384</span>                        toReturn = findSingleUnificationXref(orderedUrefs, "ncbi gene");<a name="line.384"></a>
<span class="sourceLineNo">385</span>                        if(toReturn==null)<a name="line.385"></a>
<span class="sourceLineNo">386</span>                                toReturn = findSingleUnificationXref(orderedUrefs, "entrez");<a name="line.386"></a>
<span class="sourceLineNo">387</span>                } else {<a name="line.387"></a>
<span class="sourceLineNo">388</span>                        //for other XReferrable types (BioSource or ControlledVocabulary)<a name="line.388"></a>
<span class="sourceLineNo">389</span>                        //use if there's only one xref (return null if many)<a name="line.389"></a>
<span class="sourceLineNo">390</span>                        if(orderedUrefs.size()==1)<a name="line.390"></a>
<span class="sourceLineNo">391</span>                                toReturn = orderedUrefs.iterator().next();<a name="line.391"></a>
<span class="sourceLineNo">392</span>                }<a name="line.392"></a>
<span class="sourceLineNo">393</span>                <a name="line.393"></a>
<span class="sourceLineNo">394</span>                return toReturn;<a name="line.394"></a>
<span class="sourceLineNo">395</span>        }<a name="line.395"></a>
<span class="sourceLineNo">396</span><a name="line.396"></a>
<span class="sourceLineNo">397</span>        private UnificationXref findSingleUnificationXref(Collection&lt;UnificationXref&gt; uXrefs, String dbStartsWith) {<a name="line.397"></a>
<span class="sourceLineNo">398</span>                UnificationXref ret = null;<a name="line.398"></a>
<span class="sourceLineNo">399</span><a name="line.399"></a>
<span class="sourceLineNo">400</span>                for(Xref x : uXrefs) {<a name="line.400"></a>
<span class="sourceLineNo">401</span>                        if(x instanceof UnificationXref &amp;&amp; x.getId() != null &amp;&amp; x.getDb() != null<a name="line.401"></a>
<span class="sourceLineNo">402</span>                                        &amp;&amp; x.getDb().toLowerCase().startsWith(dbStartsWith))<a name="line.402"></a>
<span class="sourceLineNo">403</span>                        {<a name="line.403"></a>
<span class="sourceLineNo">404</span>                                if(ret == null)<a name="line.404"></a>
<span class="sourceLineNo">405</span>                                {<a name="line.405"></a>
<span class="sourceLineNo">406</span>                                        ret = (UnificationXref)x;<a name="line.406"></a>
<span class="sourceLineNo">407</span>                                }<a name="line.407"></a>
<span class="sourceLineNo">408</span>                                else if (ret.getDb().equalsIgnoreCase(x.getDb()) &amp;&amp; !ret.getId().equals(x.getId())) {<a name="line.408"></a>
<span class="sourceLineNo">409</span>                                        //exactly the same kind (same 'db') xref is found;<a name="line.409"></a>
<span class="sourceLineNo">410</span>                                        //we can't test here if several IDs map to the same thing;<a name="line.410"></a>
<span class="sourceLineNo">411</span>                                        //so, give up - for safety - and return null.<a name="line.411"></a>
<span class="sourceLineNo">412</span>                                        ret = null;<a name="line.412"></a>
<span class="sourceLineNo">413</span>                                        break;<a name="line.413"></a>
<span class="sourceLineNo">414</span>                                }<a name="line.414"></a>
<span class="sourceLineNo">415</span>                        }<a name="line.415"></a>
<span class="sourceLineNo">416</span>                }<a name="line.416"></a>
<span class="sourceLineNo">417</span><a name="line.417"></a>
<span class="sourceLineNo">418</span>                return ret;<a name="line.418"></a>
<span class="sourceLineNo">419</span>        }<a name="line.419"></a>
<span class="sourceLineNo">420</span><a name="line.420"></a>
<span class="sourceLineNo">421</span>        <a name="line.421"></a>
<span class="sourceLineNo">422</span>        /**<a name="line.422"></a>
<span class="sourceLineNo">423</span>         * BioPAX normalization <a name="line.423"></a>
<span class="sourceLineNo">424</span>         * (modifies the original Model)<a name="line.424"></a>
<span class="sourceLineNo">425</span>         * <a name="line.425"></a>
<span class="sourceLineNo">426</span>         * @param model BioPAX model to normalize<a name="line.426"></a>
<span class="sourceLineNo">427</span>         * @throws NullPointerException if model is null<a name="line.427"></a>
<span class="sourceLineNo">428</span>         * @throws IllegalArgumentException if model is not Level3 BioPAX<a name="line.428"></a>
<span class="sourceLineNo">429</span>         */<a name="line.429"></a>
<span class="sourceLineNo">430</span>        public void normalize(Model model) {<a name="line.430"></a>
<span class="sourceLineNo">431</span>                <a name="line.431"></a>
<span class="sourceLineNo">432</span>                if(model.getLevel() != BioPAXLevel.L3)<a name="line.432"></a>
<span class="sourceLineNo">433</span>                        throw new IllegalArgumentException("Not Level3 model. " +<a name="line.433"></a>
<span class="sourceLineNo">434</span>                                "Consider converting it first (e.g., with the PaxTools).");<a name="line.434"></a>
<span class="sourceLineNo">435</span>                <a name="line.435"></a>
<span class="sourceLineNo">436</span>                //if set, update the xml:base<a name="line.436"></a>
<span class="sourceLineNo">437</span>                if(xmlBase != null &amp;&amp; !xmlBase.isEmpty())<a name="line.437"></a>
<span class="sourceLineNo">438</span>                        model.setXmlBase(xmlBase);<a name="line.438"></a>
<span class="sourceLineNo">439</span><a name="line.439"></a>
<span class="sourceLineNo">440</span>                // Normalize/merge xrefs, first, and then CVs<a name="line.440"></a>
<span class="sourceLineNo">441</span>                // (also because some of original xrefs might have "normalized" URIs <a name="line.441"></a>
<span class="sourceLineNo">442</span>                // that, in fact, must be used for other biopax types, such as CV or ProteinReference)<a name="line.442"></a>
<span class="sourceLineNo">443</span>                log.info("Normalizing xrefs..." + description);<a name="line.443"></a>
<span class="sourceLineNo">444</span>                normalizeXrefs(model);<a name="line.444"></a>
<span class="sourceLineNo">445</span>                <a name="line.445"></a>
<span class="sourceLineNo">446</span>                // fix displayName where possible<a name="line.446"></a>
<span class="sourceLineNo">447</span>                if(fixDisplayName) {<a name="line.447"></a>
<span class="sourceLineNo">448</span>                        log.info("Normalizing display names..." + description);<a name="line.448"></a>
<span class="sourceLineNo">449</span>                        fixDisplayName(model);<a name="line.449"></a>
<span class="sourceLineNo">450</span>                }<a name="line.450"></a>
<span class="sourceLineNo">451</span>                        <a name="line.451"></a>
<span class="sourceLineNo">452</span>                log.info("Normalizing CVs..." + description);<a name="line.452"></a>
<span class="sourceLineNo">453</span>                normalizeCVs(model);<a name="line.453"></a>
<span class="sourceLineNo">454</span>                <a name="line.454"></a>
<span class="sourceLineNo">455</span>                //normalize BioSource objects (better, as it is here, go after Xrefs and CVs)<a name="line.455"></a>
<span class="sourceLineNo">456</span>                log.info("Normalizing organisms..." + description);<a name="line.456"></a>
<span class="sourceLineNo">457</span>                normalizeBioSources(model);<a name="line.457"></a>
<span class="sourceLineNo">458</span><a name="line.458"></a>
<span class="sourceLineNo">459</span>                // auto-generate missing entity references:<a name="line.459"></a>
<span class="sourceLineNo">460</span>                for(SimplePhysicalEntity spe : new HashSet&lt;SimplePhysicalEntity&gt;(model.getObjects(SimplePhysicalEntity.class))) {<a name="line.460"></a>
<span class="sourceLineNo">461</span>                        //it skips if spe has entityReference or memberPE already<a name="line.461"></a>
<span class="sourceLineNo">462</span>                        ModelUtils.addMissingEntityReference(model, spe);<a name="line.462"></a>
<span class="sourceLineNo">463</span>                }<a name="line.463"></a>
<span class="sourceLineNo">464</span><a name="line.464"></a>
<span class="sourceLineNo">465</span>                log.info("Normalizing entity references..." + description);<a name="line.465"></a>
<span class="sourceLineNo">466</span>                normalizeERs(model);<a name="line.466"></a>
<span class="sourceLineNo">467</span>                <a name="line.467"></a>
<span class="sourceLineNo">468</span>                // find/add lost (in replace) children<a name="line.468"></a>
<span class="sourceLineNo">469</span>                log.info("Repairing..." + description);<a name="line.469"></a>
<span class="sourceLineNo">470</span>                model.repair(); // it does not remove dangling utility class objects (can be done separately, later, if needed)<a name="line.470"></a>
<span class="sourceLineNo">471</span>                <a name="line.471"></a>
<span class="sourceLineNo">472</span>                log.info("Optional tasks (reasoning)..." + description);<a name="line.472"></a>
<span class="sourceLineNo">473</span>        }<a name="line.473"></a>
<span class="sourceLineNo">474</span><a name="line.474"></a>
<span class="sourceLineNo">475</span>        <a name="line.475"></a>
<span class="sourceLineNo">476</span>        private void normalizeCVs(Model model) {<a name="line.476"></a>
<span class="sourceLineNo">477</span>                <a name="line.477"></a>
<span class="sourceLineNo">478</span>                NormalizerMap map = new NormalizerMap(model);<a name="line.478"></a>
<span class="sourceLineNo">479</span>                <a name="line.479"></a>
<span class="sourceLineNo">480</span>                // process ControlledVocabulary objects (all sub-classes)<a name="line.480"></a>
<span class="sourceLineNo">481</span>                for(ControlledVocabulary cv : model.getObjects(ControlledVocabulary.class))<a name="line.481"></a>
<span class="sourceLineNo">482</span>                {<a name="line.482"></a>
<span class="sourceLineNo">483</span>                        //it does not check/fix the CV terms though (but a validation rule can do if run before the normalizer)...<a name="line.483"></a>
<span class="sourceLineNo">484</span>                        UnificationXref uref = findPreferredUnificationXref(cv); //usually, there's only one such xref<a name="line.484"></a>
<span class="sourceLineNo">485</span>                        if (uref != null) {<a name="line.485"></a>
<span class="sourceLineNo">486</span>                                // so let's generate a consistent URI<a name="line.486"></a>
<span class="sourceLineNo">487</span>                                map.put(cv, uri(xmlBase, uref.getDb(), uref.getId(), cv.getModelInterface()));<a name="line.487"></a>
<span class="sourceLineNo">488</span>                        } else if(!cv.getTerm().isEmpty()) {<a name="line.488"></a>
<span class="sourceLineNo">489</span>                                map.put(cv, uri(xmlBase, null, cv.getTerm().iterator().next(), cv.getModelInterface()));<a name="line.489"></a>
<span class="sourceLineNo">490</span>                        } else log.info("Cannot normalize " + cv.getModelInterface().getSimpleName() <a name="line.490"></a>
<span class="sourceLineNo">491</span>                                + " : no unification xrefs nor terms found in " + cv.getUri()<a name="line.491"></a>
<span class="sourceLineNo">492</span>                                + ". " + description);<a name="line.492"></a>
<span class="sourceLineNo">493</span>                } <a name="line.493"></a>
<span class="sourceLineNo">494</span>                <a name="line.494"></a>
<span class="sourceLineNo">495</span>                // replace/update elements in the model<a name="line.495"></a>
<span class="sourceLineNo">496</span>                map.doSubs();<a name="line.496"></a>
<span class="sourceLineNo">497</span>        }<a name="line.497"></a>
<span class="sourceLineNo">498</span>        <a name="line.498"></a>
<span class="sourceLineNo">499</span>        <a name="line.499"></a>
<span class="sourceLineNo">500</span>        private void normalizeBioSources(Model model) {<a name="line.500"></a>
<span class="sourceLineNo">501</span>                <a name="line.501"></a>
<span class="sourceLineNo">502</span>                NormalizerMap map = new NormalizerMap(model);<a name="line.502"></a>
<span class="sourceLineNo">503</span>                <a name="line.503"></a>
<span class="sourceLineNo">504</span>                for(BioSource bs : model.getObjects(BioSource.class))<a name="line.504"></a>
<span class="sourceLineNo">505</span>                {<a name="line.505"></a>
<span class="sourceLineNo">506</span>                        UnificationXref uref = findPreferredUnificationXref(bs);<a name="line.506"></a>
<span class="sourceLineNo">507</span>                        //normally, the xref db is 'Taxonomy' (or a valid synonym)<a name="line.507"></a>
<span class="sourceLineNo">508</span>                        if (uref != null<a name="line.508"></a>
<span class="sourceLineNo">509</span>                                &amp;&amp; (uref.getDb().toLowerCase().contains("taxonomy") || uref.getDb().equalsIgnoreCase("newt")))<a name="line.509"></a>
<span class="sourceLineNo">510</span>                        {       <a name="line.510"></a>
<span class="sourceLineNo">511</span>                                String  idPart = uref.getId();<a name="line.511"></a>
<span class="sourceLineNo">512</span><a name="line.512"></a>
<span class="sourceLineNo">513</span>                                //tissue/cellType terms can be added below:<a name="line.513"></a>
<span class="sourceLineNo">514</span>                                if(bs.getTissue()!=null &amp;&amp; !bs.getTissue().getTerm().isEmpty()) <a name="line.514"></a>
<span class="sourceLineNo">515</span>                                        idPart += "_" + bs.getTissue().getTerm().iterator().next();<a name="line.515"></a>
<span class="sourceLineNo">516</span>                                if(bs.getCellType()!=null &amp;&amp; !bs.getCellType().getTerm().isEmpty()) <a name="line.516"></a>
<span class="sourceLineNo">517</span>                                        idPart += "_" + bs.getCellType().getTerm().iterator().next();<a name="line.517"></a>
<span class="sourceLineNo">518</span>                                <a name="line.518"></a>
<span class="sourceLineNo">519</span>                                String uri = (idPart.equals(uref.getId()) //- no tissue or celltype were attached<a name="line.519"></a>
<span class="sourceLineNo">520</span>                                                &amp;&amp; idPart.matches("^\\d+$")) //- is positive integer id<a name="line.520"></a>
<span class="sourceLineNo">521</span>                                        ? uri(xmlBase, uref.getDb(), idPart, BioSource.class)<a name="line.521"></a>
<span class="sourceLineNo">522</span>                                                : "http://identifiers.org/taxonomy/" + idPart;<a name="line.522"></a>
<span class="sourceLineNo">523</span>                                        // the latter is intentionally invalid identifiers.org/taxonomy URI - good (and important) for merging<a name="line.523"></a>
<span class="sourceLineNo">524</span>                                map.put(bs, uri);<a name="line.524"></a>
<span class="sourceLineNo">525</span><a name="line.525"></a>
<span class="sourceLineNo">526</span>                        } else <a name="line.526"></a>
<span class="sourceLineNo">527</span>                                log.debug("Won't normalize BioSource" <a name="line.527"></a>
<span class="sourceLineNo">528</span>                                        + " : no taxonomy unification xref found in " + bs.getUri()<a name="line.528"></a>
<span class="sourceLineNo">529</span>                                        + ". " + description);<a name="line.529"></a>
<span class="sourceLineNo">530</span>                } <a name="line.530"></a>
<span class="sourceLineNo">531</span>                <a name="line.531"></a>
<span class="sourceLineNo">532</span>                map.doSubs();<a name="line.532"></a>
<span class="sourceLineNo">533</span>        }<a name="line.533"></a>
<span class="sourceLineNo">534</span><a name="line.534"></a>
<span class="sourceLineNo">535</span>        private void normalizeERs(Model model) {<a name="line.535"></a>
<span class="sourceLineNo">536</span>                <a name="line.536"></a>
<span class="sourceLineNo">537</span>                NormalizerMap map = new NormalizerMap(model);<a name="line.537"></a>
<span class="sourceLineNo">538</span>                <a name="line.538"></a>
<span class="sourceLineNo">539</span>                // process the rest of utility classes (selectively though)<a name="line.539"></a>
<span class="sourceLineNo">540</span>                for (EntityReference bpe : model.getObjects(EntityReference.class)) {<a name="line.540"></a>
<span class="sourceLineNo">541</span>                        <a name="line.541"></a>
<span class="sourceLineNo">542</span>                        //skip those with already normalized URIs<a name="line.542"></a>
<span class="sourceLineNo">543</span>                        if(bpe.getUri().startsWith("http://identifiers.org/")) {<a name="line.543"></a>
<span class="sourceLineNo">544</span>                                log.info("Skip already normalized: " + bpe.getUri());<a name="line.544"></a>
<span class="sourceLineNo">545</span>                                continue;<a name="line.545"></a>
<span class="sourceLineNo">546</span>                        }                       <a name="line.546"></a>
<span class="sourceLineNo">547</span>                        <a name="line.547"></a>
<span class="sourceLineNo">548</span>                        UnificationXref uref = findPreferredUnificationXref(bpe);<a name="line.548"></a>
<span class="sourceLineNo">549</span>                        if (uref != null) {<a name="line.549"></a>
<span class="sourceLineNo">550</span>                                // Create (with a new URI made from a unif. xref) <a name="line.550"></a>
<span class="sourceLineNo">551</span>                                // and save the replacement object, if possible, <a name="line.551"></a>
<span class="sourceLineNo">552</span>                                // but do not replace yet (will call doSubs later, for all).<a name="line.552"></a>
<span class="sourceLineNo">553</span>                                final String db = uref.getDb();<a name="line.553"></a>
<span class="sourceLineNo">554</span>                                final String id = uref.getId();<a name="line.554"></a>
<span class="sourceLineNo">555</span>                                // get the standard ID<a name="line.555"></a>
<span class="sourceLineNo">556</span>                                String uri = null;<a name="line.556"></a>
<span class="sourceLineNo">557</span>                                try { // make a new ID for the element<a name="line.557"></a>
<span class="sourceLineNo">558</span>                                        uri = MiriamLink.getIdentifiersOrgURI(db, id);<a name="line.558"></a>
<span class="sourceLineNo">559</span>                                } catch (Exception e) {<a name="line.559"></a>
<span class="sourceLineNo">560</span>                                        log.error("Cannot get a Miriam standard ID for " + bpe <a name="line.560"></a>
<span class="sourceLineNo">561</span>                                                        + " (" + bpe.getModelInterface().getSimpleName()<a name="line.561"></a>
<span class="sourceLineNo">562</span>                                                        + ") " + ", using " + db + ":" + id <a name="line.562"></a>
<span class="sourceLineNo">563</span>                                                        + ". " + e.getMessage());<a name="line.563"></a>
<span class="sourceLineNo">564</span>                                        return;<a name="line.564"></a>
<span class="sourceLineNo">565</span>                                }<a name="line.565"></a>
<span class="sourceLineNo">566</span><a name="line.566"></a>
<span class="sourceLineNo">567</span>                                if(uri != null) {<a name="line.567"></a>
<span class="sourceLineNo">568</span>                                        map.put(bpe, uri);<a name="line.568"></a>
<span class="sourceLineNo">569</span>                                }       <a name="line.569"></a>
<span class="sourceLineNo">570</span>                        } else<a name="line.570"></a>
<span class="sourceLineNo">571</span>                                log.info("Cannot normalize EntityReference: "<a name="line.571"></a>
<span class="sourceLineNo">572</span>                                        + "no unification xrefs found in " + bpe.getUri()<a name="line.572"></a>
<span class="sourceLineNo">573</span>                                        + ". " + description);<a name="line.573"></a>
<span class="sourceLineNo">574</span>                }<a name="line.574"></a>
<span class="sourceLineNo">575</span>                <a name="line.575"></a>
<span class="sourceLineNo">576</span>                // replace/update elements in the model<a name="line.576"></a>
<span class="sourceLineNo">577</span>                map.doSubs();<a name="line.577"></a>
<span class="sourceLineNo">578</span>        }<a name="line.578"></a>
<span class="sourceLineNo">579</span>        <a name="line.579"></a>
<span class="sourceLineNo">580</span>        <a name="line.580"></a>
<span class="sourceLineNo">581</span>        /**<a name="line.581"></a>
<span class="sourceLineNo">582</span>         * Auto-generates standard and other names for the datasource<a name="line.582"></a>
<span class="sourceLineNo">583</span>         * from either its ID (if URN) or one of its existing names (preferably - standard name)<a name="line.583"></a>
<span class="sourceLineNo">584</span>         * <a name="line.584"></a>
<span class="sourceLineNo">585</span>         * @param pro data source (BioPAX Provenance)<a name="line.585"></a>
<span class="sourceLineNo">586</span>         */<a name="line.586"></a>
<span class="sourceLineNo">587</span>        public static void autoName(Provenance pro) {<a name="line.587"></a>
<span class="sourceLineNo">588</span>                if(!(pro.getUri().startsWith("urn:miriam:") || pro.getUri().startsWith("http://identifiers.org/"))<a name="line.588"></a>
<span class="sourceLineNo">589</span>                                &amp;&amp; pro.getName().isEmpty()) {<a name="line.589"></a>
<span class="sourceLineNo">590</span>                        log.info("Skipping: cannot normalize Provenance: " + pro.getUri());<a name="line.590"></a>
<span class="sourceLineNo">591</span>                }<a name="line.591"></a>
<span class="sourceLineNo">592</span>                else { // i.e., 'name' is not empty or ID is the URN<a name="line.592"></a>
<span class="sourceLineNo">593</span>                        final SortedSet&lt;String&gt; names = new TreeSet&lt;String&gt;();<a name="line.593"></a>
<span class="sourceLineNo">594</span>                        <a name="line.594"></a>
<span class="sourceLineNo">595</span>                        String key = null;<a name="line.595"></a>
<span class="sourceLineNo">596</span>                        if(pro.getUri().startsWith("urn:miriam:") || pro.getUri().startsWith("http://identifiers.org/")) {<a name="line.596"></a>
<span class="sourceLineNo">597</span>                                key = pro.getUri();<a name="line.597"></a>
<span class="sourceLineNo">598</span>                        } else if (pro.getStandardName() != null) {<a name="line.598"></a>
<span class="sourceLineNo">599</span>                                key = pro.getStandardName();<a name="line.599"></a>
<span class="sourceLineNo">600</span>                        } else {<a name="line.600"></a>
<span class="sourceLineNo">601</span>                                key = pro.getDisplayName(); // can be null<a name="line.601"></a>
<span class="sourceLineNo">602</span>                        }<a name="line.602"></a>
<span class="sourceLineNo">603</span>                        <a name="line.603"></a>
<span class="sourceLineNo">604</span>                        if (key != null) {<a name="line.604"></a>
<span class="sourceLineNo">605</span>                                try {<a name="line.605"></a>
<span class="sourceLineNo">606</span>                                        names.addAll(Arrays.asList(MiriamLink.getNames(key)));<a name="line.606"></a>
<span class="sourceLineNo">607</span>                                        pro.setStandardName(MiriamLink.getName(key));<a name="line.607"></a>
<span class="sourceLineNo">608</span>                                        // get the datasource description<a name="line.608"></a>
<span class="sourceLineNo">609</span>                                        String description = MiriamLink.getDataTypeDef(pro.getStandardName());<a name="line.609"></a>
<span class="sourceLineNo">610</span>                                        pro.addComment(description);<a name="line.610"></a>
<span class="sourceLineNo">611</span>                                } catch (IllegalArgumentException e) {<a name="line.611"></a>
<span class="sourceLineNo">612</span>                                        // ignore (then, names is still empty...)<a name="line.612"></a>
<span class="sourceLineNo">613</span>                                }<a name="line.613"></a>
<span class="sourceLineNo">614</span>                        } <a name="line.614"></a>
<span class="sourceLineNo">615</span>                        <a name="line.615"></a>
<span class="sourceLineNo">616</span>                        // when the above failed (no match in Miriam), or key was null -<a name="line.616"></a>
<span class="sourceLineNo">617</span>                        if(names.isEmpty()) {<a name="line.617"></a>
<span class="sourceLineNo">618</span>                                // finally, trying to find all valid names for each existing one<a name="line.618"></a>
<span class="sourceLineNo">619</span>                                        for (String name : pro.getName()) {<a name="line.619"></a>
<span class="sourceLineNo">620</span>                                                try {<a name="line.620"></a>
<span class="sourceLineNo">621</span>                                                        names.addAll(Arrays.asList(MiriamLink.getNames(name)));<a name="line.621"></a>
<span class="sourceLineNo">622</span>                                                } catch (IllegalArgumentException e) {<a name="line.622"></a>
<span class="sourceLineNo">623</span>                                                        // ignore<a name="line.623"></a>
<span class="sourceLineNo">624</span>                                                }<a name="line.624"></a>
<span class="sourceLineNo">625</span>                                        }<a name="line.625"></a>
<span class="sourceLineNo">626</span>                                        // pick up the first name, get the standard name<a name="line.626"></a>
<span class="sourceLineNo">627</span>                                        if(!names.isEmpty())<a name="line.627"></a>
<span class="sourceLineNo">628</span>                                                pro.setStandardName(MiriamLink<a name="line.628"></a>
<span class="sourceLineNo">629</span>                                                        .getName(names.iterator().next()));<a name="line.629"></a>
<span class="sourceLineNo">630</span>                        }<a name="line.630"></a>
<span class="sourceLineNo">631</span>                        <a name="line.631"></a>
<span class="sourceLineNo">632</span>                        // and add all the synonyms if any<a name="line.632"></a>
<span class="sourceLineNo">633</span>                        for(String name : names)<a name="line.633"></a>
<span class="sourceLineNo">634</span>                                pro.addName(name);<a name="line.634"></a>
<span class="sourceLineNo">635</span>                        <a name="line.635"></a>
<span class="sourceLineNo">636</span>                        //set display name if not set (standard name is set already)<a name="line.636"></a>
<span class="sourceLineNo">637</span>                        if(pro.getDisplayName() == null)<a name="line.637"></a>
<span class="sourceLineNo">638</span>                                pro.setDisplayName(pro.getStandardName());                      <a name="line.638"></a>
<span class="sourceLineNo">639</span>                }<a name="line.639"></a>
<span class="sourceLineNo">640</span>        }<a name="line.640"></a>
<span class="sourceLineNo">641</span>        <a name="line.641"></a>
<span class="sourceLineNo">642</span>        /**<a name="line.642"></a>
<span class="sourceLineNo">643</span>         * Converts BioPAX L1 or L2 RDF/XML string data to BioPAX L3 string.<a name="line.643"></a>
<span class="sourceLineNo">644</span>         *<a name="line.644"></a>
<span class="sourceLineNo">645</span>         * WARN: this is not for huge (larger than 1GB) BioPAX RDF/XML data<a name="line.645"></a>
<span class="sourceLineNo">646</span>         * due to use of (UTF-8) String and Byte Array internally.<a name="line.646"></a>
<span class="sourceLineNo">647</span>         * This can be and is used by online web apps, such as the<a name="line.647"></a>
<span class="sourceLineNo">648</span>         * BioPAX Validator.<a name="line.648"></a>
<span class="sourceLineNo">649</span>         *<a name="line.649"></a>
<span class="sourceLineNo">650</span>         * @param biopaxData String<a name="line.650"></a>
<span class="sourceLineNo">651</span>         * @return BioPAX Level3 RDF/XML string<a name="line.651"></a>
<span class="sourceLineNo">652</span>         */<a name="line.652"></a>
<span class="sourceLineNo">653</span>        public static String convertToLevel3(final String biopaxData) {<a name="line.653"></a>
<span class="sourceLineNo">654</span>                String toReturn = "";<a name="line.654"></a>
<span class="sourceLineNo">655</span>                <a name="line.655"></a>
<span class="sourceLineNo">656</span>                try {<a name="line.656"></a>
<span class="sourceLineNo">657</span>                        ByteArrayOutputStream os = new ByteArrayOutputStream();<a name="line.657"></a>
<span class="sourceLineNo">658</span>                        InputStream is = new ByteArrayInputStream(biopaxData.getBytes());<a name="line.658"></a>
<span class="sourceLineNo">659</span>                        SimpleIOHandler io = new SimpleIOHandler();<a name="line.659"></a>
<span class="sourceLineNo">660</span>                        io.mergeDuplicates(true);<a name="line.660"></a>
<span class="sourceLineNo">661</span>                        Model model = io.convertFromOWL(is);<a name="line.661"></a>
<span class="sourceLineNo">662</span>                        if (model.getLevel() != BioPAXLevel.L3) {<a name="line.662"></a>
<span class="sourceLineNo">663</span>                                log.info("Converting to BioPAX Level3... " + model.getXmlBase());<a name="line.663"></a>
<span class="sourceLineNo">664</span>                                model = (new LevelUpgrader()).filter(model);<a name="line.664"></a>
<span class="sourceLineNo">665</span>                                if (model != null) {<a name="line.665"></a>
<span class="sourceLineNo">666</span>                                        io.setFactory(model.getLevel().getDefaultFactory());<a name="line.666"></a>
<span class="sourceLineNo">667</span>                                        io.convertToOWL(model, os);<a name="line.667"></a>
<span class="sourceLineNo">668</span>                                        toReturn = os.toString();<a name="line.668"></a>
<span class="sourceLineNo">669</span>                                }<a name="line.669"></a>
<span class="sourceLineNo">670</span>                        } else {<a name="line.670"></a>
<span class="sourceLineNo">671</span>                                toReturn = biopaxData;<a name="line.671"></a>
<span class="sourceLineNo">672</span>                        }<a name="line.672"></a>
<span class="sourceLineNo">673</span>                } catch(Exception e) {<a name="line.673"></a>
<span class="sourceLineNo">674</span>                        throw new RuntimeException(<a name="line.674"></a>
<span class="sourceLineNo">675</span>                                "Cannot convert to BioPAX Level3", e);<a name="line.675"></a>
<span class="sourceLineNo">676</span>                }<a name="line.676"></a>
<span class="sourceLineNo">677</span><a name="line.677"></a>
<span class="sourceLineNo">678</span>                return toReturn;<a name="line.678"></a>
<span class="sourceLineNo">679</span>        }<a name="line.679"></a>
<span class="sourceLineNo">680</span>        <a name="line.680"></a>
<span class="sourceLineNo">681</span>        <a name="line.681"></a>
<span class="sourceLineNo">682</span>        /**<a name="line.682"></a>
<span class="sourceLineNo">683</span>         * Gets the xml:base to use with newly created BioPAX elements.<a name="line.683"></a>
<span class="sourceLineNo">684</span>         * <a name="line.684"></a>
<span class="sourceLineNo">685</span>         * @param modelToNormalize<a name="line.685"></a>
<span class="sourceLineNo">686</span>         * @return BioPAX model's xml:base or empty string if it's null<a name="line.686"></a>
<span class="sourceLineNo">687</span>         */<a name="line.687"></a>
<span class="sourceLineNo">688</span>        private String getXmlBase(Model modelToNormalize) {<a name="line.688"></a>
<span class="sourceLineNo">689</span>                if(xmlBase != null)<a name="line.689"></a>
<span class="sourceLineNo">690</span>                        return xmlBase;<a name="line.690"></a>
<span class="sourceLineNo">691</span>                else<a name="line.691"></a>
<span class="sourceLineNo">692</span>                        return "";<a name="line.692"></a>
<span class="sourceLineNo">693</span>        }<a name="line.693"></a>
<span class="sourceLineNo">694</span>        <a name="line.694"></a>
<span class="sourceLineNo">695</span>        <a name="line.695"></a>
<span class="sourceLineNo">696</span>        public boolean isFixDisplayName() {<a name="line.696"></a>
<span class="sourceLineNo">697</span>                return fixDisplayName;<a name="line.697"></a>
<span class="sourceLineNo">698</span>        }<a name="line.698"></a>
<span class="sourceLineNo">699</span>        public void setFixDisplayName(boolean fixDisplayName) {<a name="line.699"></a>
<span class="sourceLineNo">700</span>                this.fixDisplayName = fixDisplayName;<a name="line.700"></a>
<span class="sourceLineNo">701</span>        }<a name="line.701"></a>
<span class="sourceLineNo">702</span><a name="line.702"></a>
<span class="sourceLineNo">703</span>        public String getXmlBase() {<a name="line.703"></a>
<span class="sourceLineNo">704</span>                return xmlBase;<a name="line.704"></a>
<span class="sourceLineNo">705</span>        }<a name="line.705"></a>
<span class="sourceLineNo">706</span>        public void setXmlBase(String xmlBase) {<a name="line.706"></a>
<span class="sourceLineNo">707</span>                this.xmlBase = xmlBase;<a name="line.707"></a>
<span class="sourceLineNo">708</span>        }<a name="line.708"></a>
<span class="sourceLineNo">709</span><a name="line.709"></a>
<span class="sourceLineNo">710</span>        <a name="line.710"></a>
<span class="sourceLineNo">711</span>        /**<a name="line.711"></a>
<span class="sourceLineNo">712</span>         * Helper class, to associate original <a name="line.712"></a>
<span class="sourceLineNo">713</span>         * and new (either copy or existing) objects <a name="line.713"></a>
<span class="sourceLineNo">714</span>         * within some biopax model and then execute <a name="line.714"></a>
<span class="sourceLineNo">715</span>         * batch replace.<a name="line.715"></a>
<span class="sourceLineNo">716</span>         * <a name="line.716"></a>
<span class="sourceLineNo">717</span>         * @author rodche<a name="line.717"></a>
<span class="sourceLineNo">718</span>         */<a name="line.718"></a>
<span class="sourceLineNo">719</span>        private static class NormalizerMap {<a name="line.719"></a>
<span class="sourceLineNo">720</span><a name="line.720"></a>
<span class="sourceLineNo">721</span>                //a model to modify by replacing biopax objects<a name="line.721"></a>
<span class="sourceLineNo">722</span>                final Model model;<a name="line.722"></a>
<span class="sourceLineNo">723</span>                <a name="line.723"></a>
<span class="sourceLineNo">724</span>                // this is the biopax elements substitution map (old-&gt;new)<a name="line.724"></a>
<span class="sourceLineNo">725</span>                final Map&lt;BioPAXElement,BioPAXElement&gt; subs;<a name="line.725"></a>
<span class="sourceLineNo">726</span>                <a name="line.726"></a>
<span class="sourceLineNo">727</span>                //the next map is to make sure the subs.values()  all have different URIs<a name="line.727"></a>
<span class="sourceLineNo">728</span>                final Map&lt;String,BioPAXElement&gt; uriToSub;<a name="line.728"></a>
<span class="sourceLineNo">729</span>                <a name="line.729"></a>
<span class="sourceLineNo">730</span>                final ShallowCopy copier;<a name="line.730"></a>
<span class="sourceLineNo">731</span>                                <a name="line.731"></a>
<span class="sourceLineNo">732</span>                NormalizerMap(Model model) {<a name="line.732"></a>
<span class="sourceLineNo">733</span>                        subs = BPCollections.I.createMap();<a name="line.733"></a>
<span class="sourceLineNo">734</span>                        uriToSub = BPCollections.I.createMap();<a name="line.734"></a>
<span class="sourceLineNo">735</span>                        this.model = model;<a name="line.735"></a>
<span class="sourceLineNo">736</span>                        copier = new ShallowCopy();<a name="line.736"></a>
<span class="sourceLineNo">737</span>                }<a name="line.737"></a>
<span class="sourceLineNo">738</span><a name="line.738"></a>
<span class="sourceLineNo">739</span><a name="line.739"></a>
<span class="sourceLineNo">740</span>                /**<a name="line.740"></a>
<span class="sourceLineNo">741</span>                 * Creates (by URI) and saves the replacement object,<a name="line.741"></a>
<span class="sourceLineNo">742</span>                 * but does not replace yet (call doSubs to replace).<a name="line.742"></a>
<span class="sourceLineNo">743</span>                 * <a name="line.743"></a>
<span class="sourceLineNo">744</span>                 * @param bpe<a name="line.744"></a>
<span class="sourceLineNo">745</span>                 * @param newUri<a name="line.745"></a>
<span class="sourceLineNo">746</span>                 */<a name="line.746"></a>
<span class="sourceLineNo">747</span>                void put(BioPAXElement bpe, String newUri)<a name="line.747"></a>
<span class="sourceLineNo">748</span>                {<a name="line.748"></a>
<span class="sourceLineNo">749</span>                        if(model.containsID(newUri)) {<a name="line.749"></a>
<span class="sourceLineNo">750</span>                                // will use existing original (model) object that has the new Uri<a name="line.750"></a>
<span class="sourceLineNo">751</span>                                map(bpe, model.getByID(newUri));<a name="line.751"></a>
<span class="sourceLineNo">752</span>                        } else if(uriToSub.containsKey(newUri)) {<a name="line.752"></a>
<span class="sourceLineNo">753</span>                                // re-use the new object that's already added to replace another original<a name="line.753"></a>
<span class="sourceLineNo">754</span>                                map(bpe, uriToSub.get(newUri));<a name="line.754"></a>
<span class="sourceLineNo">755</span>                        } else {<a name="line.755"></a>
<span class="sourceLineNo">756</span>                                // will use the object's shallow copy that gets new Uri<a name="line.756"></a>
<span class="sourceLineNo">757</span>                                BioPAXElement copy = copier.copy(bpe, newUri);<a name="line.757"></a>
<span class="sourceLineNo">758</span>                                map(bpe, copy);<a name="line.758"></a>
<span class="sourceLineNo">759</span>                        }<a name="line.759"></a>
<span class="sourceLineNo">760</span>                }               <a name="line.760"></a>
<span class="sourceLineNo">761</span>                <a name="line.761"></a>
<span class="sourceLineNo">762</span>                /**<a name="line.762"></a>
<span class="sourceLineNo">763</span>                 * Executes the batch replace - migrating  <a name="line.763"></a>
<span class="sourceLineNo">764</span>                 * to the normalized equivalent objects.<a name="line.764"></a>
<span class="sourceLineNo">765</span>                 */<a name="line.765"></a>
<span class="sourceLineNo">766</span>                void doSubs() {<a name="line.766"></a>
<span class="sourceLineNo">767</span>                        for(BioPAXElement e : subs.keySet()) {<a name="line.767"></a>
<span class="sourceLineNo">768</span>                                model.remove(e);<a name="line.768"></a>
<span class="sourceLineNo">769</span>                        }<a name="line.769"></a>
<span class="sourceLineNo">770</span>                        <a name="line.770"></a>
<span class="sourceLineNo">771</span>                        try {<a name="line.771"></a>
<span class="sourceLineNo">772</span>                                ModelUtils.replace(model, subs);<a name="line.772"></a>
<span class="sourceLineNo">773</span>                        } catch (Exception e) {<a name="line.773"></a>
<span class="sourceLineNo">774</span>                                log.error("Failed to replace BioPAX elements.", e);<a name="line.774"></a>
<span class="sourceLineNo">775</span>                                return;<a name="line.775"></a>
<span class="sourceLineNo">776</span>                        }<a name="line.776"></a>
<span class="sourceLineNo">777</span>                        <a name="line.777"></a>
<span class="sourceLineNo">778</span>                        for(BioPAXElement e : subs.values()) {<a name="line.778"></a>
<span class="sourceLineNo">779</span>                                if(!model.contains(e))<a name="line.779"></a>
<span class="sourceLineNo">780</span>                                        model.add(e);<a name="line.780"></a>
<span class="sourceLineNo">781</span>                        }<a name="line.781"></a>
<span class="sourceLineNo">782</span>                <a name="line.782"></a>
<span class="sourceLineNo">783</span>                        for(BioPAXElement e : model.getObjects()) {<a name="line.783"></a>
<span class="sourceLineNo">784</span>                                ModelUtils.fixDanglingInverseProperties(e, model);<a name="line.784"></a>
<span class="sourceLineNo">785</span>                        }<a name="line.785"></a>
<span class="sourceLineNo">786</span>                }<a name="line.786"></a>
<span class="sourceLineNo">787</span><a name="line.787"></a>
<span class="sourceLineNo">788</span>                private void map(BioPAXElement bpe, BioPAXElement newBpe) {<a name="line.788"></a>
<span class="sourceLineNo">789</span>                        subs.put(bpe, newBpe);<a name="line.789"></a>
<span class="sourceLineNo">790</span>                        uriToSub.put(newBpe.getUri(), newBpe);<a name="line.790"></a>
<span class="sourceLineNo">791</span>                }<a name="line.791"></a>
<span class="sourceLineNo">792</span>        }<a name="line.792"></a>
<span class="sourceLineNo">793</span>        <a name="line.793"></a>
<span class="sourceLineNo">794</span>}<a name="line.794"></a>




























































</pre>
</div>
</body>
</html>
