<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head><meta http-equiv="content-type" content="text/html; charset=UTF-8" />
<title>EntityReference xref</title>
<link type="text/css" rel="stylesheet" href="../../../../../stylesheet.css" />
</head>
<body>
<div id="overview"><a href="../../../../../../apidocs/org/biopax/paxtools/model/level3/EntityReference.html">View Javadoc</a></div><pre>
<a class="jxr_linenumber" name="L1" href="#L1">1</a>   <strong class="jxr_keyword">package</strong> org.biopax.paxtools.model.level3;
<a class="jxr_linenumber" name="L2" href="#L2">2</a>   
<a class="jxr_linenumber" name="L3" href="#L3">3</a>   <strong class="jxr_keyword">import</strong> org.biopax.paxtools.util.AutoComplete;
<a class="jxr_linenumber" name="L4" href="#L4">4</a>   
<a class="jxr_linenumber" name="L5" href="#L5">5</a>   <strong class="jxr_keyword">import</strong> java.util.Set;
<a class="jxr_linenumber" name="L6" href="#L6">6</a>   
<a class="jxr_linenumber" name="L7" href="#L7">7</a>   <em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L8" href="#L8">8</a>   <em class="jxr_javadoccomment"> * Definition: An entity reference is a grouping of several physical entities across different</em>
<a class="jxr_linenumber" name="L9" href="#L9">9</a>   <em class="jxr_javadoccomment"> * contexts and molecular states, that share common physical properties and often named and treated</em>
<a class="jxr_linenumber" name="L10" href="#L10">10</a>  <em class="jxr_javadoccomment"> * as a single entity with multiple states by biologists.</em>
<a class="jxr_linenumber" name="L11" href="#L11">11</a>  <em class="jxr_javadoccomment"> *</em>
<a class="jxr_linenumber" name="L12" href="#L12">12</a>  <em class="jxr_javadoccomment"> * Rationale:   Many protein, small molecule and gene databases share this point of view, and such a</em>
<a class="jxr_linenumber" name="L13" href="#L13">13</a>  <em class="jxr_javadoccomment"> * grouping is an important prerequisite for interoperability with those databases. Biologists would</em>
<a class="jxr_linenumber" name="L14" href="#L14">14</a>  <em class="jxr_javadoccomment"> * often group different pools of molecules in different contexts under the same name. For example</em>
<a class="jxr_linenumber" name="L15" href="#L15">15</a>  <em class="jxr_javadoccomment"> * cytoplasmic and extracellular calcium have different effects on the cell's behavior, but they are</em>
<a class="jxr_linenumber" name="L16" href="#L16">16</a>  <em class="jxr_javadoccomment"> * still called calcium. For DNA, RNA and Proteins the grouping is defined based on a wildtype</em>
<a class="jxr_linenumber" name="L17" href="#L17">17</a>  <em class="jxr_javadoccomment"> * sequence, for small molecules it is defined by the chemical structure.</em>
<a class="jxr_linenumber" name="L18" href="#L18">18</a>  <em class="jxr_javadoccomment"> *</em>
<a class="jxr_linenumber" name="L19" href="#L19">19</a>  <em class="jxr_javadoccomment"> * Usage: Entity references store the information common to a set of molecules in various states</em>
<a class="jxr_linenumber" name="L20" href="#L20">20</a>  <em class="jxr_javadoccomment"> * described in the BioPAX document, including database cross-references. For instance, the P53</em>
<a class="jxr_linenumber" name="L21" href="#L21">21</a>  <em class="jxr_javadoccomment"> * protein can be phosphorylated in multiple different ways. Each separate P53 protein (pool) in a</em>
<a class="jxr_linenumber" name="L22" href="#L22">22</a>  <em class="jxr_javadoccomment"> * phosphorylation state would be represented as a different protein (child of physicalEntity) and</em>
<a class="jxr_linenumber" name="L23" href="#L23">23</a>  <em class="jxr_javadoccomment"> * all things common to all P53 proteins, including all possible phosphorylation sites, the sequence</em>
<a class="jxr_linenumber" name="L24" href="#L24">24</a>  <em class="jxr_javadoccomment"> * common to all of them and common references to protein databases containing more information</em>
<a class="jxr_linenumber" name="L25" href="#L25">25</a>  <em class="jxr_javadoccomment"> * about P53 would be stored in a Entity Reference.</em>
<a class="jxr_linenumber" name="L26" href="#L26">26</a>  <em class="jxr_javadoccomment"> *</em>
<a class="jxr_linenumber" name="L27" href="#L27">27</a>  <em class="jxr_javadoccomment"> * Comments: This grouping has three semantic implications: &lt;ol&gt; &lt;li&gt;Members of different pools</em>
<a class="jxr_linenumber" name="L28" href="#L28">28</a>  <em class="jxr_javadoccomment"> * share many physical and biochemical properties. This includes their chemical structure, sequence,</em>
<a class="jxr_linenumber" name="L29" href="#L29">29</a>  <em class="jxr_javadoccomment"> * organism and set of molecules they react with. They will also share a lot of secondary</em>
<a class="jxr_linenumber" name="L30" href="#L30">30</a>  <em class="jxr_javadoccomment"> * information such as their names, functional groupings, annotation terms and database identifiers.</em>
<a class="jxr_linenumber" name="L31" href="#L31">31</a>  <em class="jxr_javadoccomment"> * &lt;li&gt; A small number of transitions seperates these pools. In other words it is relatively easy</em>
<a class="jxr_linenumber" name="L32" href="#L32">32</a>  <em class="jxr_javadoccomment"> * and frequent for a molecule to transform from one physical entity to another that belong to the</em>
<a class="jxr_linenumber" name="L33" href="#L33">33</a>  <em class="jxr_javadoccomment"> * same reference entity. For example an extracellular calcium can become cytoplasmic, and p53 can</em>
<a class="jxr_linenumber" name="L34" href="#L34">34</a>  <em class="jxr_javadoccomment"> * become phosphorylated. However no calcium virtually becomes sodium, or no p53 becomes mdm2. In</em>
<a class="jxr_linenumber" name="L35" href="#L35">35</a>  <em class="jxr_javadoccomment"> * the former it is the sheer energy barrier of a nuclear reaction, in the latter sheer statistical</em>
<a class="jxr_linenumber" name="L36" href="#L36">36</a>  <em class="jxr_javadoccomment"> * improbability of synthesizing the same sequence without a template. If one thinks about the</em>
<a class="jxr_linenumber" name="L37" href="#L37">37</a>  <em class="jxr_javadoccomment"> * biochemical network as molecules transforming into each other, and remove edges that respond to</em>
<a class="jxr_linenumber" name="L38" href="#L38">38</a>  <em class="jxr_javadoccomment"> * transcription, translation, degradation and covalent modification of small molecules, each</em>
<a class="jxr_linenumber" name="L39" href="#L39">39</a>  <em class="jxr_javadoccomment"> * remaining component is a reference entity. &lt;li&gt; Some of the pools in the same group can overlap.</em>
<a class="jxr_linenumber" name="L40" href="#L40">40</a>  <em class="jxr_javadoccomment"> * p53-p@ser15 can overlap with p53-p@thr18. Most of the experiments in molecular biology will only</em>
<a class="jxr_linenumber" name="L41" href="#L41">41</a>  <em class="jxr_javadoccomment"> * check for one state variable, rarely multiple, and never for the all possible combinations. So</em>
<a class="jxr_linenumber" name="L42" href="#L42">42</a>  <em class="jxr_javadoccomment"> * almost all statements that refer to the state of the molecule talk about a pool that can overlap</em>
<a class="jxr_linenumber" name="L43" href="#L43">43</a>  <em class="jxr_javadoccomment"> * with other pools. However no overlaps is possible between molecules of different groups. &lt;/ol&gt;</em>
<a class="jxr_linenumber" name="L44" href="#L44">44</a>  <em class="jxr_javadoccomment"> */</em>
<a class="jxr_linenumber" name="L45" href="#L45">45</a>  
<a class="jxr_linenumber" name="L46" href="#L46">46</a>  <strong class="jxr_keyword">public</strong> <strong class="jxr_keyword">interface</strong> <a name="EntityReference" href="../../../../../org/biopax/paxtools/model/level3/EntityReference.html#EntityReference">EntityReference</a> <strong class="jxr_keyword">extends</strong> UtilityClass, Named, <a name="Observable" href="../../../../../org/biopax/paxtools/model/level3/Observable.html#Observable">Observable</a>
<a class="jxr_linenumber" name="L47" href="#L47">47</a>  {
<a class="jxr_linenumber" name="L48" href="#L48">48</a>  
<a class="jxr_linenumber" name="L49" href="#L49">49</a>  
<a class="jxr_linenumber" name="L50" href="#L50">50</a>  	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L51" href="#L51">51</a>  <em class="jxr_javadoccomment">	 * Variable features that are observed for the entities of this entityReference - such as known PTM</em>
<a class="jxr_linenumber" name="L52" href="#L52">52</a>  <em class="jxr_javadoccomment">	 * or methylation sites and non-covalent bonds. Note that this is an aggregate list of all known</em>
<a class="jxr_linenumber" name="L53" href="#L53">53</a>  <em class="jxr_javadoccomment">	 * features and it does not represent a state itself.</em>
<a class="jxr_linenumber" name="L54" href="#L54">54</a>  <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L55" href="#L55">55</a>  <em class="jxr_javadoccomment">	 * @return a set of entityFeatures</em>
<a class="jxr_linenumber" name="L56" href="#L56">56</a>  <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L57" href="#L57">57</a>  	@AutoComplete(backward = <strong class="jxr_keyword">true</strong>)
<a class="jxr_linenumber" name="L58" href="#L58">58</a>  	Set&lt;EntityFeature&gt; getEntityFeature();
<a class="jxr_linenumber" name="L59" href="#L59">59</a>  
<a class="jxr_linenumber" name="L60" href="#L60">60</a>  	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L61" href="#L61">61</a>  <em class="jxr_javadoccomment">	 * Variable features that are observed for the entities of this entityReference - such as known PTM</em>
<a class="jxr_linenumber" name="L62" href="#L62">62</a>  <em class="jxr_javadoccomment">	 * or methylation sites and non-covalent bonds. Note that this is an aggregate list of all known</em>
<a class="jxr_linenumber" name="L63" href="#L63">63</a>  <em class="jxr_javadoccomment">	 * features and it does not represent a state itself.</em>
<a class="jxr_linenumber" name="L64" href="#L64">64</a>  <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L65" href="#L65">65</a>  <em class="jxr_javadoccomment">	 * @param feature to be added.</em>
<a class="jxr_linenumber" name="L66" href="#L66">66</a>  <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L67" href="#L67">67</a>  	<strong class="jxr_keyword">void</strong> addEntityFeature(<a name="EntityFeature" href="../../../../../org/biopax/paxtools/model/level3/EntityFeature.html#EntityFeature">EntityFeature</a> feature);
<a class="jxr_linenumber" name="L68" href="#L68">68</a>  
<a class="jxr_linenumber" name="L69" href="#L69">69</a>  	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L70" href="#L70">70</a>  <em class="jxr_javadoccomment">	 * Variable features that are observed for the entities of this entityReference - such as known PTM</em>
<a class="jxr_linenumber" name="L71" href="#L71">71</a>  <em class="jxr_javadoccomment">	 * or methylation sites and non-covalent bonds. Note that this is an aggregate list of all known</em>
<a class="jxr_linenumber" name="L72" href="#L72">72</a>  <em class="jxr_javadoccomment">	 * features and it does not represent a state itself.</em>
<a class="jxr_linenumber" name="L73" href="#L73">73</a>  <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L74" href="#L74">74</a>  <em class="jxr_javadoccomment">	 * @param feature to be removed.</em>
<a class="jxr_linenumber" name="L75" href="#L75">75</a>  <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L76" href="#L76">76</a>  	<strong class="jxr_keyword">void</strong> removeEntityFeature(<a name="EntityFeature" href="../../../../../org/biopax/paxtools/model/level3/EntityFeature.html#EntityFeature">EntityFeature</a> feature);
<a class="jxr_linenumber" name="L77" href="#L77">77</a>  
<a class="jxr_linenumber" name="L78" href="#L78">78</a>  
<a class="jxr_linenumber" name="L79" href="#L79">79</a>  	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L80" href="#L80">80</a>  <em class="jxr_javadoccomment">	 * Inverse of {@link SimplePhysicalEntity#getEntityReference()}</em>
<a class="jxr_linenumber" name="L81" href="#L81">81</a>  <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L82" href="#L82">82</a>  <em class="jxr_javadoccomment">	 * @return a  set of {@link SimplePhysicalEntity} that has this EntityReference</em>
<a class="jxr_linenumber" name="L83" href="#L83">83</a>  <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L84" href="#L84">84</a>  	Set&lt;SimplePhysicalEntity&gt; getEntityReferenceOf();
<a class="jxr_linenumber" name="L85" href="#L85">85</a>  
<a class="jxr_linenumber" name="L86" href="#L86">86</a>  
<a class="jxr_linenumber" name="L87" href="#L87">87</a>  	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L88" href="#L88">88</a>  <em class="jxr_javadoccomment">	 * @return Controlled vocabulary terms that is used to describe the type of grouping such as</em>
<a class="jxr_linenumber" name="L89" href="#L89">89</a>  <em class="jxr_javadoccomment">	 *         homology or functional group.</em>
<a class="jxr_linenumber" name="L90" href="#L90">90</a>  <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L91" href="#L91">91</a>  	Set&lt;EntityReferenceTypeVocabulary&gt; getEntityReferenceType();
<a class="jxr_linenumber" name="L92" href="#L92">92</a>  
<a class="jxr_linenumber" name="L93" href="#L93">93</a>  	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L94" href="#L94">94</a>  <em class="jxr_javadoccomment">	 * Adds the given cv to the list of types</em>
<a class="jxr_linenumber" name="L95" href="#L95">95</a>  <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L96" href="#L96">96</a>  <em class="jxr_javadoccomment">	 * @param type A controlled vocabulary term that is used to describe the type of grouping such as</em>
<a class="jxr_linenumber" name="L97" href="#L97">97</a>  <em class="jxr_javadoccomment">	 *             homology or functional group.</em>
<a class="jxr_linenumber" name="L98" href="#L98">98</a>  <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L99" href="#L99">99</a>  	<strong class="jxr_keyword">void</strong> addEntityReferenceType(<a name="EntityReferenceTypeVocabulary" href="../../../../../org/biopax/paxtools/model/level3/EntityReferenceTypeVocabulary.html#EntityReferenceTypeVocabulary">EntityReferenceTypeVocabulary</a> type);
<a class="jxr_linenumber" name="L100" href="#L100">100</a> 
<a class="jxr_linenumber" name="L101" href="#L101">101</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L102" href="#L102">102</a> <em class="jxr_javadoccomment">	 * Removes the given cv from the list of types</em>
<a class="jxr_linenumber" name="L103" href="#L103">103</a> <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L104" href="#L104">104</a> <em class="jxr_javadoccomment">	 * @param type A controlled vocabulary term that is used to describe the type of grouping such as</em>
<a class="jxr_linenumber" name="L105" href="#L105">105</a> <em class="jxr_javadoccomment">	 *             homology or functional group.</em>
<a class="jxr_linenumber" name="L106" href="#L106">106</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L107" href="#L107">107</a> 	<strong class="jxr_keyword">void</strong> removeEntityReferenceType(<a name="EntityReferenceTypeVocabulary" href="../../../../../org/biopax/paxtools/model/level3/EntityReferenceTypeVocabulary.html#EntityReferenceTypeVocabulary">EntityReferenceTypeVocabulary</a> type);
<a class="jxr_linenumber" name="L108" href="#L108">108</a> 
<a class="jxr_linenumber" name="L109" href="#L109">109</a> 
<a class="jxr_linenumber" name="L110" href="#L110">110</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L111" href="#L111">111</a> <em class="jxr_javadoccomment">	 * @return Entity references that qualifies for the definition of this group. For example a member</em>
<a class="jxr_linenumber" name="L112" href="#L112">112</a> <em class="jxr_javadoccomment">	 *         of a PFAM protein family.</em>
<a class="jxr_linenumber" name="L113" href="#L113">113</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L114" href="#L114">114</a> 	Set&lt;EntityReference&gt; getMemberEntityReference();  <em class="jxr_comment">//todo generify?</em>
<a class="jxr_linenumber" name="L115" href="#L115">115</a> 
<a class="jxr_linenumber" name="L116" href="#L116">116</a> <em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L117" href="#L117">117</a> <em class="jxr_javadoccomment">	 * Adds the given entityReference to the member list</em>
<a class="jxr_linenumber" name="L118" href="#L118">118</a> <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L119" href="#L119">119</a> <em class="jxr_javadoccomment">	 * @param entityReference An entity reference that qualifies for the definition of this group. For</em>
<a class="jxr_linenumber" name="L120" href="#L120">120</a> <em class="jxr_javadoccomment">	 *                        example a member of a PFAM protein family.</em>
<a class="jxr_linenumber" name="L121" href="#L121">121</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L122" href="#L122">122</a> 	<strong class="jxr_keyword">void</strong> addMemberEntityReference(<a name="EntityReference" href="../../../../../org/biopax/paxtools/model/level3/EntityReference.html#EntityReference">EntityReference</a> entityReference);
<a class="jxr_linenumber" name="L123" href="#L123">123</a> 
<a class="jxr_linenumber" name="L124" href="#L124">124</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L125" href="#L125">125</a> <em class="jxr_javadoccomment">	 * Removes the given entityReference from the member list</em>
<a class="jxr_linenumber" name="L126" href="#L126">126</a> <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L127" href="#L127">127</a> <em class="jxr_javadoccomment">	 * @param entityReference An entity reference that qualifies for the definition of this group. For</em>
<a class="jxr_linenumber" name="L128" href="#L128">128</a> <em class="jxr_javadoccomment">	 *                        example a member of a PFAM protein family.</em>
<a class="jxr_linenumber" name="L129" href="#L129">129</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L130" href="#L130">130</a> 	<strong class="jxr_keyword">void</strong> removeMemberEntityReference(<a name="EntityReference" href="../../../../../org/biopax/paxtools/model/level3/EntityReference.html#EntityReference">EntityReference</a> entityReference);
<a class="jxr_linenumber" name="L131" href="#L131">131</a> 
<a class="jxr_linenumber" name="L132" href="#L132">132</a> 	<em class="jxr_javadoccomment">/**</em>
<a class="jxr_linenumber" name="L133" href="#L133">133</a> <em class="jxr_javadoccomment">	 * Reverse of {@link #getMemberEntityReference()}</em>
<a class="jxr_linenumber" name="L134" href="#L134">134</a> <em class="jxr_javadoccomment">	 *</em>
<a class="jxr_linenumber" name="L135" href="#L135">135</a> <em class="jxr_javadoccomment">	 * @return EntityReferences that this EntityReference is a member of.</em>
<a class="jxr_linenumber" name="L136" href="#L136">136</a> <em class="jxr_javadoccomment">	 */</em>
<a class="jxr_linenumber" name="L137" href="#L137">137</a> 	<strong class="jxr_keyword">public</strong> Set&lt;EntityReference&gt; getMemberEntityReferenceOf();
<a class="jxr_linenumber" name="L138" href="#L138">138</a> 
<a class="jxr_linenumber" name="L139" href="#L139">139</a> }
</pre>
<hr/>
<div id="footer">Copyright &#169; 2023 <a href="http://www.biopax.org">BioPAX</a>. All rights reserved.</div>
</body>
</html>
